// @git-diff-view/lowlight/dist/esm/index.mjs
import { createLowlight, all } from 'lowlight';

const processAST = (ast) => {
    let lineNumber = 1;
    const syntaxObj = {};
    const loopAST = (nodes, wrapper) => {
        nodes.forEach((node) => {
            if (node.type === "text") {
                if (node.value.indexOf("\n") === -1) {
                    const valueLength = node.value.length;
                    if (!syntaxObj[lineNumber]) {
                        node.startIndex = 0;
                        node.endIndex = valueLength - 1;
                        const item = {
                            value: node.value,
                            lineNumber,
                            valueLength,
                            nodeList: [{ node, wrapper }],
                        };
                        syntaxObj[lineNumber] = item;
                    }
                    else {
                        node.startIndex = syntaxObj[lineNumber].valueLength;
                        node.endIndex = node.startIndex + valueLength - 1;
                        syntaxObj[lineNumber].value += node.value;
                        syntaxObj[lineNumber].valueLength += valueLength;
                        syntaxObj[lineNumber].nodeList.push({ node, wrapper });
                    }
                    node.lineNumber = lineNumber;
                    return;
                }
                const lines = node.value.split("\n");
                node.children = node.children || [];
                for (let i = 0; i < lines.length; i++) {
                    const _value = i === lines.length - 1 ? lines[i] : lines[i] + "\n";
                    const _lineNumber = i === 0 ? lineNumber : ++lineNumber;
                    const _valueLength = _value.length;
                    const _node = {
                        type: "text",
                        value: _value,
                        startIndex: Infinity,
                        endIndex: Infinity,
                        lineNumber: _lineNumber,
                    };
                    if (!syntaxObj[_lineNumber]) {
                        _node.startIndex = 0;
                        _node.endIndex = _valueLength - 1;
                        const item = {
                            value: _value,
                            lineNumber: _lineNumber,
                            valueLength: _valueLength,
                            nodeList: [{ node: _node, wrapper }],
                        };
                        syntaxObj[_lineNumber] = item;
                    }
                    else {
                        _node.startIndex = syntaxObj[_lineNumber].valueLength;
                        _node.endIndex = _node.startIndex + _valueLength - 1;
                        syntaxObj[_lineNumber].value += _value;
                        syntaxObj[_lineNumber].valueLength += _valueLength;
                        syntaxObj[_lineNumber].nodeList.push({ node: _node, wrapper });
                    }
                    node.children.push(_node);
                }
                node.lineNumber = lineNumber;
                return;
            }
            if (node.children) {
                loopAST(node.children, node);
                node.lineNumber = lineNumber;
            }
        });
    };
    loopAST(ast.children);
    return { syntaxFileObject: syntaxObj, syntaxFileLineNumber: lineNumber };
};

function _getAST(_raw, _fileName, _lang, _theme) {
    return {};
}

const lowlight = createLowlight(all);
// !SEE https://github.com/highlightjs/highlightjs-vue
lowlight.register("vue", function hljsDefineVue(hljs) {
    return {
        subLanguage: "xml",
        contains: [
            hljs.COMMENT("<!--", "-->", {
                relevance: 10,
            }),
            {
                begin: /^(\s*)(<script>)/gm,
                end: /^(\s*)(<\/script>)/gm,
                subLanguage: "javascript",
                excludeBegin: true,
                excludeEnd: true,
            },
            {
                begin: /^(?:\s*)(?:<script\s+lang=(["'])ts\1>)/gm,
                end: /^(\s*)(<\/script>)/gm,
                subLanguage: "typescript",
                excludeBegin: true,
                excludeEnd: true,
            },
            {
                begin: /^(\s*)(<style(\s+scoped)?>)/gm,
                end: /^(\s*)(<\/style>)/gm,
                subLanguage: "css",
                excludeBegin: true,
                excludeEnd: true,
            },
            {
                begin: /^(?:\s*)(?:<style(?:\s+scoped)?\s+lang=(["'])(?:s[ca]ss)\1(?:\s+scoped)?>)/gm,
                end: /^(\s*)(<\/style>)/gm,
                subLanguage: "scss",
                excludeBegin: true,
                excludeEnd: true,
            },
            {
                begin: /^(?:\s*)(?:<style(?:\s+scoped)?\s+lang=(["'])stylus\1(?:\s+scoped)?>)/gm,
                end: /^(\s*)(<\/style>)/gm,
                subLanguage: "stylus",
                excludeBegin: true,
                excludeEnd: true,
            },
        ],
    };
});
const instance = { name: "lowlight" };
let _maxLineToIgnoreSyntax = 2000;
const _ignoreSyntaxHighlightList = [];
Object.defineProperty(instance, "maxLineToIgnoreSyntax", {
    get: () => _maxLineToIgnoreSyntax,
});
Object.defineProperty(instance, "setMaxLineToIgnoreSyntax", {
    value: (v) => {
        _maxLineToIgnoreSyntax = v;
    },
});
Object.defineProperty(instance, "ignoreSyntaxHighlightList", {
    get: () => _ignoreSyntaxHighlightList,
});
Object.defineProperty(instance, "setIgnoreSyntaxHighlightList", {
    value: (v) => {
        _ignoreSyntaxHighlightList.length = 0;
        _ignoreSyntaxHighlightList.push(...v);
    },
});
Object.defineProperty(instance, "getAST", {
    value: (raw, fileName, lang) => {
        let hasRegisteredLang = true;
        if (!lowlight.registered(lang)) {
            if (process.env.NODE_ENV === "development") {
                console.warn(`not support current lang: ${lang} yet`);
            }
            hasRegisteredLang = false;
        }
        if (fileName &&
            highlighter.ignoreSyntaxHighlightList.some((item) => item instanceof RegExp ? item.test(fileName) : fileName === item)) {
            if (process.env.NODE_ENV === "development") {
                console.warn(`ignore syntax for current file, because the fileName is in the ignoreSyntaxHighlightList: ${fileName}`);
            }
            return;
        }
        if (hasRegisteredLang) {
            return lowlight.highlight(lang, raw);
        }
        else {
            return lowlight.highlightAuto(raw);
        }
    },
});
Object.defineProperty(instance, "processAST", {
    value: (ast) => {
        return processAST(ast);
    },
});
Object.defineProperty(instance, "hasRegisteredCurrentLang", {
    value: (lang) => {
        return lowlight.registered(lang);
    },
});
Object.defineProperty(instance, "getHighlighterEngine", {
    value: () => lowlight,
});
Object.defineProperty(instance, "type", { value: "class" });
const versions = "0.0.30";
const highlighter = instance;

export { _getAST, highlighter, processAST, versions };
