// @git-diff-view/lowlight/dist/cjs/index.development.js
'use strict';

var HighlightJs = require('highlight.js/lib/core');
var $1c = require('highlight.js/lib/languages/1c');
var abnf = require('highlight.js/lib/languages/abnf');
var accesslog = require('highlight.js/lib/languages/accesslog');
var actionscript = require('highlight.js/lib/languages/actionscript');
var ada = require('highlight.js/lib/languages/ada');
var angelscript = require('highlight.js/lib/languages/angelscript');
var apache = require('highlight.js/lib/languages/apache');
var applescript = require('highlight.js/lib/languages/applescript');
var arcade = require('highlight.js/lib/languages/arcade');
var armasm = require('highlight.js/lib/languages/armasm');
var asciidoc = require('highlight.js/lib/languages/asciidoc');
var aspectj = require('highlight.js/lib/languages/aspectj');
var autohotkey = require('highlight.js/lib/languages/autohotkey');
var autoit = require('highlight.js/lib/languages/autoit');
var avrasm = require('highlight.js/lib/languages/avrasm');
var awk = require('highlight.js/lib/languages/awk');
var axapta = require('highlight.js/lib/languages/axapta');
var basic = require('highlight.js/lib/languages/basic');
var bnf = require('highlight.js/lib/languages/bnf');
var brainfuck = require('highlight.js/lib/languages/brainfuck');
var cal = require('highlight.js/lib/languages/cal');
var capnproto = require('highlight.js/lib/languages/capnproto');
var ceylon = require('highlight.js/lib/languages/ceylon');
var clean = require('highlight.js/lib/languages/clean');
var clojure = require('highlight.js/lib/languages/clojure');
var clojureRepl = require('highlight.js/lib/languages/clojure-repl');
var cmake = require('highlight.js/lib/languages/cmake');
var coffeescript = require('highlight.js/lib/languages/coffeescript');
var coq = require('highlight.js/lib/languages/coq');
var cos = require('highlight.js/lib/languages/cos');
var crmsh = require('highlight.js/lib/languages/crmsh');
var crystal = require('highlight.js/lib/languages/crystal');
var csp = require('highlight.js/lib/languages/csp');
var d = require('highlight.js/lib/languages/d');
var dart = require('highlight.js/lib/languages/dart');
var delphi = require('highlight.js/lib/languages/delphi');
var django = require('highlight.js/lib/languages/django');
var dns = require('highlight.js/lib/languages/dns');
var dockerfile = require('highlight.js/lib/languages/dockerfile');
var dos = require('highlight.js/lib/languages/dos');
var dsconfig = require('highlight.js/lib/languages/dsconfig');
var dts = require('highlight.js/lib/languages/dts');
var dust = require('highlight.js/lib/languages/dust');
var ebnf = require('highlight.js/lib/languages/ebnf');
var elixir = require('highlight.js/lib/languages/elixir');
var elm = require('highlight.js/lib/languages/elm');
var erb = require('highlight.js/lib/languages/erb');
var erlang = require('highlight.js/lib/languages/erlang');
var erlangRepl = require('highlight.js/lib/languages/erlang-repl');
var excel = require('highlight.js/lib/languages/excel');
var fix = require('highlight.js/lib/languages/fix');
var flix = require('highlight.js/lib/languages/flix');
var fortran = require('highlight.js/lib/languages/fortran');
var fsharp = require('highlight.js/lib/languages/fsharp');
var gams = require('highlight.js/lib/languages/gams');
var gauss = require('highlight.js/lib/languages/gauss');
var gcode = require('highlight.js/lib/languages/gcode');
var gherkin = require('highlight.js/lib/languages/gherkin');
var glsl = require('highlight.js/lib/languages/glsl');
var gml = require('highlight.js/lib/languages/gml');
var golo = require('highlight.js/lib/languages/golo');
var gradle = require('highlight.js/lib/languages/gradle');
var groovy = require('highlight.js/lib/languages/groovy');
var haml = require('highlight.js/lib/languages/haml');
var handlebars = require('highlight.js/lib/languages/handlebars');
var haskell = require('highlight.js/lib/languages/haskell');
var haxe = require('highlight.js/lib/languages/haxe');
var hsp = require('highlight.js/lib/languages/hsp');
var http = require('highlight.js/lib/languages/http');
var hy = require('highlight.js/lib/languages/hy');
var inform7 = require('highlight.js/lib/languages/inform7');
var irpf90 = require('highlight.js/lib/languages/irpf90');
var isbl = require('highlight.js/lib/languages/isbl');
var jbossCli = require('highlight.js/lib/languages/jboss-cli');
var julia = require('highlight.js/lib/languages/julia');
var juliaRepl = require('highlight.js/lib/languages/julia-repl');
var lasso = require('highlight.js/lib/languages/lasso');
var latex = require('highlight.js/lib/languages/latex');
var ldif = require('highlight.js/lib/languages/ldif');
var leaf = require('highlight.js/lib/languages/leaf');
var lisp = require('highlight.js/lib/languages/lisp');
var livecodeserver = require('highlight.js/lib/languages/livecodeserver');
var livescript = require('highlight.js/lib/languages/livescript');
var llvm = require('highlight.js/lib/languages/llvm');
var lsl = require('highlight.js/lib/languages/lsl');
var mathematica = require('highlight.js/lib/languages/mathematica');
var matlab = require('highlight.js/lib/languages/matlab');
var maxima = require('highlight.js/lib/languages/maxima');
var mel = require('highlight.js/lib/languages/mel');
var mercury = require('highlight.js/lib/languages/mercury');
var mipsasm = require('highlight.js/lib/languages/mipsasm');
var mizar = require('highlight.js/lib/languages/mizar');
var mojolicious = require('highlight.js/lib/languages/mojolicious');
var monkey = require('highlight.js/lib/languages/monkey');
var moonscript = require('highlight.js/lib/languages/moonscript');
var n1ql = require('highlight.js/lib/languages/n1ql');
var nestedtext = require('highlight.js/lib/languages/nestedtext');
var nginx = require('highlight.js/lib/languages/nginx');
var nim = require('highlight.js/lib/languages/nim');
var nix = require('highlight.js/lib/languages/nix');
var nodeRepl = require('highlight.js/lib/languages/node-repl');
var nsis = require('highlight.js/lib/languages/nsis');
var ocaml = require('highlight.js/lib/languages/ocaml');
var openscad = require('highlight.js/lib/languages/openscad');
var oxygene = require('highlight.js/lib/languages/oxygene');
var parser3 = require('highlight.js/lib/languages/parser3');
var pf = require('highlight.js/lib/languages/pf');
var pgsql = require('highlight.js/lib/languages/pgsql');
var pony = require('highlight.js/lib/languages/pony');
var powershell = require('highlight.js/lib/languages/powershell');
var processing = require('highlight.js/lib/languages/processing');
var profile = require('highlight.js/lib/languages/profile');
var prolog = require('highlight.js/lib/languages/prolog');
var properties = require('highlight.js/lib/languages/properties');
var protobuf = require('highlight.js/lib/languages/protobuf');
var puppet = require('highlight.js/lib/languages/puppet');
var purebasic = require('highlight.js/lib/languages/purebasic');
var q = require('highlight.js/lib/languages/q');
var qml = require('highlight.js/lib/languages/qml');
var reasonml = require('highlight.js/lib/languages/reasonml');
var rib = require('highlight.js/lib/languages/rib');
var roboconf = require('highlight.js/lib/languages/roboconf');
var routeros = require('highlight.js/lib/languages/routeros');
var rsl = require('highlight.js/lib/languages/rsl');
var ruleslanguage = require('highlight.js/lib/languages/ruleslanguage');
var sas = require('highlight.js/lib/languages/sas');
var scala = require('highlight.js/lib/languages/scala');
var scheme = require('highlight.js/lib/languages/scheme');
var scilab = require('highlight.js/lib/languages/scilab');
var smali = require('highlight.js/lib/languages/smali');
var smalltalk = require('highlight.js/lib/languages/smalltalk');
var sml = require('highlight.js/lib/languages/sml');
var sqf = require('highlight.js/lib/languages/sqf');
var stan = require('highlight.js/lib/languages/stan');
var stata = require('highlight.js/lib/languages/stata');
var step21 = require('highlight.js/lib/languages/step21');
var stylus = require('highlight.js/lib/languages/stylus');
var subunit = require('highlight.js/lib/languages/subunit');
var taggerscript = require('highlight.js/lib/languages/taggerscript');
var tap = require('highlight.js/lib/languages/tap');
var tcl = require('highlight.js/lib/languages/tcl');
var thrift = require('highlight.js/lib/languages/thrift');
var tp = require('highlight.js/lib/languages/tp');
var twig = require('highlight.js/lib/languages/twig');
var vala = require('highlight.js/lib/languages/vala');
var vbscript = require('highlight.js/lib/languages/vbscript');
var vbscriptHtml = require('highlight.js/lib/languages/vbscript-html');
var verilog = require('highlight.js/lib/languages/verilog');
var vhdl = require('highlight.js/lib/languages/vhdl');
var vim = require('highlight.js/lib/languages/vim');
var wren = require('highlight.js/lib/languages/wren');
var x86asm = require('highlight.js/lib/languages/x86asm');
var xl = require('highlight.js/lib/languages/xl');
var xquery = require('highlight.js/lib/languages/xquery');
var zephir = require('highlight.js/lib/languages/zephir');
var arduino = require('highlight.js/lib/languages/arduino');
var bash = require('highlight.js/lib/languages/bash');
var c = require('highlight.js/lib/languages/c');
var cpp = require('highlight.js/lib/languages/cpp');
var csharp = require('highlight.js/lib/languages/csharp');
var css = require('highlight.js/lib/languages/css');
var diff = require('highlight.js/lib/languages/diff');
var go = require('highlight.js/lib/languages/go');
var graphql = require('highlight.js/lib/languages/graphql');
var ini = require('highlight.js/lib/languages/ini');
var java = require('highlight.js/lib/languages/java');
var javascript = require('highlight.js/lib/languages/javascript');
var json = require('highlight.js/lib/languages/json');
var kotlin = require('highlight.js/lib/languages/kotlin');
var less = require('highlight.js/lib/languages/less');
var lua = require('highlight.js/lib/languages/lua');
var makefile = require('highlight.js/lib/languages/makefile');
var markdown = require('highlight.js/lib/languages/markdown');
var objectivec = require('highlight.js/lib/languages/objectivec');
var perl = require('highlight.js/lib/languages/perl');
var php = require('highlight.js/lib/languages/php');
var phpTemplate = require('highlight.js/lib/languages/php-template');
var plaintext = require('highlight.js/lib/languages/plaintext');
var python = require('highlight.js/lib/languages/python');
var pythonRepl = require('highlight.js/lib/languages/python-repl');
var r = require('highlight.js/lib/languages/r');
var ruby = require('highlight.js/lib/languages/ruby');
var rust = require('highlight.js/lib/languages/rust');
var scss = require('highlight.js/lib/languages/scss');
var shell = require('highlight.js/lib/languages/shell');
var sql = require('highlight.js/lib/languages/sql');
var swift = require('highlight.js/lib/languages/swift');
var typescript = require('highlight.js/lib/languages/typescript');
var vbnet = require('highlight.js/lib/languages/vbnet');
var wasm = require('highlight.js/lib/languages/wasm');
var xml = require('highlight.js/lib/languages/xml');
var yaml = require('highlight.js/lib/languages/yaml');

/**
 * @import {LanguageFn} from 'highlight.js'
 */


/**
 * Map of grammars.
 *
 * @type {Record<string, LanguageFn>}
 */
const grammars$1 = {
  arduino,
  bash,
  c,
  cpp,
  csharp,
  css,
  diff,
  go,
  graphql,
  ini,
  java,
  javascript,
  json,
  kotlin,
  less,
  lua,
  makefile,
  markdown,
  objectivec,
  perl,
  php,
  'php-template': phpTemplate,
  plaintext,
  python,
  'python-repl': pythonRepl,
  r,
  ruby,
  rust,
  scss,
  shell,
  sql,
  swift,
  typescript,
  vbnet,
  wasm,
  xml,
  yaml
};

/**
 * @import {LanguageFn} from 'highlight.js'
 */


/**
 * Map of grammars.
 *
 * @type {Record<string, LanguageFn>}
 */
const grammars = {
  ...grammars$1,
  '1c': $1c,
  abnf,
  accesslog,
  actionscript,
  ada,
  angelscript,
  apache,
  applescript,
  arcade,
  armasm,
  asciidoc,
  aspectj,
  autohotkey,
  autoit,
  avrasm,
  awk,
  axapta,
  basic,
  bnf,
  brainfuck,
  cal,
  capnproto,
  ceylon,
  clean,
  clojure,
  'clojure-repl': clojureRepl,
  cmake,
  coffeescript,
  coq,
  cos,
  crmsh,
  crystal,
  csp,
  d,
  dart,
  delphi,
  django,
  dns,
  dockerfile,
  dos,
  dsconfig,
  dts,
  dust,
  ebnf,
  elixir,
  elm,
  erb,
  erlang,
  'erlang-repl': erlangRepl,
  excel,
  fix,
  flix,
  fortran,
  fsharp,
  gams,
  gauss,
  gcode,
  gherkin,
  glsl,
  gml,
  golo,
  gradle,
  groovy,
  haml,
  handlebars,
  haskell,
  haxe,
  hsp,
  http,
  hy,
  inform7,
  irpf90,
  isbl,
  'jboss-cli': jbossCli,
  julia,
  'julia-repl': juliaRepl,
  lasso,
  latex,
  ldif,
  leaf,
  lisp,
  livecodeserver,
  livescript,
  llvm,
  lsl,
  mathematica,
  matlab,
  maxima,
  mel,
  mercury,
  mipsasm,
  mizar,
  mojolicious,
  monkey,
  moonscript,
  n1ql,
  nestedtext,
  nginx,
  nim,
  nix,
  'node-repl': nodeRepl,
  nsis,
  ocaml,
  openscad,
  oxygene,
  parser3,
  pf,
  pgsql,
  pony,
  powershell,
  processing,
  profile,
  prolog,
  properties,
  protobuf,
  puppet,
  purebasic,
  q,
  qml,
  reasonml,
  rib,
  roboconf,
  routeros,
  rsl,
  ruleslanguage,
  sas,
  scala,
  scheme,
  scilab,
  smali,
  smalltalk,
  sml,
  sqf,
  stan,
  stata,
  step21,
  stylus,
  subunit,
  taggerscript,
  tap,
  tcl,
  thrift,
  tp,
  twig,
  vala,
  vbscript,
  'vbscript-html': vbscriptHtml,
  verilog,
  vhdl,
  vim,
  wren,
  x86asm,
  xl,
  xquery,
  zephir
};

/**
 * @import {ElementContent, Element, RootData, Root} from 'hast'
 * @import {Emitter, HLJSOptions as HljsOptions, HighlightResult, LanguageFn} from 'highlight.js'
 */


/** @type {AutoOptions} */
const emptyOptions = {};

const defaultPrefix = 'hljs-';

/**
 * Create a `lowlight` instance.
 *
 * @param {Readonly<Record<string, LanguageFn>> | null | undefined} [grammars]
 *   Grammars to add (optional).
 * @returns
 *   Lowlight.
 */
function createLowlight(grammars) {
  const high = HighlightJs.newInstance();

  if (grammars) {
    register(grammars);
  }

  return {
    highlight,
    highlightAuto,
    listLanguages,
    register,
    registerAlias,
    registered
  }

  /**
   * Highlight `value` (code) as `language` (name).
   *
   * @example
   *   ```js
   *   import {common, createLowlight} from 'lowlight'
   *
   *   const lowlight = createLowlight(common)
   *
   *   console.log(lowlight.highlight('css', 'em { color: red }'))
   *   ```
   *
   *   Yields:
   *
   *   ```js
   *   {type: 'root', children: [Array], data: {language: 'css', relevance: 3}}
   *   ```
   *
   * @param {string} language
   *   Programming language name.
   * @param {string} value
   *   Code to highlight.
   * @param {Readonly<Options> | null | undefined} [options={}]
   *   Configuration (optional).
   * @returns {Root}
   *   Tree; with the following `data` fields: `language` (`string`), detected
   *   programming language name; `relevance` (`number`), how sure lowlight is
   *   that the given code is in the language.
   */
  function highlight(language, value, options) {
    const settings = options || emptyOptions;
    const prefix =
      typeof settings.prefix === 'string' ? settings.prefix : defaultPrefix;

    if (!high.getLanguage(language)) {
      throw new Error('Unknown language: `' + language + '` is not registered')
    }

    // See: <https://github.com/highlightjs/highlight.js/issues/3621#issuecomment-1528841888>
    high.configure({__emitter: HastEmitter, classPrefix: prefix});

    const result = /** @type {HighlightResult & {_emitter: HastEmitter}} */ (
      high.highlight(value, {ignoreIllegals: true, language})
    );

    // `highlight.js` seems to use this (currently) for broken grammars, so let’s
    // keep it in there just to be sure.
    /* c8 ignore next 5 */
    if (result.errorRaised) {
      throw new Error('Could not highlight with `Highlight.js`', {
        cause: result.errorRaised
      })
    }

    const root = result._emitter.root;

    // Cast because it is always defined.
    const data = /** @type {RootData} */ (root.data);

    data.language = result.language;
    data.relevance = result.relevance;

    return root
  }

  /**
   * Highlight `value` (code) and guess its programming language.
   *
   * @example
   *   ```js
   *   import {common, createLowlight} from 'lowlight'
   *
   *   const lowlight = createLowlight(common)
   *
   *   console.log(lowlight.highlightAuto('"hello, " + name + "!"'))
   *   ```
   *
   *   Yields:
   *
   *   ```js
   *   {type: 'root', children: [Array], data: {language: 'arduino', relevance: 2}}
   *   ```
   *
   * @param {string} value
   *   Code to highlight.
   * @param {Readonly<AutoOptions> | null | undefined} [options={}]
   *   Configuration (optional).
   * @returns {Root}
   *   Tree; with the following `data` fields: `language` (`string`), detected
   *   programming language name; `relevance` (`number`), how sure lowlight is
   *   that the given code is in the language.
   */
  function highlightAuto(value, options) {
    const settings = options || emptyOptions;
    const subset = settings.subset || listLanguages();

    let index = -1;
    let relevance = 0;
    /** @type {Root | undefined} */
    let result;

    while (++index < subset.length) {
      const name = subset[index];

      if (!high.getLanguage(name)) continue

      const current = highlight(name, value, options);

      if (
        current.data &&
        current.data.relevance !== undefined &&
        current.data.relevance > relevance
      ) {
        relevance = current.data.relevance;
        result = current;
      }
    }

    return (
      result || {
        type: 'root',
        children: [],
        data: {language: undefined, relevance}
      }
    )
  }

  /**
   * List registered languages.
   *
   * @example
   *   ```js
   *   import {createLowlight} from 'lowlight'
   *   import markdown from 'highlight.js/lib/languages/markdown'
   *
   *   const lowlight = createLowlight()
   *
   *   console.log(lowlight.listLanguages()) // => []
   *
   *   lowlight.register({markdown})
   *
   *   console.log(lowlight.listLanguages()) // => ['markdown']
   *   ```
   *
   * @returns {Array<string>}
   *   Names of registered language.
   */
  function listLanguages() {
    return high.listLanguages()
  }

  /**
   * Register languages.
   *
   * @example
   *   ```js
   *   import {createLowlight} from 'lowlight'
   *   import xml from 'highlight.js/lib/languages/xml'
   *
   *   const lowlight = createLowlight()
   *
   *   lowlight.register({xml})
   *
   *   // Note: `html` is an alias for `xml`.
   *   console.log(lowlight.highlight('html', '<em>Emphasis</em>'))
   *   ```
   *
   *   Yields:
   *
   *   ```js
   *   {type: 'root', children: [Array], data: {language: 'html', relevance: 2}}
   *   ```
   *
   * @overload
   * @param {Readonly<Record<string, LanguageFn>>} grammars
   * @returns {undefined}
   *
   * @overload
   * @param {string} name
   * @param {LanguageFn} grammar
   * @returns {undefined}
   *
   * @param {Readonly<Record<string, LanguageFn>> | string} grammarsOrName
   *   Grammars or programming language name.
   * @param {LanguageFn | undefined} [grammar]
   *   Grammar, if with name.
   * @returns {undefined}
   *   Nothing.
   */
  function register(grammarsOrName, grammar) {
    if (typeof grammarsOrName === 'string') {
      high.registerLanguage(grammarsOrName, grammar);
    } else {
      /** @type {string} */
      let name;

      for (name in grammarsOrName) {
        if (Object.hasOwn(grammarsOrName, name)) {
          high.registerLanguage(name, grammarsOrName[name]);
        }
      }
    }
  }

  /**
   * Register aliases.
   *
   * @example
   *   ```js
   *   import {createLowlight} from 'lowlight'
   *   import markdown from 'highlight.js/lib/languages/markdown'
   *
   *   const lowlight = createLowlight()
   *
   *   lowlight.register({markdown})
   *
   *   // lowlight.highlight('mdown', '<em>Emphasis</em>')
   *   // ^ would throw: Error: Unknown language: `mdown` is not registered
   *
   *   lowlight.registerAlias({markdown: ['mdown', 'mkdn', 'mdwn', 'ron']})
   *   lowlight.highlight('mdown', '<em>Emphasis</em>')
   *   // ^ Works!
   *   ```
   *
   * @overload
   * @param {Readonly<Record<string, ReadonlyArray<string> | string>>} aliases
   * @returns {undefined}
   *
   * @overload
   * @param {string} language
   * @param {ReadonlyArray<string> | string} alias
   * @returns {undefined}
   *
   * @param {Readonly<Record<string, ReadonlyArray<string> | string>> | string} aliasesOrName
   *   Map of programming language names to one or more aliases, or programming
   *   language name.
   * @param {ReadonlyArray<string> | string | undefined} [alias]
   *   One or more aliases for the programming language, if with `name`.
   * @returns {undefined}
   *   Nothing.
   */
  function registerAlias(aliasesOrName, alias) {
    if (typeof aliasesOrName === 'string') {
      high.registerAliases(
        // Note: copy needed because hljs doesn’t accept readonly arrays yet.
        typeof alias === 'string' ? alias : [...alias],
        {languageName: aliasesOrName}
      );
    } else {
      /** @type {string} */
      let key;

      for (key in aliasesOrName) {
        if (Object.hasOwn(aliasesOrName, key)) {
          const aliases = aliasesOrName[key];
          high.registerAliases(
            // Note: copy needed because hljs doesn’t accept readonly arrays yet.
            typeof aliases === 'string' ? aliases : [...aliases],
            {languageName: key}
          );
        }
      }
    }
  }

  /**
   * Check whether an alias or name is registered.
   *
   * @example
   *   ```js
   *   import {createLowlight} from 'lowlight'
   *   import javascript from 'highlight.js/lib/languages/javascript'
   *
   *   const lowlight = createLowlight({javascript})
   *
   *   console.log(lowlight.registered('funkyscript')) // => `false`
   *
   *   lowlight.registerAlias({javascript: 'funkyscript'})
   *   console.log(lowlight.registered('funkyscript')) // => `true`
   *   ```
   *
   * @param {string} aliasOrName
   *   Name of a language or alias for one.
   * @returns {boolean}
   *   Whether `aliasOrName` is registered.
   */
  function registered(aliasOrName) {
    return Boolean(high.getLanguage(aliasOrName))
  }
}

/** @type {Emitter} */
class HastEmitter {
  /**
   * @param {Readonly<HljsOptions>} options
   *   Configuration.
   * @returns
   *   Instance.
   */
  constructor(options) {
    /** @type {HljsOptions} */
    this.options = options;
    /** @type {Root} */
    this.root = {
      type: 'root',
      children: [],
      data: {language: undefined, relevance: 0}
    };
    /** @type {[Root, ...Array<Element>]} */
    this.stack = [this.root];
  }

  /**
   * @param {string} value
   *   Text to add.
   * @returns {undefined}
   *   Nothing.
   *
   */
  addText(value) {
    if (value === '') return

    const current = this.stack[this.stack.length - 1];
    const tail = current.children[current.children.length - 1];

    if (tail && tail.type === 'text') {
      tail.value += value;
    } else {
      current.children.push({type: 'text', value});
    }
  }

  /**
   *
   * @param {unknown} rawName
   *   Name to add.
   * @returns {undefined}
   *   Nothing.
   */
  startScope(rawName) {
    this.openNode(String(rawName));
  }

  /**
   * @returns {undefined}
   *   Nothing.
   */
  endScope() {
    this.closeNode();
  }

  /**
   * @param {HastEmitter} other
   *   Other emitter.
   * @param {string} name
   *   Name of the sublanguage.
   * @returns {undefined}
   *   Nothing.
   */
  __addSublanguage(other, name) {
    const current = this.stack[this.stack.length - 1];
    // Assume only element content.
    const results = /** @type {Array<ElementContent>} */ (other.root.children);

    if (name) {
      current.children.push({
        type: 'element',
        tagName: 'span',
        properties: {className: [name]},
        children: results
      });
    } else {
      current.children.push(...results);
    }
  }

  /**
   * @param {string} name
   *   Name to add.
   * @returns {undefined}
   *   Nothing.
   */
  openNode(name) {
    const self = this;
    // First “class” gets the prefix. Rest gets a repeated underscore suffix.
    // See: <https://github.com/highlightjs/highlight.js/commit/51806aa>
    // See: <https://github.com/wooorm/lowlight/issues/43>
    const className = name.split('.').map(function (d, i) {
      return i ? d + '_'.repeat(i) : self.options.classPrefix + d
    });
    const current = this.stack[this.stack.length - 1];
    /** @type {Element} */
    const child = {
      type: 'element',
      tagName: 'span',
      properties: {className},
      children: []
    };

    current.children.push(child);
    this.stack.push(child);
  }

  /**
   * @returns {undefined}
   *   Nothing.
   */
  closeNode() {
    this.stack.pop();
  }

  /**
   * @returns {undefined}
   *   Nothing.
   */
  finalize() {}

  /**
   * @returns {string}
   *   Nothing.
   */
  toHTML() {
    return ''
  }
}

const processAST = (ast) => {
    let lineNumber = 1;
    const syntaxObj = {};
    const loopAST = (nodes, wrapper) => {
        nodes.forEach((node) => {
            if (node.type === "text") {
                if (node.value.indexOf("\n") === -1) {
                    const valueLength = node.value.length;
                    if (!syntaxObj[lineNumber]) {
                        node.startIndex = 0;
                        node.endIndex = valueLength - 1;
                        const item = {
                            value: node.value,
                            lineNumber,
                            valueLength,
                            nodeList: [{ node, wrapper }],
                        };
                        syntaxObj[lineNumber] = item;
                    }
                    else {
                        node.startIndex = syntaxObj[lineNumber].valueLength;
                        node.endIndex = node.startIndex + valueLength - 1;
                        syntaxObj[lineNumber].value += node.value;
                        syntaxObj[lineNumber].valueLength += valueLength;
                        syntaxObj[lineNumber].nodeList.push({ node, wrapper });
                    }
                    node.lineNumber = lineNumber;
                    return;
                }
                const lines = node.value.split("\n");
                node.children = node.children || [];
                for (let i = 0; i < lines.length; i++) {
                    const _value = i === lines.length - 1 ? lines[i] : lines[i] + "\n";
                    const _lineNumber = i === 0 ? lineNumber : ++lineNumber;
                    const _valueLength = _value.length;
                    const _node = {
                        type: "text",
                        value: _value,
                        startIndex: Infinity,
                        endIndex: Infinity,
                        lineNumber: _lineNumber,
                    };
                    if (!syntaxObj[_lineNumber]) {
                        _node.startIndex = 0;
                        _node.endIndex = _valueLength - 1;
                        const item = {
                            value: _value,
                            lineNumber: _lineNumber,
                            valueLength: _valueLength,
                            nodeList: [{ node: _node, wrapper }],
                        };
                        syntaxObj[_lineNumber] = item;
                    }
                    else {
                        _node.startIndex = syntaxObj[_lineNumber].valueLength;
                        _node.endIndex = _node.startIndex + _valueLength - 1;
                        syntaxObj[_lineNumber].value += _value;
                        syntaxObj[_lineNumber].valueLength += _valueLength;
                        syntaxObj[_lineNumber].nodeList.push({ node: _node, wrapper });
                    }
                    node.children.push(_node);
                }
                node.lineNumber = lineNumber;
                return;
            }
            if (node.children) {
                loopAST(node.children, node);
                node.lineNumber = lineNumber;
            }
        });
    };
    loopAST(ast.children);
    return { syntaxFileObject: syntaxObj, syntaxFileLineNumber: lineNumber };
};

function _getAST(_raw, _fileName, _lang, _theme) {
    return {};
}

const lowlight = createLowlight(grammars);
// !SEE https://github.com/highlightjs/highlightjs-vue
lowlight.register("vue", function hljsDefineVue(hljs) {
    return {
        subLanguage: "xml",
        contains: [
            hljs.COMMENT("<!--", "-->", {
                relevance: 10,
            }),
            {
                begin: /^(\s*)(<script>)/gm,
                end: /^(\s*)(<\/script>)/gm,
                subLanguage: "javascript",
                excludeBegin: true,
                excludeEnd: true,
            },
            {
                begin: /^(?:\s*)(?:<script\s+lang=(["'])ts\1>)/gm,
                end: /^(\s*)(<\/script>)/gm,
                subLanguage: "typescript",
                excludeBegin: true,
                excludeEnd: true,
            },
            {
                begin: /^(\s*)(<style(\s+scoped)?>)/gm,
                end: /^(\s*)(<\/style>)/gm,
                subLanguage: "css",
                excludeBegin: true,
                excludeEnd: true,
            },
            {
                begin: /^(?:\s*)(?:<style(?:\s+scoped)?\s+lang=(["'])(?:s[ca]ss)\1(?:\s+scoped)?>)/gm,
                end: /^(\s*)(<\/style>)/gm,
                subLanguage: "scss",
                excludeBegin: true,
                excludeEnd: true,
            },
            {
                begin: /^(?:\s*)(?:<style(?:\s+scoped)?\s+lang=(["'])stylus\1(?:\s+scoped)?>)/gm,
                end: /^(\s*)(<\/style>)/gm,
                subLanguage: "stylus",
                excludeBegin: true,
                excludeEnd: true,
            },
        ],
    };
});
const instance = { name: "lowlight" };
let _maxLineToIgnoreSyntax = 2000;
const _ignoreSyntaxHighlightList = [];
Object.defineProperty(instance, "maxLineToIgnoreSyntax", {
    get: () => _maxLineToIgnoreSyntax,
});
Object.defineProperty(instance, "setMaxLineToIgnoreSyntax", {
    value: (v) => {
        _maxLineToIgnoreSyntax = v;
    },
});
Object.defineProperty(instance, "ignoreSyntaxHighlightList", {
    get: () => _ignoreSyntaxHighlightList,
});
Object.defineProperty(instance, "setIgnoreSyntaxHighlightList", {
    value: (v) => {
        _ignoreSyntaxHighlightList.length = 0;
        _ignoreSyntaxHighlightList.push(...v);
    },
});
Object.defineProperty(instance, "getAST", {
    value: (raw, fileName, lang) => {
        let hasRegisteredLang = true;
        if (!lowlight.registered(lang)) {
            {
                console.warn(`not support current lang: ${lang} yet`);
            }
            hasRegisteredLang = false;
        }
        if (fileName &&
            highlighter.ignoreSyntaxHighlightList.some((item) => item instanceof RegExp ? item.test(fileName) : fileName === item)) {
            {
                console.warn(`ignore syntax for current file, because the fileName is in the ignoreSyntaxHighlightList: ${fileName}`);
            }
            return;
        }
        if (hasRegisteredLang) {
            return lowlight.highlight(lang, raw);
        }
        else {
            return lowlight.highlightAuto(raw);
        }
    },
});
Object.defineProperty(instance, "processAST", {
    value: (ast) => {
        return processAST(ast);
    },
});
Object.defineProperty(instance, "hasRegisteredCurrentLang", {
    value: (lang) => {
        return lowlight.registered(lang);
    },
});
Object.defineProperty(instance, "getHighlighterEngine", {
    value: () => lowlight,
});
Object.defineProperty(instance, "type", { value: "class" });
const versions = "0.0.30";
const highlighter = instance;

exports._getAST = _getAST;
exports.highlighter = highlighter;
exports.processAST = processAST;
exports.versions = versions;
