// @rjsf/shadcn/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Form: () => Form_default,
  Templates: () => Templates_default,
  Theme: () => Theme_default,
  Widgets: () => Widgets_default,
  default: () => index_default,
  generateForm: () => generateForm,
  generateTemplates: () => generateTemplates,
  generateTheme: () => generateTheme,
  generateWidgets: () => generateWidgets
});
module.exports = __toCommonJS(index_exports);

// src/Form/Form.tsx
var import_core3 = require("@rjsf/core");

// src/AddButton/AddButton.tsx
var import_utils2 = require("@rjsf/utils");
var import_lucide_react = require("lucide-react");

// src/components/ui/button.tsx
var import_react_slot = require("@radix-ui/react-slot");
var import_class_variance_authority = require("class-variance-authority");

// src/lib/utils.ts
var import_clsx = require("clsx");
var import_tailwind_merge = require("tailwind-merge");
function cn(...inputs) {
  return (0, import_tailwind_merge.twMerge)((0, import_clsx.clsx)(inputs));
}

// src/components/ui/button.tsx
var import_jsx_runtime = require("react/jsx-runtime");
var buttonVariants = (0, import_class_variance_authority.cva)(
  "inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive",
  {
    variants: {
      variant: {
        default: "bg-primary text-primary-foreground shadow-xs hover:bg-primary/90",
        destructive: "bg-destructive text-white shadow-xs hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60",
        outline: "border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50",
        secondary: "bg-secondary text-secondary-foreground shadow-xs hover:bg-secondary/80",
        ghost: "hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50",
        link: "text-primary underline-offset-4 hover:underline"
      },
      size: {
        default: "h-9 px-4 py-2 has-[>svg]:px-3",
        sm: "h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5",
        lg: "h-10 rounded-md px-6 has-[>svg]:px-4",
        xs: "h-6 rounded-md gap-1 px-2 has-[>svg]:px-2",
        icon: "size-9"
      }
    },
    defaultVariants: {
      variant: "default",
      size: "default"
    }
  }
);
function Button({
  className,
  variant,
  size,
  asChild = false,
  ...props
}) {
  const Comp = asChild ? import_react_slot.Slot : "button";
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Comp, { "data-slot": "button", className: cn(buttonVariants({ variant, size, className })), ...props });
}

// src/AddButton/AddButton.tsx
var import_jsx_runtime2 = require("react/jsx-runtime");
function AddButton({
  uiSchema,
  registry,
  className,
  ...props
}) {
  const { translateString } = registry;
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)("div", { className: "p-0 m-0", children: /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)(Button, { ...props, className: cn("w-fit gap-2", className), variant: "outline", type: "button", children: [
    /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(import_lucide_react.PlusCircle, { size: 16 }),
    " ",
    translateString(import_utils2.TranslatableString.AddItemButton)
  ] }) });
}

// src/ArrayFieldItemTemplate/ArrayFieldItemTemplate.tsx
var import_utils4 = require("@rjsf/utils");
var import_jsx_runtime3 = require("react/jsx-runtime");
function ArrayFieldItemTemplate(props) {
  const { children, buttonsProps, displayLabel, hasDescription, hasToolbar, uiSchema, registry } = props;
  const uiOptions = (0, import_utils4.getUiOptions)(uiSchema);
  const ArrayFieldItemButtonsTemplate = (0, import_utils4.getTemplate)(
    "ArrayFieldItemButtonsTemplate",
    registry,
    uiOptions
  );
  const margin = hasDescription ? -6 : 22;
  return /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("div", { children: /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)("div", { className: "mb-2 flex flex-row flex-wrap items-center", children: [
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("div", { className: "grow shrink", children }),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("div", { className: "flex items-end justify-end p-0.5", children: hasToolbar && /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "div",
      {
        className: "flex gap-2",
        style: { marginLeft: "5px", marginTop: displayLabel ? `${margin}px` : void 0 },
        children: /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(ArrayFieldItemButtonsTemplate, { ...buttonsProps })
      }
    ) })
  ] }) });
}

// src/ArrayFieldTemplate/ArrayFieldTemplate.tsx
var import_utils5 = require("@rjsf/utils");
var import_jsx_runtime4 = require("react/jsx-runtime");
function ArrayFieldTemplate(props) {
  const {
    canAdd,
    disabled,
    fieldPathId,
    uiSchema,
    items,
    optionalDataControl,
    onAddClick,
    readonly,
    registry,
    required,
    schema,
    title
  } = props;
  const uiOptions = (0, import_utils5.getUiOptions)(uiSchema);
  const ArrayFieldDescriptionTemplate = (0, import_utils5.getTemplate)(
    "ArrayFieldDescriptionTemplate",
    registry,
    uiOptions
  );
  const ArrayFieldTitleTemplate = (0, import_utils5.getTemplate)(
    "ArrayFieldTitleTemplate",
    registry,
    uiOptions
  );
  const showOptionalDataControlInTitle = !readonly && !disabled;
  const {
    ButtonTemplates: { AddButton: AddButton2 }
  } = registry.templates;
  return /* @__PURE__ */ (0, import_jsx_runtime4.jsx)("div", { children: /* @__PURE__ */ (0, import_jsx_runtime4.jsx)("div", { className: "m-0 flex p-0", children: /* @__PURE__ */ (0, import_jsx_runtime4.jsxs)("div", { className: "m-0 w-full p-0", children: [
    /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
      ArrayFieldTitleTemplate,
      {
        fieldPathId,
        title: uiOptions.title || title,
        schema,
        uiSchema,
        required,
        registry,
        optionalDataControl: showOptionalDataControlInTitle ? optionalDataControl : void 0
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
      ArrayFieldDescriptionTemplate,
      {
        fieldPathId,
        description: uiOptions.description || schema.description,
        schema,
        uiSchema,
        registry
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime4.jsxs)("div", { className: "p-0 m-0 w-full mb-2", children: [
      !showOptionalDataControlInTitle ? optionalDataControl : void 0,
      items,
      canAdd && /* @__PURE__ */ (0, import_jsx_runtime4.jsx)("div", { className: "mt-2 flex justify-end", children: /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
        AddButton2,
        {
          id: (0, import_utils5.buttonId)(fieldPathId, "add"),
          className: "rjsf-array-item-add",
          onClick: onAddClick,
          disabled: disabled || readonly,
          uiSchema,
          registry
        }
      ) })
    ] }, `array-item-list-${fieldPathId.$id}`)
  ] }) }) });
}

// src/BaseInputTemplate/BaseInputTemplate.tsx
var import_utils7 = require("@rjsf/utils");

// src/components/ui/input.tsx
var import_jsx_runtime5 = require("react/jsx-runtime");
function Input({ className, type, ...props }) {
  return /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(
    "input",
    {
      type,
      "data-slot": "input",
      className: cn(
        "file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm",
        "focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px]",
        "aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive",
        className
      ),
      ...props
    }
  );
}

// src/BaseInputTemplate/BaseInputTemplate.tsx
var import_jsx_runtime6 = require("react/jsx-runtime");
function BaseInputTemplate({
  id,
  htmlName,
  placeholder,
  required,
  readonly,
  disabled,
  type,
  value,
  onChange,
  onChangeOverride,
  onBlur,
  onFocus,
  autofocus,
  options,
  schema,
  rawErrors = [],
  children,
  extraProps,
  className
}) {
  const inputProps = {
    ...extraProps,
    ...(0, import_utils7.getInputProps)(schema, type, options)
  };
  const _onChange = ({ target: { value: value2 } }) => onChange(value2 === "" ? options.emptyValue : value2);
  const _onBlur = ({ target }) => onBlur(id, target && target.value);
  const _onFocus = ({ target }) => onFocus(id, target && target.value);
  return /* @__PURE__ */ (0, import_jsx_runtime6.jsxs)("div", { className: "p-0.5", children: [
    /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
      Input,
      {
        id,
        name: htmlName || id,
        type,
        placeholder,
        autoFocus: autofocus,
        required,
        disabled,
        readOnly: readonly,
        className: cn({ "border-destructive focus-visible:ring-0": rawErrors.length > 0 }, className),
        list: schema.examples ? (0, import_utils7.examplesId)(id) : void 0,
        ...inputProps,
        value: value || value === 0 ? value : "",
        onChange: onChangeOverride || _onChange,
        onBlur: _onBlur,
        onFocus: _onFocus,
        "aria-describedby": (0, import_utils7.ariaDescribedByIds)(id, !!schema.examples)
      }
    ),
    children,
    Array.isArray(schema.examples) ? /* @__PURE__ */ (0, import_jsx_runtime6.jsx)("datalist", { id: (0, import_utils7.examplesId)(id), children: schema.examples.concat(schema.default && !schema.examples.includes(schema.default) ? [schema.default] : []).map((example) => {
      return /* @__PURE__ */ (0, import_jsx_runtime6.jsx)("option", { value: example }, example);
    }) }) : null
  ] });
}

// src/DescriptionField/DescriptionField.tsx
var import_core = require("@rjsf/core");
var import_jsx_runtime7 = require("react/jsx-runtime");
function DescriptionField({ id, description, registry, uiSchema }) {
  if (!description) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime7.jsx)("div", { children: /* @__PURE__ */ (0, import_jsx_runtime7.jsx)("div", { id, className: "text-sm text-muted-foreground", children: /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(import_core.RichDescription, { description, registry, uiSchema }) }) });
}

// src/ErrorList/ErrorList.tsx
var import_utils10 = require("@rjsf/utils");
var import_lucide_react2 = require("lucide-react");

// src/components/ui/alert.tsx
var import_class_variance_authority2 = require("class-variance-authority");
var import_jsx_runtime8 = require("react/jsx-runtime");
var alertVariants = (0, import_class_variance_authority2.cva)(
  "relative w-full rounded-lg border px-4 py-3 text-sm grid has-[>svg]:grid-cols-[calc(var(--spacing)*4)_1fr] grid-cols-[0_1fr] has-[>svg]:gap-x-3 gap-y-0.5 items-start [&>svg]:size-4 [&>svg]:translate-y-0.5 [&>svg]:text-current",
  {
    variants: {
      variant: {
        default: "bg-card text-card-foreground",
        destructive: "text-destructive bg-card [&>svg]:text-current *:data-[slot=alert-description]:text-destructive/90"
      }
    },
    defaultVariants: {
      variant: "default"
    }
  }
);
function Alert({ className, variant, ...props }) {
  return /* @__PURE__ */ (0, import_jsx_runtime8.jsx)("div", { "data-slot": "alert", role: "alert", className: cn(alertVariants({ variant }), className), ...props });
}
function AlertTitle({ className, ...props }) {
  return /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
    "div",
    {
      "data-slot": "alert-title",
      className: cn("col-start-2 line-clamp-1 min-h-4 font-medium tracking-tight", className),
      ...props
    }
  );
}
function AlertDescription({ className, ...props }) {
  return /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
    "div",
    {
      "data-slot": "alert-description",
      className: cn(
        "text-muted-foreground col-start-2 grid justify-items-start gap-1 text-sm [&_p]:leading-relaxed",
        className
      ),
      ...props
    }
  );
}

// src/ErrorList/ErrorList.tsx
var import_jsx_runtime9 = require("react/jsx-runtime");
function ErrorList({
  errors,
  registry
}) {
  const { translateString } = registry;
  return /* @__PURE__ */ (0, import_jsx_runtime9.jsxs)(Alert, { variant: "destructive", className: "mb-2", children: [
    /* @__PURE__ */ (0, import_jsx_runtime9.jsx)(import_lucide_react2.AlertCircle, { className: "h-4 w-4" }),
    /* @__PURE__ */ (0, import_jsx_runtime9.jsx)(AlertTitle, { children: translateString(import_utils10.TranslatableString.ErrorsLabel) }),
    /* @__PURE__ */ (0, import_jsx_runtime9.jsx)(AlertDescription, { className: "flex flex-col gap-1", children: errors.map((error, i) => {
      return /* @__PURE__ */ (0, import_jsx_runtime9.jsxs)("span", { children: [
        "\u2022 ",
        error.stack
      ] }, i);
    }) })
  ] });
}

// src/FieldErrorTemplate/FieldErrorTemplate.tsx
var import_utils11 = require("@rjsf/utils");
var import_jsx_runtime10 = require("react/jsx-runtime");
function FieldErrorTemplate(props) {
  const { errors = [], fieldPathId } = props;
  if (errors.length === 0) {
    return null;
  }
  const id = (0, import_utils11.errorId)(fieldPathId);
  return /* @__PURE__ */ (0, import_jsx_runtime10.jsx)("div", { className: "flex flex-col gap-1", id, children: errors.map((error, i) => {
    return /* @__PURE__ */ (0, import_jsx_runtime10.jsx)("span", { className: "text-xs font-medium text-destructive mb-1", children: error }, i);
  }) });
}

// src/FieldHelpTemplate/FieldHelpTemplate.tsx
var import_utils12 = require("@rjsf/utils");
var import_core2 = require("@rjsf/core");
var import_jsx_runtime11 = require("react/jsx-runtime");
function FieldHelpTemplate(props) {
  const { fieldPathId, help, uiSchema, registry, hasErrors } = props;
  if (!help) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(
    "span",
    {
      className: cn("text-xs font-medium text-muted-foreground", { "text-destructive": hasErrors }),
      id: (0, import_utils12.helpId)(fieldPathId),
      children: /* @__PURE__ */ (0, import_jsx_runtime11.jsx)(import_core2.RichHelp, { help, registry, uiSchema })
    }
  );
}

// src/FieldTemplate/FieldTemplate.tsx
var import_utils14 = require("@rjsf/utils");
var import_jsx_runtime12 = require("react/jsx-runtime");
function FieldTemplate({
  id,
  children,
  displayLabel,
  rawErrors = [],
  errors,
  help,
  description,
  rawDescription,
  classNames,
  style,
  disabled,
  label,
  hidden,
  onKeyRename,
  onKeyRenameBlur,
  onRemoveProperty,
  readonly,
  required,
  schema,
  uiSchema,
  registry
}) {
  const uiOptions = (0, import_utils14.getUiOptions)(uiSchema);
  const WrapIfAdditionalTemplate2 = (0, import_utils14.getTemplate)(
    "WrapIfAdditionalTemplate",
    registry,
    uiOptions
  );
  if (hidden) {
    return /* @__PURE__ */ (0, import_jsx_runtime12.jsx)("div", { className: "hidden", children });
  }
  const isCheckbox = uiOptions.widget === "checkbox";
  return /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(
    WrapIfAdditionalTemplate2,
    {
      classNames,
      style,
      disabled,
      id,
      label,
      displayLabel,
      onKeyRename,
      onKeyRenameBlur,
      onRemoveProperty,
      rawDescription,
      readonly,
      required,
      schema,
      uiSchema,
      registry,
      children: /* @__PURE__ */ (0, import_jsx_runtime12.jsxs)("div", { className: "flex flex-col gap-2", children: [
        displayLabel && !isCheckbox && /* @__PURE__ */ (0, import_jsx_runtime12.jsxs)(
          "label",
          {
            className: cn(
              "text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70",
              { " text-destructive": rawErrors.length > 0 }
            ),
            htmlFor: id,
            children: [
              label,
              required ? "*" : null
            ]
          }
        ),
        children,
        displayLabel && rawDescription && !isCheckbox && /* @__PURE__ */ (0, import_jsx_runtime12.jsx)(
          "span",
          {
            className: cn("text-xs font-medium text-muted-foreground", { " text-destructive": rawErrors.length > 0 }),
            children: description
          }
        ),
        errors,
        help
      ] })
    }
  );
}

// src/GridTemplate/GridTemplate.tsx
var import_jsx_runtime13 = require("react/jsx-runtime");
function GridTemplate(props) {
  const { children, column, className, ...rest } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime13.jsx)("div", { className: cn("grid gap-2", className), ...rest, children });
}

// src/IconButton/IconButton.tsx
var import_utils17 = require("@rjsf/utils");
var import_lucide_react3 = require("lucide-react");
var import_jsx_runtime14 = require("react/jsx-runtime");
function IconButton(props) {
  const { icon, iconType, className, uiSchema, registry, ...otherProps } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime14.jsx)(Button, { size: "icon", variant: "outline", className, ...otherProps, type: "button", children: icon });
}
function CopyButton(props) {
  const {
    registry: { translateString }
  } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime14.jsx)(IconButton, { title: translateString(import_utils17.TranslatableString.CopyButton), ...props, icon: /* @__PURE__ */ (0, import_jsx_runtime14.jsx)(import_lucide_react3.Copy, { className: "h-4 w-4" }) });
}
function MoveDownButton(props) {
  const {
    registry: { translateString }
  } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime14.jsx)(
    IconButton,
    {
      title: translateString(import_utils17.TranslatableString.MoveDownButton),
      ...props,
      icon: /* @__PURE__ */ (0, import_jsx_runtime14.jsx)(import_lucide_react3.ChevronDown, { className: "h-4 w-4" })
    }
  );
}
function MoveUpButton(props) {
  const {
    registry: { translateString }
  } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime14.jsx)(
    IconButton,
    {
      title: translateString(import_utils17.TranslatableString.MoveUpButton),
      ...props,
      icon: /* @__PURE__ */ (0, import_jsx_runtime14.jsx)(import_lucide_react3.ChevronUp, { className: "h-4 w-4" })
    }
  );
}
function RemoveButton(props) {
  const {
    registry: { translateString }
  } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime14.jsx)(
    IconButton,
    {
      title: translateString(import_utils17.TranslatableString.RemoveButton),
      ...props,
      className: "border-destructive",
      icon: /* @__PURE__ */ (0, import_jsx_runtime14.jsx)(import_lucide_react3.Trash2, { className: "h-4 w-4 stroke-destructive" })
    }
  );
}

// src/MultiSchemaFieldTemplate/MultiSchemaFieldTemplate.tsx
var import_jsx_runtime15 = require("react/jsx-runtime");
function MultiSchemaFieldTemplate({ selector, optionSchemaField }) {
  return /* @__PURE__ */ (0, import_jsx_runtime15.jsxs)("div", { className: cn("p-4 border rounded-md bg-background shadow-sm"), children: [
    /* @__PURE__ */ (0, import_jsx_runtime15.jsx)("div", { className: cn("mb-4"), children: selector }),
    optionSchemaField
  ] });
}

// src/ObjectFieldTemplate/ObjectFieldTemplate.tsx
var import_utils19 = require("@rjsf/utils");
var import_jsx_runtime16 = require("react/jsx-runtime");
function ObjectFieldTemplate({
  description,
  title,
  properties,
  required,
  uiSchema,
  fieldPathId,
  schema,
  formData,
  optionalDataControl,
  onAddProperty,
  disabled,
  readonly,
  registry
}) {
  const uiOptions = (0, import_utils19.getUiOptions)(uiSchema);
  const TitleFieldTemplate = (0, import_utils19.getTemplate)("TitleFieldTemplate", registry, uiOptions);
  const DescriptionFieldTemplate = (0, import_utils19.getTemplate)(
    "DescriptionFieldTemplate",
    registry,
    uiOptions
  );
  const showOptionalDataControlInTitle = !readonly && !disabled;
  const {
    ButtonTemplates: { AddButton: AddButton2 }
  } = registry.templates;
  return /* @__PURE__ */ (0, import_jsx_runtime16.jsxs)(import_jsx_runtime16.Fragment, { children: [
    title && /* @__PURE__ */ (0, import_jsx_runtime16.jsx)(
      TitleFieldTemplate,
      {
        id: (0, import_utils19.titleId)(fieldPathId),
        title,
        required,
        schema,
        uiSchema,
        registry,
        optionalDataControl: showOptionalDataControlInTitle ? optionalDataControl : void 0
      }
    ),
    description && /* @__PURE__ */ (0, import_jsx_runtime16.jsx)(
      DescriptionFieldTemplate,
      {
        id: (0, import_utils19.descriptionId)(fieldPathId),
        description,
        schema,
        uiSchema,
        registry
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime16.jsxs)("div", { className: "flex flex-col gap-2", children: [
      !showOptionalDataControlInTitle ? optionalDataControl : void 0,
      properties.map((element, index) => /* @__PURE__ */ (0, import_jsx_runtime16.jsx)("div", { className: `${element.hidden ? "hidden" : ""} flex`, children: /* @__PURE__ */ (0, import_jsx_runtime16.jsx)("div", { className: "w-full", children: element.content }) }, index)),
      (0, import_utils19.canExpand)(schema, uiSchema, formData) ? /* @__PURE__ */ (0, import_jsx_runtime16.jsx)("div", { className: "mt-2 flex justify-end", children: /* @__PURE__ */ (0, import_jsx_runtime16.jsx)(
        AddButton2,
        {
          id: (0, import_utils19.buttonId)(fieldPathId, "add"),
          onClick: onAddProperty,
          disabled: disabled || readonly,
          className: "rjsf-object-property-expand",
          uiSchema,
          registry
        }
      ) }) : null
    ] })
  ] });
}

// src/OptionalDataControlsTemplate/OptionalDataControlsTemplate.tsx
var import_lucide_react4 = require("lucide-react");
var import_jsx_runtime17 = require("react/jsx-runtime");
function OptionalDataControlsTemplate(props) {
  const { id, registry, label, onAddClick, onRemoveClick } = props;
  if (onAddClick) {
    return /* @__PURE__ */ (0, import_jsx_runtime17.jsx)(
      IconButton,
      {
        id,
        registry,
        className: "rjsf-add-optional-data",
        onClick: onAddClick,
        title: label,
        icon: /* @__PURE__ */ (0, import_jsx_runtime17.jsx)(import_lucide_react4.PlusCircle, {}),
        size: "xs"
      }
    );
  } else if (onRemoveClick) {
    return /* @__PURE__ */ (0, import_jsx_runtime17.jsx)(
      RemoveButton,
      {
        id,
        registry,
        className: "rjsf-remove-optional-data",
        onClick: onRemoveClick,
        title: label,
        size: "xs"
      }
    );
  }
  return /* @__PURE__ */ (0, import_jsx_runtime17.jsx)("em", { id, children: label });
}

// src/SubmitButton/SubmitButton.tsx
var import_utils20 = require("@rjsf/utils");
var import_jsx_runtime18 = require("react/jsx-runtime");
function SubmitButton(props) {
  const { submitText, norender, props: submitButtonProps } = (0, import_utils20.getSubmitButtonOptions)(props.uiSchema);
  if (norender) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime18.jsx)("div", { children: /* @__PURE__ */ (0, import_jsx_runtime18.jsx)(Button, { type: "submit", ...submitButtonProps, className: cn("my-2", submitButtonProps?.className), children: submitText }) });
}

// src/TitleField/TitleField.tsx
var import_utils23 = require("@rjsf/utils");

// src/components/ui/separator.tsx
var import_react_separator = require("@radix-ui/react-separator");
var import_jsx_runtime19 = require("react/jsx-runtime");
function Separator({
  className,
  orientation = "horizontal",
  decorative = true,
  ...props
}) {
  return /* @__PURE__ */ (0, import_jsx_runtime19.jsx)(
    import_react_separator.Root,
    {
      "data-slot": "separator",
      decorative,
      orientation,
      className: cn(
        "bg-border shrink-0 data-[orientation=horizontal]:h-px data-[orientation=horizontal]:w-full data-[orientation=vertical]:h-full data-[orientation=vertical]:w-px",
        className
      ),
      ...props
    }
  );
}

// src/TitleField/TitleField.tsx
var import_jsx_runtime20 = require("react/jsx-runtime");
function TitleField({
  id,
  title,
  uiSchema,
  optionalDataControl
}) {
  const uiOptions = (0, import_utils23.getUiOptions)(uiSchema);
  let heading = /* @__PURE__ */ (0, import_jsx_runtime20.jsx)("h5", { children: uiOptions.title || title });
  if (optionalDataControl) {
    heading = /* @__PURE__ */ (0, import_jsx_runtime20.jsxs)("div", { className: "flex flex-row", children: [
      /* @__PURE__ */ (0, import_jsx_runtime20.jsx)("div", { className: "flex-grow", children: heading }),
      /* @__PURE__ */ (0, import_jsx_runtime20.jsx)("div", { className: "flex", children: optionalDataControl })
    ] });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime20.jsxs)("div", { id, className: "my-1 flex flex-col gap-0.5", children: [
    heading,
    /* @__PURE__ */ (0, import_jsx_runtime20.jsx)(Separator, { dir: "horizontal", style: { height: "1px" }, className: "my-1" })
  ] });
}

// src/WrapIfAdditionalTemplate/WrapIfAdditionalTemplate.tsx
var import_utils24 = require("@rjsf/utils");
var import_jsx_runtime21 = require("react/jsx-runtime");
function WrapIfAdditionalTemplate({
  classNames,
  style,
  children,
  disabled,
  id,
  label,
  displayLabel,
  onRemoveProperty,
  onKeyRenameBlur,
  rawDescription,
  readonly,
  required,
  schema,
  uiSchema,
  registry
}) {
  const { templates, translateString } = registry;
  const { RemoveButton: RemoveButton2 } = templates.ButtonTemplates;
  const keyLabel = translateString(import_utils24.TranslatableString.KeyLabel, [label]);
  const additional = import_utils24.ADDITIONAL_PROPERTY_FLAG in schema;
  if (!additional) {
    return /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("div", { className: classNames, style, children });
  }
  const marginDesc = rawDescription ? -28 : 0;
  const margin = displayLabel ? 22 + marginDesc : 0;
  const keyId = `${id}-key`;
  return /* @__PURE__ */ (0, import_jsx_runtime21.jsxs)(import_jsx_runtime21.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime21.jsxs)("div", { className: `grid grid-cols-12 col-span-12 items-center gap-2 ${classNames}`, style, children: [
      /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("div", { className: "grid gap-2 col-span-5", children: /* @__PURE__ */ (0, import_jsx_runtime21.jsxs)("div", { className: "flex flex-col gap-2", children: [
        displayLabel && /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("label", { htmlFor: keyId, className: "text-sm font-medium text-muted-foreground leading-none", children: keyLabel }),
        /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("div", { className: "pl-0.5", children: /* @__PURE__ */ (0, import_jsx_runtime21.jsx)(
          Input,
          {
            required,
            defaultValue: label,
            disabled: disabled || readonly,
            id: keyId,
            name: keyId,
            onBlur: !readonly ? onKeyRenameBlur : void 0,
            type: "text",
            className: "w-full border shadow-sm"
          }
        ) }),
        !!rawDescription && /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("span", { className: "text-xs font-medium text-muted-foreground", children: /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("div", { className: "text-sm text-muted-foreground", children: "\xA0" }) })
      ] }) }),
      /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("div", { className: "grid gap-2 col-span-6 pr-0.5", children }),
      /* @__PURE__ */ (0, import_jsx_runtime21.jsx)("div", { className: "grid gap-2 col-span-1", style: { marginTop: `${margin}px` }, children: /* @__PURE__ */ (0, import_jsx_runtime21.jsx)(
        RemoveButton2,
        {
          id: (0, import_utils24.buttonId)(id, "remove"),
          iconType: "block",
          className: "rjsf-object-property-remove w-full",
          disabled: disabled || readonly,
          onClick: onRemoveProperty,
          uiSchema,
          registry
        }
      ) })
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime21.jsx)(Separator, { dir: "horizontal", className: "mt-2" })
  ] });
}

// src/Templates/Templates.ts
function generateTemplates() {
  return {
    ArrayFieldItemTemplate,
    ArrayFieldTemplate,
    BaseInputTemplate,
    ButtonTemplates: {
      AddButton,
      CopyButton,
      MoveDownButton,
      MoveUpButton,
      RemoveButton,
      SubmitButton
    },
    DescriptionFieldTemplate: DescriptionField,
    ErrorListTemplate: ErrorList,
    FieldErrorTemplate,
    FieldHelpTemplate,
    FieldTemplate,
    GridTemplate,
    MultiSchemaFieldTemplate,
    ObjectFieldTemplate,
    OptionalDataControlsTemplate,
    TitleFieldTemplate: TitleField,
    WrapIfAdditionalTemplate
  };
}
var Templates_default = generateTemplates();

// src/CheckboxWidget/CheckboxWidget.tsx
var import_utils27 = require("@rjsf/utils");

// src/components/ui/checkbox.tsx
var import_react_checkbox = require("@radix-ui/react-checkbox");
var import_react_icons = require("@radix-ui/react-icons");
var import_jsx_runtime22 = require("react/jsx-runtime");
function Checkbox({ className, ...props }) {
  return /* @__PURE__ */ (0, import_jsx_runtime22.jsx)(
    import_react_checkbox.Root,
    {
      "data-slot": "checkbox",
      className: cn(
        "peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50",
        className
      ),
      ...props,
      children: /* @__PURE__ */ (0, import_jsx_runtime22.jsx)(
        import_react_checkbox.Indicator,
        {
          "data-slot": "checkbox-indicator",
          className: "flex items-center justify-center text-current transition-none",
          children: /* @__PURE__ */ (0, import_jsx_runtime22.jsx)(import_react_icons.CheckIcon, { className: "size-3.5" })
        }
      )
    }
  );
}

// src/components/ui/label.tsx
var import_react_label = require("@radix-ui/react-label");
var import_jsx_runtime23 = require("react/jsx-runtime");
function Label({ className, ...props }) {
  return /* @__PURE__ */ (0, import_jsx_runtime23.jsx)(
    import_react_label.Root,
    {
      "data-slot": "label",
      className: cn(
        "flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50",
        className
      ),
      ...props
    }
  );
}

// src/CheckboxWidget/CheckboxWidget.tsx
var import_jsx_runtime24 = require("react/jsx-runtime");
function CheckboxWidget(props) {
  const {
    id,
    htmlName,
    value,
    disabled,
    readonly,
    label,
    hideLabel,
    schema,
    autofocus,
    options,
    onChange,
    onBlur,
    onFocus,
    registry,
    uiSchema,
    className
  } = props;
  const required = (0, import_utils27.schemaRequiresTrueValue)(schema);
  const DescriptionFieldTemplate = (0, import_utils27.getTemplate)(
    "DescriptionFieldTemplate",
    registry,
    options
  );
  const _onChange = (checked) => onChange(checked);
  const _onBlur = () => onBlur(id, value);
  const _onFocus = () => onFocus(id, value);
  const description = options.description || schema.description;
  return /* @__PURE__ */ (0, import_jsx_runtime24.jsxs)(
    "div",
    {
      className: `relative ${disabled || readonly ? "cursor-not-allowed opacity-50" : ""}`,
      "aria-describedby": (0, import_utils27.ariaDescribedByIds)(id),
      children: [
        !hideLabel && description && /* @__PURE__ */ (0, import_jsx_runtime24.jsx)(
          DescriptionFieldTemplate,
          {
            id: (0, import_utils27.descriptionId)(id),
            description,
            schema,
            uiSchema,
            registry
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime24.jsxs)("div", { className: "flex items-center gap-2 my-2", children: [
          /* @__PURE__ */ (0, import_jsx_runtime24.jsx)(
            Checkbox,
            {
              id,
              name: htmlName || id,
              checked: typeof value === "undefined" ? false : Boolean(value),
              required,
              disabled: disabled || readonly,
              autoFocus: autofocus,
              onCheckedChange: _onChange,
              onBlur: _onBlur,
              onFocus: _onFocus,
              className
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime24.jsx)(Label, { className: "leading-tight", htmlFor: id, children: (0, import_utils27.labelValue)(label, hideLabel || !label) })
        ] })
      ]
    }
  );
}

// src/CheckboxesWidget/CheckboxesWidget.tsx
var import_utils28 = require("@rjsf/utils");
var import_jsx_runtime25 = require("react/jsx-runtime");
function CheckboxesWidget({
  id,
  htmlName,
  disabled,
  options,
  value,
  autofocus,
  readonly,
  required,
  onChange,
  onBlur,
  onFocus,
  className
}) {
  const { enumOptions, enumDisabled, inline, emptyValue } = options;
  const checkboxesValues = Array.isArray(value) ? value : [value];
  const _onBlur = ({ target }) => onBlur(id, (0, import_utils28.enumOptionsValueForIndex)(target && target.value, enumOptions, emptyValue));
  const _onFocus = ({ target }) => onFocus(id, (0, import_utils28.enumOptionsValueForIndex)(target && target.value, enumOptions, emptyValue));
  return /* @__PURE__ */ (0, import_jsx_runtime25.jsx)("div", { className: cn({ "flex flex-col gap-2": !inline, "flex flex-row gap-4 flex-wrap": inline }), children: Array.isArray(enumOptions) && enumOptions.map((option, index) => {
    const checked = (0, import_utils28.enumOptionsIsSelected)(option.value, checkboxesValues);
    const itemDisabled = Array.isArray(enumDisabled) && enumDisabled.indexOf(option.value) !== -1;
    const indexOptionId = (0, import_utils28.optionId)(id, index);
    return /* @__PURE__ */ (0, import_jsx_runtime25.jsxs)("div", { className: "flex items-center gap-2", children: [
      /* @__PURE__ */ (0, import_jsx_runtime25.jsx)(
        Checkbox,
        {
          id: indexOptionId,
          name: htmlName || id,
          required,
          disabled: disabled || itemDisabled || readonly,
          onCheckedChange: (state) => {
            if (state) {
              onChange((0, import_utils28.enumOptionsSelectValue)(index, checkboxesValues, enumOptions));
            } else {
              onChange((0, import_utils28.enumOptionsDeselectValue)(index, checkboxesValues, enumOptions));
            }
          },
          className,
          checked,
          autoFocus: autofocus && index === 0,
          onBlur: _onBlur,
          onFocus: _onFocus,
          "aria-describedby": (0, import_utils28.ariaDescribedByIds)(id)
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime25.jsx)(Label, { className: "leading-tight", htmlFor: (0, import_utils28.optionId)(id, index), children: option.label })
    ] }, indexOptionId);
  }) });
}

// src/RadioWidget/RadioWidget.tsx
var import_utils31 = require("@rjsf/utils");

// src/components/ui/radio-group.tsx
var import_react_radio_group = require("@radix-ui/react-radio-group");
var import_lucide_react5 = require("lucide-react");
var import_jsx_runtime26 = require("react/jsx-runtime");
function RadioGroup({ className, ...props }) {
  return /* @__PURE__ */ (0, import_jsx_runtime26.jsx)(import_react_radio_group.Root, { "data-slot": "radio-group", className: cn("grid gap-3", className), ...props });
}
function RadioGroupItem({ className, ...props }) {
  return /* @__PURE__ */ (0, import_jsx_runtime26.jsx)(
    import_react_radio_group.Item,
    {
      "data-slot": "radio-group-item",
      className: cn(
        "border-input text-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 aspect-square size-4 shrink-0 rounded-full border shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50",
        className
      ),
      ...props,
      children: /* @__PURE__ */ (0, import_jsx_runtime26.jsx)(import_react_radio_group.Indicator, { "data-slot": "radio-group-indicator", className: "relative flex items-center justify-center", children: /* @__PURE__ */ (0, import_jsx_runtime26.jsx)(import_lucide_react5.CircleIcon, { className: "fill-primary absolute top-1/2 left-1/2 size-2 -translate-x-1/2 -translate-y-1/2" }) })
    }
  );
}

// src/RadioWidget/RadioWidget.tsx
var import_jsx_runtime27 = require("react/jsx-runtime");
function RadioWidget({
  id,
  options,
  value,
  required,
  disabled,
  readonly,
  onChange,
  onBlur,
  onFocus,
  className
}) {
  const { enumOptions, enumDisabled, emptyValue } = options;
  const _onChange = (value2) => onChange((0, import_utils31.enumOptionsValueForIndex)(value2, enumOptions, emptyValue));
  const _onBlur = ({ target }) => onBlur(id, (0, import_utils31.enumOptionsValueForIndex)(target && target.value, enumOptions, emptyValue));
  const _onFocus = ({ target }) => onFocus(id, (0, import_utils31.enumOptionsValueForIndex)(target && target.value, enumOptions, emptyValue));
  const inline = Boolean(options && options.inline);
  return /* @__PURE__ */ (0, import_jsx_runtime27.jsx)("div", { className: "mb-0", children: /* @__PURE__ */ (0, import_jsx_runtime27.jsx)(
    RadioGroup,
    {
      defaultValue: value?.toString(),
      required,
      disabled: disabled || readonly,
      onValueChange: (e) => {
        _onChange(e);
      },
      onBlur: _onBlur,
      onFocus: _onFocus,
      "aria-describedby": (0, import_utils31.ariaDescribedByIds)(id),
      orientation: inline ? "horizontal" : "vertical",
      className: cn("flex flex-wrap", { "flex-col": !inline }, className),
      children: Array.isArray(enumOptions) && enumOptions.map((option, index) => {
        const itemDisabled = Array.isArray(enumDisabled) && enumDisabled.indexOf(option.value) !== -1;
        const checked = (0, import_utils31.enumOptionsIsSelected)(option.value, value);
        return /* @__PURE__ */ (0, import_jsx_runtime27.jsxs)("div", { className: "flex items-center gap-2", children: [
          /* @__PURE__ */ (0, import_jsx_runtime27.jsx)(
            RadioGroupItem,
            {
              checked,
              value: index.toString(),
              id: (0, import_utils31.optionId)(id, index),
              disabled: itemDisabled
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime27.jsx)(Label, { className: "leading-tight", htmlFor: (0, import_utils31.optionId)(id, index), children: option.label })
        ] }, (0, import_utils31.optionId)(id, index));
      })
    }
  ) });
}

// src/RangeWidget/RangeWidget.tsx
var import_utils34 = require("@rjsf/utils");
var import_pick = __toESM(require("lodash/pick"), 1);

// src/components/ui/slider.tsx
var import_react_slider = require("@radix-ui/react-slider");
var import_react = require("react");
var import_jsx_runtime28 = require("react/jsx-runtime");
function Slider({ className, defaultValue, value, min = 0, max = 100, ...props }) {
  const _values = (0, import_react.useMemo)(
    () => Array.isArray(value) ? value : Array.isArray(defaultValue) ? defaultValue : [min, max],
    [value, defaultValue, min, max]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime28.jsxs)(
    import_react_slider.Root,
    {
      "data-slot": "slider",
      defaultValue,
      value,
      min,
      max,
      className: cn(
        "relative flex w-full touch-none items-center select-none data-[disabled]:opacity-50 data-[orientation=vertical]:h-full data-[orientation=vertical]:min-h-44 data-[orientation=vertical]:w-auto data-[orientation=vertical]:flex-col",
        className
      ),
      ...props,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime28.jsx)(
          import_react_slider.Track,
          {
            "data-slot": "slider-track",
            className: cn(
              "bg-muted relative grow overflow-hidden rounded-full data-[orientation=horizontal]:h-1.5 data-[orientation=horizontal]:w-full data-[orientation=vertical]:h-full data-[orientation=vertical]:w-1.5"
            ),
            children: /* @__PURE__ */ (0, import_jsx_runtime28.jsx)(
              import_react_slider.Range,
              {
                "data-slot": "slider-range",
                className: cn("bg-primary absolute data-[orientation=horizontal]:h-full data-[orientation=vertical]:w-full")
              }
            )
          }
        ),
        Array.from({ length: _values.length }, (_, index) => /* @__PURE__ */ (0, import_jsx_runtime28.jsx)(
          import_react_slider.Thumb,
          {
            "data-slot": "slider-thumb",
            className: "border-primary bg-background ring-ring/50 block size-4 shrink-0 rounded-full border shadow-sm transition-[color,box-shadow] hover:ring-4 focus-visible:ring-4 focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50"
          },
          index
        ))
      ]
    }
  );
}

// src/RangeWidget/RangeWidget.tsx
var import_jsx_runtime29 = require("react/jsx-runtime");
var allowedProps = [
  "name",
  "min",
  "max",
  "step",
  "orientation",
  "disabled",
  "defaultValue",
  "value",
  "onValueChange",
  "className",
  "dir",
  "inverted",
  "minStepsBetweenThumbs"
];
function RangeWidget({
  value,
  readonly,
  disabled,
  options,
  schema,
  onChange,
  label,
  id
}) {
  const _onChange = (value2) => onChange(value2[0]);
  const sliderProps = { value, label, id, ...(0, import_utils34.rangeSpec)(schema) };
  const uiProps = { id, ...(0, import_pick.default)(options.props || {}, allowedProps) };
  return /* @__PURE__ */ (0, import_jsx_runtime29.jsxs)(import_jsx_runtime29.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime29.jsx)(
      Slider,
      {
        disabled: disabled || readonly,
        min: sliderProps.min,
        max: sliderProps.max,
        step: sliderProps.step,
        value: [value],
        onValueChange: _onChange,
        ...uiProps,
        "aria-describedby": (0, import_utils34.ariaDescribedByIds)(id)
      }
    ),
    value
  ] });
}

// src/SelectWidget/SelectWidget.tsx
var import_utils40 = require("@rjsf/utils");

// src/components/ui/fancy-multi-select.tsx
var import_cmdk2 = require("cmdk");
var import_isEqual = __toESM(require("lodash/isEqual"), 1);
var import_lucide_react8 = require("lucide-react");
var import_react3 = require("react");

// src/components/ui/badge.tsx
var import_class_variance_authority3 = require("class-variance-authority");
var import_react_slot2 = require("@radix-ui/react-slot");
var import_jsx_runtime30 = require("react/jsx-runtime");
var badgeVariants = (0, import_class_variance_authority3.cva)(
  "inline-flex items-center justify-center rounded-md border px-2 py-0.5 text-xs font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden",
  {
    variants: {
      variant: {
        default: "border-transparent bg-primary text-primary-foreground [a&]:hover:bg-primary/90",
        secondary: "border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90",
        destructive: "border-transparent bg-destructive text-white [a&]:hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60",
        outline: "text-foreground [a&]:hover:bg-accent [a&]:hover:text-accent-foreground"
      }
    },
    defaultVariants: {
      variant: "default"
    }
  }
);
function Badge({
  className,
  variant,
  asChild = false,
  ...props
}) {
  const Comp = asChild ? import_react_slot2.Slot : "span";
  return /* @__PURE__ */ (0, import_jsx_runtime30.jsx)(Comp, { "data-slot": "badge", className: cn(badgeVariants({ variant }), className), ...props });
}

// src/components/ui/command.tsx
var import_cmdk = require("cmdk");
var import_lucide_react7 = require("lucide-react");
var import_react2 = require("react");

// src/components/ui/dialog.tsx
var import_react_dialog = require("@radix-ui/react-dialog");
var import_lucide_react6 = require("lucide-react");
var import_jsx_runtime31 = require("react/jsx-runtime");

// src/components/ui/command.tsx
var import_jsx_runtime32 = require("react/jsx-runtime");
var Command = (0, import_react2.forwardRef)(
  ({ className, ...props }, ref) => {
    return /* @__PURE__ */ (0, import_jsx_runtime32.jsx)(
      import_cmdk.Command,
      {
        ref,
        "data-slot": "command",
        className: cn(
          "bg-popover text-popover-foreground flex h-full w-full flex-col overflow-hidden rounded-md",
          className
        ),
        ...props
      }
    );
  }
);
var CommandInput = (0, import_react2.forwardRef)(({ className, ...props }, ref) => /* @__PURE__ */ (0, import_jsx_runtime32.jsxs)("div", { className: "flex items-center border-b px-3", ...{ "cmdk-input-wrapper": "" }, children: [
  /* @__PURE__ */ (0, import_jsx_runtime32.jsx)(import_lucide_react7.Search, { className: "mr-2 h-4 w-4 shrink-0 opacity-50" }),
  /* @__PURE__ */ (0, import_jsx_runtime32.jsx)(
    import_cmdk.Command.Input,
    {
      ref,
      className: cn(
        "flex h-10 w-full rounded-md bg-transparent py-3 text-sm outline-none placeholder:text-muted-foreground disabled:cursor-not-allowed disabled:opacity-50",
        className
      ),
      ...props
    }
  )
] }));
CommandInput.displayName = import_cmdk.Command.Input.displayName;
function CommandList({ className, ...props }) {
  return /* @__PURE__ */ (0, import_jsx_runtime32.jsx)(
    import_cmdk.Command.List,
    {
      "data-slot": "command-list",
      className: cn("max-h-[300px] scroll-py-1 overflow-x-hidden overflow-y-auto", className),
      ...props
    }
  );
}
function CommandGroup({ className, ...props }) {
  return /* @__PURE__ */ (0, import_jsx_runtime32.jsx)(
    import_cmdk.Command.Group,
    {
      "data-slot": "command-group",
      className: cn(
        "text-foreground [&_[cmdk-group-heading]]:text-muted-foreground overflow-hidden p-1 [&_[cmdk-group-heading]]:px-2 [&_[cmdk-group-heading]]:py-1.5 [&_[cmdk-group-heading]]:text-xs [&_[cmdk-group-heading]]:font-medium",
        className
      ),
      ...props
    }
  );
}
function CommandItem({ className, ...props }) {
  return /* @__PURE__ */ (0, import_jsx_runtime32.jsx)(
    import_cmdk.Command.Item,
    {
      "data-slot": "command-item",
      className: cn(
        "data-[selected=true]:bg-accent data-[selected=true]:text-accent-foreground [&_svg:not([class*='text-'])]:text-muted-foreground relative flex cursor-default items-center gap-2 rounded-sm px-2 py-1.5 text-sm outline-hidden select-none data-[disabled=true]:pointer-events-none data-[disabled=true]:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
        className
      ),
      ...props
    }
  );
}

// src/components/ui/fancy-multi-select.tsx
var import_jsx_runtime33 = require("react/jsx-runtime");
function FancyMultiSelect({
  multiple,
  items = [],
  selected,
  onValueChange,
  autoFocus = false,
  disabled = false,
  ariaDescribedby,
  ariaPlaceholder,
  onFocus,
  onBlur,
  className,
  id
}) {
  const inputRef = (0, import_react3.useRef)(null);
  const [open, setOpen] = (0, import_react3.useState)(false);
  const [inputValue, setInputValue] = (0, import_react3.useState)("");
  const selectedItems = (0, import_react3.useMemo)(
    () => items.filter((item) => selected.some((selectedValue) => (0, import_isEqual.default)(item.value, selectedValue))),
    [items, selected]
  );
  const selectables = (0, import_react3.useMemo)(
    () => items.filter((framework) => !selectedItems.some((item) => (0, import_isEqual.default)(item.value, framework.value))),
    [items, selectedItems]
  );
  const handleUnselect = (0, import_react3.useCallback)(
    (framework) => {
      if (disabled) {
        return;
      }
      const newSelected = selectedItems.filter((s) => !(0, import_isEqual.default)(s.value, framework.value));
      onValueChange?.(newSelected.map((item) => item.index));
    },
    [selectedItems, onValueChange, disabled]
  );
  const handleKeyDown = (0, import_react3.useCallback)(
    (e) => {
      if (disabled || !inputRef.current || inputRef.current.value !== "") {
        return;
      }
      if (e.key === "Delete" || e.key === "Backspace") {
        const newSelected = selectedItems.slice(0, -1);
        onValueChange?.(newSelected.map((item) => item.index));
      } else if (e.key === "Escape") {
        inputRef.current.blur();
      }
    },
    [selectedItems, onValueChange, disabled]
  );
  const handleSelect = (0, import_react3.useCallback)(
    (item) => {
      if (disabled) {
        return;
      }
      setInputValue("");
      const newSelected = multiple ? [...selectedItems, item] : [item];
      onValueChange?.(newSelected.map((item2) => item2.index));
    },
    [multiple, selectedItems, onValueChange, disabled]
  );
  const handleFocus = (0, import_react3.useCallback)(
    (e) => {
      if (!disabled) {
        setOpen(true);
      }
      onFocus?.(e);
    },
    [disabled, onFocus]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime33.jsxs)(
    Command,
    {
      onKeyDown: handleKeyDown,
      className: cn("overflow-visible bg-transparent", className),
      autoFocus,
      "aria-disabled": disabled,
      onBlur,
      onFocus: handleFocus,
      "aria-describedby": ariaDescribedby,
      "aria-placeholder": ariaPlaceholder,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime33.jsx)(
          "div",
          {
            className: cn(
              "group border border-input px-3 py-2 text-sm ring-offset-background rounded-md focus-within:ring-1 focus-within:ring-ring focus-within:ring-offset-1",
              disabled && "opacity-50 cursor-not-allowed"
            ),
            children: /* @__PURE__ */ (0, import_jsx_runtime33.jsxs)("div", { className: "flex gap-1 flex-wrap", children: [
              selectedItems.map((item) => /* @__PURE__ */ (0, import_jsx_runtime33.jsxs)(Badge, { variant: "secondary", children: [
                item.label,
                /* @__PURE__ */ (0, import_jsx_runtime33.jsx)(
                  "button",
                  {
                    type: "button",
                    className: "rtl:mr-1 ltr:ml-1 ring-offset-background rounded-full outline-none focus:ring-1 focus:ring-ring focus:ring-offset-1",
                    onKeyDown: (e) => e.key === "Enter" && !disabled && handleUnselect(item),
                    onMouseDown: (e) => {
                      e.preventDefault();
                      e.stopPropagation();
                    },
                    onClick: () => handleUnselect(item),
                    disabled,
                    children: /* @__PURE__ */ (0, import_jsx_runtime33.jsx)(
                      import_lucide_react8.X,
                      {
                        className: cn(
                          "h-3 w-3 text-muted-foreground hover:text-foreground",
                          disabled && "pointer-events-none"
                        )
                      }
                    )
                  }
                )
              ] }, item.value)),
              /* @__PURE__ */ (0, import_jsx_runtime33.jsx)(
                import_cmdk2.Command.Input,
                {
                  ref: inputRef,
                  value: inputValue,
                  onValueChange: setInputValue,
                  onBlur: () => setOpen(false),
                  onFocus: () => !disabled && setOpen(true),
                  placeholder: "Select ...",
                  className: "rtl:mr-2 ltr:ml-2 bg-transparent outline-none placeholder:text-muted-foreground flex-1",
                  disabled,
                  "aria-controls": `command-item-input-${id}`,
                  "aria-labelledby": `command-item-input-${id}`,
                  id: `command-item-input-${id}`
                }
              )
            ] })
          }
        ),
        open && !disabled && selectables.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime33.jsx)("div", { className: "relative mt-2", children: /* @__PURE__ */ (0, import_jsx_runtime33.jsx)("div", { className: "absolute w-full z-10 top-0 rounded-md border bg-popover text-popover-foreground shadow-md outline-none animate-in", children: /* @__PURE__ */ (0, import_jsx_runtime33.jsx)(CommandGroup, { className: "h-full overflow-auto", children: /* @__PURE__ */ (0, import_jsx_runtime33.jsx)(CommandList, { children: selectables.map((item) => /* @__PURE__ */ (0, import_jsx_runtime33.jsx)(
          CommandItem,
          {
            disabled: item.disabled,
            onMouseDown: (e) => {
              e.preventDefault();
              e.stopPropagation();
            },
            "aria-controls": `${item.value}-command-item`,
            "aria-labelledby": `${item.value}-command-item`,
            id: `${item.value}-command-item`,
            onSelect: () => handleSelect(item),
            className: "cursor-pointer",
            children: item.label
          },
          `${item.value}-command-item`
        )) }) }) }) })
      ]
    }
  );
}

// src/components/ui/fancy-select.tsx
var import_lucide_react9 = require("lucide-react");
var import_react4 = require("react");
var import_jsx_runtime34 = require("react/jsx-runtime");
function FancySelect({
  items,
  selected,
  onValueChange,
  autoFocus = false,
  disabled = false,
  required = false,
  placeholder = "Select...",
  ariaDescribedby,
  ariaPlaceholder,
  onFocus,
  onBlur,
  className
}) {
  const [open, setOpen] = (0, import_react4.useState)(false);
  const selectedItem = items?.find((item) => item.value === selected);
  const selectedRef = (0, import_react4.useRef)(null);
  const containerRef = (0, import_react4.useRef)(null);
  (0, import_react4.useEffect)(() => {
    if (open && selectedRef.current) {
      requestAnimationFrame(() => {
        selectedRef.current?.scrollIntoView({ block: "nearest" });
      });
    }
  }, [open]);
  const handleBlur = (e) => {
    if (containerRef.current && !containerRef.current.contains(e.relatedTarget)) {
      setOpen(false);
    }
    onBlur?.(e);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime34.jsxs)(
    Command,
    {
      ref: containerRef,
      className: cn("overflow-visible bg-transparent", className),
      autoFocus,
      "aria-disabled": disabled,
      onBlur: handleBlur,
      onFocus,
      "aria-describedby": ariaDescribedby,
      "aria-placeholder": ariaPlaceholder,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime34.jsxs)(
          "div",
          {
            onClick: () => !disabled && setOpen(!open),
            className: cn(
              "flex h-9 w-full items-center justify-between gap-2 whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50",
              !selectedItem && required && "border-red-500",
              disabled && "opacity-50 cursor-not-allowed",
              className
            ),
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime34.jsx)("span", { className: cn("flex-1 line-clamp-1", !selectedItem && "text-muted-foreground"), children: selectedItem?.label || placeholder }),
              /* @__PURE__ */ (0, import_jsx_runtime34.jsx)(import_lucide_react9.ChevronDown, { className: "h-4 w-4 opacity-50" })
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime34.jsx)("div", { className: "relative", children: open && items && items.length > 0 ? /* @__PURE__ */ (0, import_jsx_runtime34.jsx)(
          "div",
          {
            style: { top: "0.5rem" },
            className: "absolute w-full z-10 rounded-md border bg-popover text-popover-foreground shadow-md outline-none",
            children: /* @__PURE__ */ (0, import_jsx_runtime34.jsx)(CommandGroup, { className: "h-full overflow-auto", children: /* @__PURE__ */ (0, import_jsx_runtime34.jsx)(CommandList, { children: items.map((item) => /* @__PURE__ */ (0, import_jsx_runtime34.jsxs)(
              CommandItem,
              {
                ref: item.value === selected ? selectedRef : void 0,
                onMouseDown: (e) => {
                  e.preventDefault();
                  e.stopPropagation();
                },
                onSelect: () => {
                  if (!item.disabled) {
                    onValueChange?.(item.value);
                    setOpen(false);
                  }
                },
                className: cn(
                  "cursor-pointer relative flex items-center justify-between rounded-sm py-1.5 gap-2 rtl:flex-row-reverse",
                  item.value === selected && "font-semibold",
                  item.disabled && "opacity-50 cursor-not-allowed"
                ),
                children: [
                  /* @__PURE__ */ (0, import_jsx_runtime34.jsx)("span", { children: item.label }),
                  /* @__PURE__ */ (0, import_jsx_runtime34.jsx)("span", { className: "flex h-3.5 w-3.5 items-center justify-center", children: item.value === selected && /* @__PURE__ */ (0, import_jsx_runtime34.jsx)(import_lucide_react9.Check, { className: "h-4 w-4" }) })
                ]
              },
              item.value
            )) }) })
          }
        ) : null })
      ]
    }
  );
}

// src/SelectWidget/SelectWidget.tsx
var import_jsx_runtime35 = require("react/jsx-runtime");
function SelectWidget({
  id,
  options,
  required,
  disabled,
  readonly,
  value,
  multiple,
  autofocus,
  onChange,
  onBlur,
  onFocus,
  defaultValue,
  placeholder,
  rawErrors = [],
  className
}) {
  const { enumOptions, enumDisabled, emptyValue: optEmptyValue } = options;
  const _onFancyFocus = () => {
    onFocus(id, (0, import_utils40.enumOptionsValueForIndex)(value, enumOptions, optEmptyValue));
  };
  const _onFancyBlur = () => {
    onBlur(id, (0, import_utils40.enumOptionsValueForIndex)(value, enumOptions, optEmptyValue));
  };
  const items = enumOptions?.map(({ value: value2, label }, index) => ({
    value: multiple ? value2 : index.toString(),
    label,
    index,
    disabled: Array.isArray(enumDisabled) && enumDisabled.includes(value2)
  }));
  const cnClassName = cn({ "border-destructive": rawErrors.length > 0 }, className);
  return /* @__PURE__ */ (0, import_jsx_runtime35.jsx)("div", { className: "p-0.5", children: !multiple ? /* @__PURE__ */ (0, import_jsx_runtime35.jsx)(
    FancySelect,
    {
      items,
      selected: (0, import_utils40.enumOptionsIndexForValue)(value ?? defaultValue, enumOptions, false),
      onValueChange: (selectedValue) => {
        onChange((0, import_utils40.enumOptionsValueForIndex)(selectedValue, enumOptions, optEmptyValue));
      },
      autoFocus: autofocus,
      disabled: disabled || readonly,
      required,
      placeholder,
      className: cnClassName,
      onFocus: _onFancyFocus,
      onBlur: _onFancyBlur,
      ariaDescribedby: (0, import_utils40.ariaDescribedByIds)(id)
    }
  ) : /* @__PURE__ */ (0, import_jsx_runtime35.jsx)(
    FancyMultiSelect,
    {
      id,
      autoFocus: autofocus,
      disabled: disabled || readonly,
      multiple: true,
      className: cnClassName,
      items,
      selected: value,
      onValueChange: (values) => {
        onChange((0, import_utils40.enumOptionsValueForIndex)(values, enumOptions, optEmptyValue));
      },
      onFocus: _onFancyFocus,
      onBlur: _onFancyBlur
    }
  ) });
}

// src/TextareaWidget/TextareaWidget.tsx
var import_utils43 = require("@rjsf/utils");

// src/components/ui/textarea.tsx
var import_jsx_runtime36 = require("react/jsx-runtime");
function Textarea({ className, ...props }) {
  return /* @__PURE__ */ (0, import_jsx_runtime36.jsx)(
    "textarea",
    {
      "data-slot": "textarea",
      className: cn(
        "border-input placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 flex field-sizing-content min-h-16 w-full rounded-md border bg-transparent px-3 py-2 text-base shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 md:text-sm",
        className
      ),
      ...props
    }
  );
}

// src/TextareaWidget/TextareaWidget.tsx
var import_jsx_runtime37 = require("react/jsx-runtime");
function TextareaWidget({
  id,
  htmlName,
  placeholder,
  value,
  required,
  disabled,
  autofocus,
  readonly,
  onBlur,
  onFocus,
  onChange,
  options,
  className
}) {
  const _onChange = ({ target: { value: value2 } }) => onChange(value2 === "" ? options.emptyValue : value2);
  const _onBlur = ({ target }) => onBlur(id, target && target.value);
  const _onFocus = ({ target }) => onFocus(id, target && target.value);
  return /* @__PURE__ */ (0, import_jsx_runtime37.jsx)("div", { className: "flex p-0.5", children: /* @__PURE__ */ (0, import_jsx_runtime37.jsx)(
    Textarea,
    {
      id,
      name: htmlName || id,
      placeholder,
      disabled,
      readOnly: readonly,
      value: value ?? "",
      required,
      autoFocus: autofocus,
      rows: options.rows || 5,
      onChange: _onChange,
      onBlur: _onBlur,
      onFocus: _onFocus,
      "aria-describedby": (0, import_utils43.ariaDescribedByIds)(id),
      className
    }
  ) });
}

// src/Widgets/Widgets.ts
function generateWidgets() {
  return {
    CheckboxWidget,
    CheckboxesWidget,
    RadioWidget,
    RangeWidget,
    SelectWidget,
    TextareaWidget
  };
}
var Widgets_default = generateWidgets();

// src/Theme/Theme.tsx
function generateTheme() {
  return {
    templates: generateTemplates(),
    widgets: generateWidgets()
  };
}
var Theme_default = generateTheme();

// src/Form/Form.tsx
function generateForm() {
  return (0, import_core3.withTheme)(generateTheme());
}
var Form_default = generateForm();

// src/index.ts
var index_default = Form_default;
