// @rjsf/shadcn/dist/rjsf-shadcn.esm.js
// src/Form/Form.tsx
import { withTheme } from "@rjsf/core";

// src/AddButton/AddButton.tsx
import { TranslatableString } from "@rjsf/utils";
import { PlusCircle } from "lucide-react";

// src/components/ui/button.tsx
import { Slot } from "@radix-ui/react-slot";
import { cva } from "class-variance-authority";

// src/lib/utils.ts
import { clsx } from "clsx";
import { twMerge } from "tailwind-merge";
function cn(...inputs) {
  return twMerge(clsx(inputs));
}

// src/components/ui/button.tsx
import { jsx } from "react/jsx-runtime";
var buttonVariants = cva(
  "inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive",
  {
    variants: {
      variant: {
        default: "bg-primary text-primary-foreground shadow-xs hover:bg-primary/90",
        destructive: "bg-destructive text-white shadow-xs hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60",
        outline: "border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50",
        secondary: "bg-secondary text-secondary-foreground shadow-xs hover:bg-secondary/80",
        ghost: "hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50",
        link: "text-primary underline-offset-4 hover:underline"
      },
      size: {
        default: "h-9 px-4 py-2 has-[>svg]:px-3",
        sm: "h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5",
        lg: "h-10 rounded-md px-6 has-[>svg]:px-4",
        xs: "h-6 rounded-md gap-1 px-2 has-[>svg]:px-2",
        icon: "size-9"
      }
    },
    defaultVariants: {
      variant: "default",
      size: "default"
    }
  }
);
function Button({
  className,
  variant,
  size,
  asChild = false,
  ...props
}) {
  const Comp = asChild ? Slot : "button";
  return /* @__PURE__ */ jsx(Comp, { "data-slot": "button", className: cn(buttonVariants({ variant, size, className })), ...props });
}

// src/AddButton/AddButton.tsx
import { jsx as jsx2, jsxs } from "react/jsx-runtime";
function AddButton({
  uiSchema,
  registry,
  className,
  ...props
}) {
  const { translateString } = registry;
  return /* @__PURE__ */ jsx2("div", { className: "p-0 m-0", children: /* @__PURE__ */ jsxs(Button, { ...props, className: cn("w-fit gap-2", className), variant: "outline", type: "button", children: [
    /* @__PURE__ */ jsx2(PlusCircle, { size: 16 }),
    " ",
    translateString(TranslatableString.AddItemButton)
  ] }) });
}

// src/ArrayFieldItemTemplate/ArrayFieldItemTemplate.tsx
import {
  getTemplate,
  getUiOptions
} from "@rjsf/utils";
import { jsx as jsx3, jsxs as jsxs2 } from "react/jsx-runtime";
function ArrayFieldItemTemplate(props) {
  const { children, buttonsProps, displayLabel, hasDescription, hasToolbar, uiSchema, registry } = props;
  const uiOptions = getUiOptions(uiSchema);
  const ArrayFieldItemButtonsTemplate = getTemplate(
    "ArrayFieldItemButtonsTemplate",
    registry,
    uiOptions
  );
  const margin = hasDescription ? -6 : 22;
  return /* @__PURE__ */ jsx3("div", { children: /* @__PURE__ */ jsxs2("div", { className: "mb-2 flex flex-row flex-wrap items-center", children: [
    /* @__PURE__ */ jsx3("div", { className: "grow shrink", children }),
    /* @__PURE__ */ jsx3("div", { className: "flex items-end justify-end p-0.5", children: hasToolbar && /* @__PURE__ */ jsx3(
      "div",
      {
        className: "flex gap-2",
        style: { marginLeft: "5px", marginTop: displayLabel ? `${margin}px` : void 0 },
        children: /* @__PURE__ */ jsx3(ArrayFieldItemButtonsTemplate, { ...buttonsProps })
      }
    ) })
  ] }) });
}

// src/ArrayFieldTemplate/ArrayFieldTemplate.tsx
import {
  buttonId,
  getTemplate as getTemplate2,
  getUiOptions as getUiOptions2
} from "@rjsf/utils";
import { jsx as jsx4, jsxs as jsxs3 } from "react/jsx-runtime";
function ArrayFieldTemplate(props) {
  const {
    canAdd,
    disabled,
    fieldPathId,
    uiSchema,
    items,
    optionalDataControl,
    onAddClick,
    readonly,
    registry,
    required,
    schema,
    title
  } = props;
  const uiOptions = getUiOptions2(uiSchema);
  const ArrayFieldDescriptionTemplate = getTemplate2(
    "ArrayFieldDescriptionTemplate",
    registry,
    uiOptions
  );
  const ArrayFieldTitleTemplate = getTemplate2(
    "ArrayFieldTitleTemplate",
    registry,
    uiOptions
  );
  const showOptionalDataControlInTitle = !readonly && !disabled;
  const {
    ButtonTemplates: { AddButton: AddButton2 }
  } = registry.templates;
  return /* @__PURE__ */ jsx4("div", { children: /* @__PURE__ */ jsx4("div", { className: "m-0 flex p-0", children: /* @__PURE__ */ jsxs3("div", { className: "m-0 w-full p-0", children: [
    /* @__PURE__ */ jsx4(
      ArrayFieldTitleTemplate,
      {
        fieldPathId,
        title: uiOptions.title || title,
        schema,
        uiSchema,
        required,
        registry,
        optionalDataControl: showOptionalDataControlInTitle ? optionalDataControl : void 0
      }
    ),
    /* @__PURE__ */ jsx4(
      ArrayFieldDescriptionTemplate,
      {
        fieldPathId,
        description: uiOptions.description || schema.description,
        schema,
        uiSchema,
        registry
      }
    ),
    /* @__PURE__ */ jsxs3("div", { className: "p-0 m-0 w-full mb-2", children: [
      !showOptionalDataControlInTitle ? optionalDataControl : void 0,
      items,
      canAdd && /* @__PURE__ */ jsx4("div", { className: "mt-2 flex justify-end", children: /* @__PURE__ */ jsx4(
        AddButton2,
        {
          id: buttonId(fieldPathId, "add"),
          className: "rjsf-array-item-add",
          onClick: onAddClick,
          disabled: disabled || readonly,
          uiSchema,
          registry
        }
      ) })
    ] }, `array-item-list-${fieldPathId.$id}`)
  ] }) }) });
}

// src/BaseInputTemplate/BaseInputTemplate.tsx
import {
  ariaDescribedByIds,
  examplesId,
  getInputProps
} from "@rjsf/utils";

// src/components/ui/input.tsx
import { jsx as jsx5 } from "react/jsx-runtime";
function Input({ className, type, ...props }) {
  return /* @__PURE__ */ jsx5(
    "input",
    {
      type,
      "data-slot": "input",
      className: cn(
        "file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm",
        "focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px]",
        "aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive",
        className
      ),
      ...props
    }
  );
}

// src/BaseInputTemplate/BaseInputTemplate.tsx
import { jsx as jsx6, jsxs as jsxs4 } from "react/jsx-runtime";
function BaseInputTemplate({
  id,
  htmlName,
  placeholder,
  required,
  readonly,
  disabled,
  type,
  value,
  onChange,
  onChangeOverride,
  onBlur,
  onFocus,
  autofocus,
  options,
  schema,
  rawErrors = [],
  children,
  extraProps,
  className
}) {
  const inputProps = {
    ...extraProps,
    ...getInputProps(schema, type, options)
  };
  const _onChange = ({ target: { value: value2 } }) => onChange(value2 === "" ? options.emptyValue : value2);
  const _onBlur = ({ target }) => onBlur(id, target && target.value);
  const _onFocus = ({ target }) => onFocus(id, target && target.value);
  return /* @__PURE__ */ jsxs4("div", { className: "p-0.5", children: [
    /* @__PURE__ */ jsx6(
      Input,
      {
        id,
        name: htmlName || id,
        type,
        placeholder,
        autoFocus: autofocus,
        required,
        disabled,
        readOnly: readonly,
        className: cn({ "border-destructive focus-visible:ring-0": rawErrors.length > 0 }, className),
        list: schema.examples ? examplesId(id) : void 0,
        ...inputProps,
        value: value || value === 0 ? value : "",
        onChange: onChangeOverride || _onChange,
        onBlur: _onBlur,
        onFocus: _onFocus,
        "aria-describedby": ariaDescribedByIds(id, !!schema.examples)
      }
    ),
    children,
    Array.isArray(schema.examples) ? /* @__PURE__ */ jsx6("datalist", { id: examplesId(id), children: schema.examples.concat(schema.default && !schema.examples.includes(schema.default) ? [schema.default] : []).map((example) => {
      return /* @__PURE__ */ jsx6("option", { value: example }, example);
    }) }) : null
  ] });
}

// src/DescriptionField/DescriptionField.tsx
import { RichDescription } from "@rjsf/core";
import { jsx as jsx7 } from "react/jsx-runtime";
function DescriptionField({ id, description, registry, uiSchema }) {
  if (!description) {
    return null;
  }
  return /* @__PURE__ */ jsx7("div", { children: /* @__PURE__ */ jsx7("div", { id, className: "text-sm text-muted-foreground", children: /* @__PURE__ */ jsx7(RichDescription, { description, registry, uiSchema }) }) });
}

// src/ErrorList/ErrorList.tsx
import { TranslatableString as TranslatableString2 } from "@rjsf/utils";
import { AlertCircle } from "lucide-react";

// src/components/ui/alert.tsx
import { cva as cva2 } from "class-variance-authority";
import { jsx as jsx8 } from "react/jsx-runtime";
var alertVariants = cva2(
  "relative w-full rounded-lg border px-4 py-3 text-sm grid has-[>svg]:grid-cols-[calc(var(--spacing)*4)_1fr] grid-cols-[0_1fr] has-[>svg]:gap-x-3 gap-y-0.5 items-start [&>svg]:size-4 [&>svg]:translate-y-0.5 [&>svg]:text-current",
  {
    variants: {
      variant: {
        default: "bg-card text-card-foreground",
        destructive: "text-destructive bg-card [&>svg]:text-current *:data-[slot=alert-description]:text-destructive/90"
      }
    },
    defaultVariants: {
      variant: "default"
    }
  }
);
function Alert({ className, variant, ...props }) {
  return /* @__PURE__ */ jsx8("div", { "data-slot": "alert", role: "alert", className: cn(alertVariants({ variant }), className), ...props });
}
function AlertTitle({ className, ...props }) {
  return /* @__PURE__ */ jsx8(
    "div",
    {
      "data-slot": "alert-title",
      className: cn("col-start-2 line-clamp-1 min-h-4 font-medium tracking-tight", className),
      ...props
    }
  );
}
function AlertDescription({ className, ...props }) {
  return /* @__PURE__ */ jsx8(
    "div",
    {
      "data-slot": "alert-description",
      className: cn(
        "text-muted-foreground col-start-2 grid justify-items-start gap-1 text-sm [&_p]:leading-relaxed",
        className
      ),
      ...props
    }
  );
}

// src/ErrorList/ErrorList.tsx
import { jsx as jsx9, jsxs as jsxs5 } from "react/jsx-runtime";
function ErrorList({
  errors,
  registry
}) {
  const { translateString } = registry;
  return /* @__PURE__ */ jsxs5(Alert, { variant: "destructive", className: "mb-2", children: [
    /* @__PURE__ */ jsx9(AlertCircle, { className: "h-4 w-4" }),
    /* @__PURE__ */ jsx9(AlertTitle, { children: translateString(TranslatableString2.ErrorsLabel) }),
    /* @__PURE__ */ jsx9(AlertDescription, { className: "flex flex-col gap-1", children: errors.map((error, i) => {
      return /* @__PURE__ */ jsxs5("span", { children: [
        "\u2022 ",
        error.stack
      ] }, i);
    }) })
  ] });
}

// src/FieldErrorTemplate/FieldErrorTemplate.tsx
import { errorId } from "@rjsf/utils";
import { jsx as jsx10 } from "react/jsx-runtime";
function FieldErrorTemplate(props) {
  const { errors = [], fieldPathId } = props;
  if (errors.length === 0) {
    return null;
  }
  const id = errorId(fieldPathId);
  return /* @__PURE__ */ jsx10("div", { className: "flex flex-col gap-1", id, children: errors.map((error, i) => {
    return /* @__PURE__ */ jsx10("span", { className: "text-xs font-medium text-destructive mb-1", children: error }, i);
  }) });
}

// src/FieldHelpTemplate/FieldHelpTemplate.tsx
import { helpId } from "@rjsf/utils";
import { RichHelp } from "@rjsf/core";
import { jsx as jsx11 } from "react/jsx-runtime";
function FieldHelpTemplate(props) {
  const { fieldPathId, help, uiSchema, registry, hasErrors } = props;
  if (!help) {
    return null;
  }
  return /* @__PURE__ */ jsx11(
    "span",
    {
      className: cn("text-xs font-medium text-muted-foreground", { "text-destructive": hasErrors }),
      id: helpId(fieldPathId),
      children: /* @__PURE__ */ jsx11(RichHelp, { help, registry, uiSchema })
    }
  );
}

// src/FieldTemplate/FieldTemplate.tsx
import {
  getTemplate as getTemplate3,
  getUiOptions as getUiOptions3
} from "@rjsf/utils";
import { jsx as jsx12, jsxs as jsxs6 } from "react/jsx-runtime";
function FieldTemplate({
  id,
  children,
  displayLabel,
  rawErrors = [],
  errors,
  help,
  description,
  rawDescription,
  classNames,
  style,
  disabled,
  label,
  hidden,
  onKeyRename,
  onKeyRenameBlur,
  onRemoveProperty,
  readonly,
  required,
  schema,
  uiSchema,
  registry
}) {
  const uiOptions = getUiOptions3(uiSchema);
  const WrapIfAdditionalTemplate2 = getTemplate3(
    "WrapIfAdditionalTemplate",
    registry,
    uiOptions
  );
  if (hidden) {
    return /* @__PURE__ */ jsx12("div", { className: "hidden", children });
  }
  const isCheckbox = uiOptions.widget === "checkbox";
  return /* @__PURE__ */ jsx12(
    WrapIfAdditionalTemplate2,
    {
      classNames,
      style,
      disabled,
      id,
      label,
      displayLabel,
      onKeyRename,
      onKeyRenameBlur,
      onRemoveProperty,
      rawDescription,
      readonly,
      required,
      schema,
      uiSchema,
      registry,
      children: /* @__PURE__ */ jsxs6("div", { className: "flex flex-col gap-2", children: [
        displayLabel && !isCheckbox && /* @__PURE__ */ jsxs6(
          "label",
          {
            className: cn(
              "text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70",
              { " text-destructive": rawErrors.length > 0 }
            ),
            htmlFor: id,
            children: [
              label,
              required ? "*" : null
            ]
          }
        ),
        children,
        displayLabel && rawDescription && !isCheckbox && /* @__PURE__ */ jsx12(
          "span",
          {
            className: cn("text-xs font-medium text-muted-foreground", { " text-destructive": rawErrors.length > 0 }),
            children: description
          }
        ),
        errors,
        help
      ] })
    }
  );
}

// src/GridTemplate/GridTemplate.tsx
import { jsx as jsx13 } from "react/jsx-runtime";
function GridTemplate(props) {
  const { children, column, className, ...rest } = props;
  return /* @__PURE__ */ jsx13("div", { className: cn("grid gap-2", className), ...rest, children });
}

// src/IconButton/IconButton.tsx
import { TranslatableString as TranslatableString3 } from "@rjsf/utils";
import { ChevronDown, ChevronUp, Copy, Trash2 } from "lucide-react";
import { jsx as jsx14 } from "react/jsx-runtime";
function IconButton(props) {
  const { icon, iconType, className, uiSchema, registry, ...otherProps } = props;
  return /* @__PURE__ */ jsx14(Button, { size: "icon", variant: "outline", className, ...otherProps, type: "button", children: icon });
}
function CopyButton(props) {
  const {
    registry: { translateString }
  } = props;
  return /* @__PURE__ */ jsx14(IconButton, { title: translateString(TranslatableString3.CopyButton), ...props, icon: /* @__PURE__ */ jsx14(Copy, { className: "h-4 w-4" }) });
}
function MoveDownButton(props) {
  const {
    registry: { translateString }
  } = props;
  return /* @__PURE__ */ jsx14(
    IconButton,
    {
      title: translateString(TranslatableString3.MoveDownButton),
      ...props,
      icon: /* @__PURE__ */ jsx14(ChevronDown, { className: "h-4 w-4" })
    }
  );
}
function MoveUpButton(props) {
  const {
    registry: { translateString }
  } = props;
  return /* @__PURE__ */ jsx14(
    IconButton,
    {
      title: translateString(TranslatableString3.MoveUpButton),
      ...props,
      icon: /* @__PURE__ */ jsx14(ChevronUp, { className: "h-4 w-4" })
    }
  );
}
function RemoveButton(props) {
  const {
    registry: { translateString }
  } = props;
  return /* @__PURE__ */ jsx14(
    IconButton,
    {
      title: translateString(TranslatableString3.RemoveButton),
      ...props,
      className: "border-destructive",
      icon: /* @__PURE__ */ jsx14(Trash2, { className: "h-4 w-4 stroke-destructive" })
    }
  );
}

// src/MultiSchemaFieldTemplate/MultiSchemaFieldTemplate.tsx
import { jsx as jsx15, jsxs as jsxs7 } from "react/jsx-runtime";
function MultiSchemaFieldTemplate({ selector, optionSchemaField }) {
  return /* @__PURE__ */ jsxs7("div", { className: cn("p-4 border rounded-md bg-background shadow-sm"), children: [
    /* @__PURE__ */ jsx15("div", { className: cn("mb-4"), children: selector }),
    optionSchemaField
  ] });
}

// src/ObjectFieldTemplate/ObjectFieldTemplate.tsx
import {
  buttonId as buttonId2,
  canExpand,
  descriptionId,
  getTemplate as getTemplate4,
  getUiOptions as getUiOptions4,
  titleId
} from "@rjsf/utils";
import { Fragment, jsx as jsx16, jsxs as jsxs8 } from "react/jsx-runtime";
function ObjectFieldTemplate({
  description,
  title,
  properties,
  required,
  uiSchema,
  fieldPathId,
  schema,
  formData,
  optionalDataControl,
  onAddProperty,
  disabled,
  readonly,
  registry
}) {
  const uiOptions = getUiOptions4(uiSchema);
  const TitleFieldTemplate = getTemplate4("TitleFieldTemplate", registry, uiOptions);
  const DescriptionFieldTemplate = getTemplate4(
    "DescriptionFieldTemplate",
    registry,
    uiOptions
  );
  const showOptionalDataControlInTitle = !readonly && !disabled;
  const {
    ButtonTemplates: { AddButton: AddButton2 }
  } = registry.templates;
  return /* @__PURE__ */ jsxs8(Fragment, { children: [
    title && /* @__PURE__ */ jsx16(
      TitleFieldTemplate,
      {
        id: titleId(fieldPathId),
        title,
        required,
        schema,
        uiSchema,
        registry,
        optionalDataControl: showOptionalDataControlInTitle ? optionalDataControl : void 0
      }
    ),
    description && /* @__PURE__ */ jsx16(
      DescriptionFieldTemplate,
      {
        id: descriptionId(fieldPathId),
        description,
        schema,
        uiSchema,
        registry
      }
    ),
    /* @__PURE__ */ jsxs8("div", { className: "flex flex-col gap-2", children: [
      !showOptionalDataControlInTitle ? optionalDataControl : void 0,
      properties.map((element, index) => /* @__PURE__ */ jsx16("div", { className: `${element.hidden ? "hidden" : ""} flex`, children: /* @__PURE__ */ jsx16("div", { className: "w-full", children: element.content }) }, index)),
      canExpand(schema, uiSchema, formData) ? /* @__PURE__ */ jsx16("div", { className: "mt-2 flex justify-end", children: /* @__PURE__ */ jsx16(
        AddButton2,
        {
          id: buttonId2(fieldPathId, "add"),
          onClick: onAddProperty,
          disabled: disabled || readonly,
          className: "rjsf-object-property-expand",
          uiSchema,
          registry
        }
      ) }) : null
    ] })
  ] });
}

// src/OptionalDataControlsTemplate/OptionalDataControlsTemplate.tsx
import { PlusCircle as PlusCircle2 } from "lucide-react";
import { jsx as jsx17 } from "react/jsx-runtime";
function OptionalDataControlsTemplate(props) {
  const { id, registry, label, onAddClick, onRemoveClick } = props;
  if (onAddClick) {
    return /* @__PURE__ */ jsx17(
      IconButton,
      {
        id,
        registry,
        className: "rjsf-add-optional-data",
        onClick: onAddClick,
        title: label,
        icon: /* @__PURE__ */ jsx17(PlusCircle2, {}),
        size: "xs"
      }
    );
  } else if (onRemoveClick) {
    return /* @__PURE__ */ jsx17(
      RemoveButton,
      {
        id,
        registry,
        className: "rjsf-remove-optional-data",
        onClick: onRemoveClick,
        title: label,
        size: "xs"
      }
    );
  }
  return /* @__PURE__ */ jsx17("em", { id, children: label });
}

// src/SubmitButton/SubmitButton.tsx
import { getSubmitButtonOptions } from "@rjsf/utils";
import { jsx as jsx18 } from "react/jsx-runtime";
function SubmitButton(props) {
  const { submitText, norender, props: submitButtonProps } = getSubmitButtonOptions(props.uiSchema);
  if (norender) {
    return null;
  }
  return /* @__PURE__ */ jsx18("div", { children: /* @__PURE__ */ jsx18(Button, { type: "submit", ...submitButtonProps, className: cn("my-2", submitButtonProps?.className), children: submitText }) });
}

// src/TitleField/TitleField.tsx
import { getUiOptions as getUiOptions5 } from "@rjsf/utils";

// src/components/ui/separator.tsx
import { Root } from "@radix-ui/react-separator";
import { jsx as jsx19 } from "react/jsx-runtime";
function Separator({
  className,
  orientation = "horizontal",
  decorative = true,
  ...props
}) {
  return /* @__PURE__ */ jsx19(
    Root,
    {
      "data-slot": "separator",
      decorative,
      orientation,
      className: cn(
        "bg-border shrink-0 data-[orientation=horizontal]:h-px data-[orientation=horizontal]:w-full data-[orientation=vertical]:h-full data-[orientation=vertical]:w-px",
        className
      ),
      ...props
    }
  );
}

// src/TitleField/TitleField.tsx
import { jsx as jsx20, jsxs as jsxs9 } from "react/jsx-runtime";
function TitleField({
  id,
  title,
  uiSchema,
  optionalDataControl
}) {
  const uiOptions = getUiOptions5(uiSchema);
  let heading = /* @__PURE__ */ jsx20("h5", { children: uiOptions.title || title });
  if (optionalDataControl) {
    heading = /* @__PURE__ */ jsxs9("div", { className: "flex flex-row", children: [
      /* @__PURE__ */ jsx20("div", { className: "flex-grow", children: heading }),
      /* @__PURE__ */ jsx20("div", { className: "flex", children: optionalDataControl })
    ] });
  }
  return /* @__PURE__ */ jsxs9("div", { id, className: "my-1 flex flex-col gap-0.5", children: [
    heading,
    /* @__PURE__ */ jsx20(Separator, { dir: "horizontal", style: { height: "1px" }, className: "my-1" })
  ] });
}

// src/WrapIfAdditionalTemplate/WrapIfAdditionalTemplate.tsx
import {
  ADDITIONAL_PROPERTY_FLAG,
  buttonId as buttonId3,
  TranslatableString as TranslatableString4
} from "@rjsf/utils";
import { Fragment as Fragment2, jsx as jsx21, jsxs as jsxs10 } from "react/jsx-runtime";
function WrapIfAdditionalTemplate({
  classNames,
  style,
  children,
  disabled,
  id,
  label,
  displayLabel,
  onRemoveProperty,
  onKeyRenameBlur,
  rawDescription,
  readonly,
  required,
  schema,
  uiSchema,
  registry
}) {
  const { templates, translateString } = registry;
  const { RemoveButton: RemoveButton2 } = templates.ButtonTemplates;
  const keyLabel = translateString(TranslatableString4.KeyLabel, [label]);
  const additional = ADDITIONAL_PROPERTY_FLAG in schema;
  if (!additional) {
    return /* @__PURE__ */ jsx21("div", { className: classNames, style, children });
  }
  const marginDesc = rawDescription ? -28 : 0;
  const margin = displayLabel ? 22 + marginDesc : 0;
  const keyId = `${id}-key`;
  return /* @__PURE__ */ jsxs10(Fragment2, { children: [
    /* @__PURE__ */ jsxs10("div", { className: `grid grid-cols-12 col-span-12 items-center gap-2 ${classNames}`, style, children: [
      /* @__PURE__ */ jsx21("div", { className: "grid gap-2 col-span-5", children: /* @__PURE__ */ jsxs10("div", { className: "flex flex-col gap-2", children: [
        displayLabel && /* @__PURE__ */ jsx21("label", { htmlFor: keyId, className: "text-sm font-medium text-muted-foreground leading-none", children: keyLabel }),
        /* @__PURE__ */ jsx21("div", { className: "pl-0.5", children: /* @__PURE__ */ jsx21(
          Input,
          {
            required,
            defaultValue: label,
            disabled: disabled || readonly,
            id: keyId,
            name: keyId,
            onBlur: !readonly ? onKeyRenameBlur : void 0,
            type: "text",
            className: "w-full border shadow-sm"
          }
        ) }),
        !!rawDescription && /* @__PURE__ */ jsx21("span", { className: "text-xs font-medium text-muted-foreground", children: /* @__PURE__ */ jsx21("div", { className: "text-sm text-muted-foreground", children: "\xA0" }) })
      ] }) }),
      /* @__PURE__ */ jsx21("div", { className: "grid gap-2 col-span-6 pr-0.5", children }),
      /* @__PURE__ */ jsx21("div", { className: "grid gap-2 col-span-1", style: { marginTop: `${margin}px` }, children: /* @__PURE__ */ jsx21(
        RemoveButton2,
        {
          id: buttonId3(id, "remove"),
          iconType: "block",
          className: "rjsf-object-property-remove w-full",
          disabled: disabled || readonly,
          onClick: onRemoveProperty,
          uiSchema,
          registry
        }
      ) })
    ] }),
    /* @__PURE__ */ jsx21(Separator, { dir: "horizontal", className: "mt-2" })
  ] });
}

// src/Templates/Templates.ts
function generateTemplates() {
  return {
    ArrayFieldItemTemplate,
    ArrayFieldTemplate,
    BaseInputTemplate,
    ButtonTemplates: {
      AddButton,
      CopyButton,
      MoveDownButton,
      MoveUpButton,
      RemoveButton,
      SubmitButton
    },
    DescriptionFieldTemplate: DescriptionField,
    ErrorListTemplate: ErrorList,
    FieldErrorTemplate,
    FieldHelpTemplate,
    FieldTemplate,
    GridTemplate,
    MultiSchemaFieldTemplate,
    ObjectFieldTemplate,
    OptionalDataControlsTemplate,
    TitleFieldTemplate: TitleField,
    WrapIfAdditionalTemplate
  };
}
var Templates_default = generateTemplates();

// src/CheckboxWidget/CheckboxWidget.tsx
import {
  ariaDescribedByIds as ariaDescribedByIds2,
  descriptionId as descriptionId2,
  getTemplate as getTemplate5,
  labelValue,
  schemaRequiresTrueValue
} from "@rjsf/utils";

// src/components/ui/checkbox.tsx
import { Indicator, Root as Root2 } from "@radix-ui/react-checkbox";
import { CheckIcon } from "@radix-ui/react-icons";
import { jsx as jsx22 } from "react/jsx-runtime";
function Checkbox({ className, ...props }) {
  return /* @__PURE__ */ jsx22(
    Root2,
    {
      "data-slot": "checkbox",
      className: cn(
        "peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50",
        className
      ),
      ...props,
      children: /* @__PURE__ */ jsx22(
        Indicator,
        {
          "data-slot": "checkbox-indicator",
          className: "flex items-center justify-center text-current transition-none",
          children: /* @__PURE__ */ jsx22(CheckIcon, { className: "size-3.5" })
        }
      )
    }
  );
}

// src/components/ui/label.tsx
import { Root as Root3 } from "@radix-ui/react-label";
import { jsx as jsx23 } from "react/jsx-runtime";
function Label({ className, ...props }) {
  return /* @__PURE__ */ jsx23(
    Root3,
    {
      "data-slot": "label",
      className: cn(
        "flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50",
        className
      ),
      ...props
    }
  );
}

// src/CheckboxWidget/CheckboxWidget.tsx
import { jsx as jsx24, jsxs as jsxs11 } from "react/jsx-runtime";
function CheckboxWidget(props) {
  const {
    id,
    htmlName,
    value,
    disabled,
    readonly,
    label,
    hideLabel,
    schema,
    autofocus,
    options,
    onChange,
    onBlur,
    onFocus,
    registry,
    uiSchema,
    className
  } = props;
  const required = schemaRequiresTrueValue(schema);
  const DescriptionFieldTemplate = getTemplate5(
    "DescriptionFieldTemplate",
    registry,
    options
  );
  const _onChange = (checked) => onChange(checked);
  const _onBlur = () => onBlur(id, value);
  const _onFocus = () => onFocus(id, value);
  const description = options.description || schema.description;
  return /* @__PURE__ */ jsxs11(
    "div",
    {
      className: `relative ${disabled || readonly ? "cursor-not-allowed opacity-50" : ""}`,
      "aria-describedby": ariaDescribedByIds2(id),
      children: [
        !hideLabel && description && /* @__PURE__ */ jsx24(
          DescriptionFieldTemplate,
          {
            id: descriptionId2(id),
            description,
            schema,
            uiSchema,
            registry
          }
        ),
        /* @__PURE__ */ jsxs11("div", { className: "flex items-center gap-2 my-2", children: [
          /* @__PURE__ */ jsx24(
            Checkbox,
            {
              id,
              name: htmlName || id,
              checked: typeof value === "undefined" ? false : Boolean(value),
              required,
              disabled: disabled || readonly,
              autoFocus: autofocus,
              onCheckedChange: _onChange,
              onBlur: _onBlur,
              onFocus: _onFocus,
              className
            }
          ),
          /* @__PURE__ */ jsx24(Label, { className: "leading-tight", htmlFor: id, children: labelValue(label, hideLabel || !label) })
        ] })
      ]
    }
  );
}

// src/CheckboxesWidget/CheckboxesWidget.tsx
import {
  ariaDescribedByIds as ariaDescribedByIds3,
  enumOptionsDeselectValue,
  enumOptionsIsSelected,
  enumOptionsSelectValue,
  enumOptionsValueForIndex,
  optionId
} from "@rjsf/utils";
import { jsx as jsx25, jsxs as jsxs12 } from "react/jsx-runtime";
function CheckboxesWidget({
  id,
  htmlName,
  disabled,
  options,
  value,
  autofocus,
  readonly,
  required,
  onChange,
  onBlur,
  onFocus,
  className
}) {
  const { enumOptions, enumDisabled, inline, emptyValue } = options;
  const checkboxesValues = Array.isArray(value) ? value : [value];
  const _onBlur = ({ target }) => onBlur(id, enumOptionsValueForIndex(target && target.value, enumOptions, emptyValue));
  const _onFocus = ({ target }) => onFocus(id, enumOptionsValueForIndex(target && target.value, enumOptions, emptyValue));
  return /* @__PURE__ */ jsx25("div", { className: cn({ "flex flex-col gap-2": !inline, "flex flex-row gap-4 flex-wrap": inline }), children: Array.isArray(enumOptions) && enumOptions.map((option, index) => {
    const checked = enumOptionsIsSelected(option.value, checkboxesValues);
    const itemDisabled = Array.isArray(enumDisabled) && enumDisabled.indexOf(option.value) !== -1;
    const indexOptionId = optionId(id, index);
    return /* @__PURE__ */ jsxs12("div", { className: "flex items-center gap-2", children: [
      /* @__PURE__ */ jsx25(
        Checkbox,
        {
          id: indexOptionId,
          name: htmlName || id,
          required,
          disabled: disabled || itemDisabled || readonly,
          onCheckedChange: (state) => {
            if (state) {
              onChange(enumOptionsSelectValue(index, checkboxesValues, enumOptions));
            } else {
              onChange(enumOptionsDeselectValue(index, checkboxesValues, enumOptions));
            }
          },
          className,
          checked,
          autoFocus: autofocus && index === 0,
          onBlur: _onBlur,
          onFocus: _onFocus,
          "aria-describedby": ariaDescribedByIds3(id)
        }
      ),
      /* @__PURE__ */ jsx25(Label, { className: "leading-tight", htmlFor: optionId(id, index), children: option.label })
    ] }, indexOptionId);
  }) });
}

// src/RadioWidget/RadioWidget.tsx
import {
  ariaDescribedByIds as ariaDescribedByIds4,
  enumOptionsIsSelected as enumOptionsIsSelected2,
  enumOptionsValueForIndex as enumOptionsValueForIndex2,
  optionId as optionId2
} from "@rjsf/utils";

// src/components/ui/radio-group.tsx
import { Root as Root4, Indicator as Indicator2, Item } from "@radix-ui/react-radio-group";
import { CircleIcon } from "lucide-react";
import { jsx as jsx26 } from "react/jsx-runtime";
function RadioGroup({ className, ...props }) {
  return /* @__PURE__ */ jsx26(Root4, { "data-slot": "radio-group", className: cn("grid gap-3", className), ...props });
}
function RadioGroupItem({ className, ...props }) {
  return /* @__PURE__ */ jsx26(
    Item,
    {
      "data-slot": "radio-group-item",
      className: cn(
        "border-input text-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 aspect-square size-4 shrink-0 rounded-full border shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50",
        className
      ),
      ...props,
      children: /* @__PURE__ */ jsx26(Indicator2, { "data-slot": "radio-group-indicator", className: "relative flex items-center justify-center", children: /* @__PURE__ */ jsx26(CircleIcon, { className: "fill-primary absolute top-1/2 left-1/2 size-2 -translate-x-1/2 -translate-y-1/2" }) })
    }
  );
}

// src/RadioWidget/RadioWidget.tsx
import { jsx as jsx27, jsxs as jsxs13 } from "react/jsx-runtime";
function RadioWidget({
  id,
  options,
  value,
  required,
  disabled,
  readonly,
  onChange,
  onBlur,
  onFocus,
  className
}) {
  const { enumOptions, enumDisabled, emptyValue } = options;
  const _onChange = (value2) => onChange(enumOptionsValueForIndex2(value2, enumOptions, emptyValue));
  const _onBlur = ({ target }) => onBlur(id, enumOptionsValueForIndex2(target && target.value, enumOptions, emptyValue));
  const _onFocus = ({ target }) => onFocus(id, enumOptionsValueForIndex2(target && target.value, enumOptions, emptyValue));
  const inline = Boolean(options && options.inline);
  return /* @__PURE__ */ jsx27("div", { className: "mb-0", children: /* @__PURE__ */ jsx27(
    RadioGroup,
    {
      defaultValue: value?.toString(),
      required,
      disabled: disabled || readonly,
      onValueChange: (e) => {
        _onChange(e);
      },
      onBlur: _onBlur,
      onFocus: _onFocus,
      "aria-describedby": ariaDescribedByIds4(id),
      orientation: inline ? "horizontal" : "vertical",
      className: cn("flex flex-wrap", { "flex-col": !inline }, className),
      children: Array.isArray(enumOptions) && enumOptions.map((option, index) => {
        const itemDisabled = Array.isArray(enumDisabled) && enumDisabled.indexOf(option.value) !== -1;
        const checked = enumOptionsIsSelected2(option.value, value);
        return /* @__PURE__ */ jsxs13("div", { className: "flex items-center gap-2", children: [
          /* @__PURE__ */ jsx27(
            RadioGroupItem,
            {
              checked,
              value: index.toString(),
              id: optionId2(id, index),
              disabled: itemDisabled
            }
          ),
          /* @__PURE__ */ jsx27(Label, { className: "leading-tight", htmlFor: optionId2(id, index), children: option.label })
        ] }, optionId2(id, index));
      })
    }
  ) });
}

// src/RangeWidget/RangeWidget.tsx
import { ariaDescribedByIds as ariaDescribedByIds5, rangeSpec } from "@rjsf/utils";
import _pick from "lodash/pick";

// src/components/ui/slider.tsx
import { Range, Root as Root5, Thumb, Track } from "@radix-ui/react-slider";
import { useMemo } from "react";
import { jsx as jsx28, jsxs as jsxs14 } from "react/jsx-runtime";
function Slider({ className, defaultValue, value, min = 0, max = 100, ...props }) {
  const _values = useMemo(
    () => Array.isArray(value) ? value : Array.isArray(defaultValue) ? defaultValue : [min, max],
    [value, defaultValue, min, max]
  );
  return /* @__PURE__ */ jsxs14(
    Root5,
    {
      "data-slot": "slider",
      defaultValue,
      value,
      min,
      max,
      className: cn(
        "relative flex w-full touch-none items-center select-none data-[disabled]:opacity-50 data-[orientation=vertical]:h-full data-[orientation=vertical]:min-h-44 data-[orientation=vertical]:w-auto data-[orientation=vertical]:flex-col",
        className
      ),
      ...props,
      children: [
        /* @__PURE__ */ jsx28(
          Track,
          {
            "data-slot": "slider-track",
            className: cn(
              "bg-muted relative grow overflow-hidden rounded-full data-[orientation=horizontal]:h-1.5 data-[orientation=horizontal]:w-full data-[orientation=vertical]:h-full data-[orientation=vertical]:w-1.5"
            ),
            children: /* @__PURE__ */ jsx28(
              Range,
              {
                "data-slot": "slider-range",
                className: cn("bg-primary absolute data-[orientation=horizontal]:h-full data-[orientation=vertical]:w-full")
              }
            )
          }
        ),
        Array.from({ length: _values.length }, (_, index) => /* @__PURE__ */ jsx28(
          Thumb,
          {
            "data-slot": "slider-thumb",
            className: "border-primary bg-background ring-ring/50 block size-4 shrink-0 rounded-full border shadow-sm transition-[color,box-shadow] hover:ring-4 focus-visible:ring-4 focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50"
          },
          index
        ))
      ]
    }
  );
}

// src/RangeWidget/RangeWidget.tsx
import { Fragment as Fragment3, jsx as jsx29, jsxs as jsxs15 } from "react/jsx-runtime";
var allowedProps = [
  "name",
  "min",
  "max",
  "step",
  "orientation",
  "disabled",
  "defaultValue",
  "value",
  "onValueChange",
  "className",
  "dir",
  "inverted",
  "minStepsBetweenThumbs"
];
function RangeWidget({
  value,
  readonly,
  disabled,
  options,
  schema,
  onChange,
  label,
  id
}) {
  const _onChange = (value2) => onChange(value2[0]);
  const sliderProps = { value, label, id, ...rangeSpec(schema) };
  const uiProps = { id, ..._pick(options.props || {}, allowedProps) };
  return /* @__PURE__ */ jsxs15(Fragment3, { children: [
    /* @__PURE__ */ jsx29(
      Slider,
      {
        disabled: disabled || readonly,
        min: sliderProps.min,
        max: sliderProps.max,
        step: sliderProps.step,
        value: [value],
        onValueChange: _onChange,
        ...uiProps,
        "aria-describedby": ariaDescribedByIds5(id)
      }
    ),
    value
  ] });
}

// src/SelectWidget/SelectWidget.tsx
import {
  ariaDescribedByIds as ariaDescribedByIds6,
  enumOptionsIndexForValue,
  enumOptionsValueForIndex as enumOptionsValueForIndex3
} from "@rjsf/utils";

// src/components/ui/fancy-multi-select.tsx
import { Command as CommandPrimitive2 } from "cmdk";
import isEqual from "lodash/isEqual";
import { X } from "lucide-react";
import {
  useCallback,
  useMemo as useMemo2,
  useRef,
  useState
} from "react";

// src/components/ui/badge.tsx
import { cva as cva3 } from "class-variance-authority";
import { Slot as Slot2 } from "@radix-ui/react-slot";
import { jsx as jsx30 } from "react/jsx-runtime";
var badgeVariants = cva3(
  "inline-flex items-center justify-center rounded-md border px-2 py-0.5 text-xs font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden",
  {
    variants: {
      variant: {
        default: "border-transparent bg-primary text-primary-foreground [a&]:hover:bg-primary/90",
        secondary: "border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90",
        destructive: "border-transparent bg-destructive text-white [a&]:hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60",
        outline: "text-foreground [a&]:hover:bg-accent [a&]:hover:text-accent-foreground"
      }
    },
    defaultVariants: {
      variant: "default"
    }
  }
);
function Badge({
  className,
  variant,
  asChild = false,
  ...props
}) {
  const Comp = asChild ? Slot2 : "span";
  return /* @__PURE__ */ jsx30(Comp, { "data-slot": "badge", className: cn(badgeVariants({ variant }), className), ...props });
}

// src/components/ui/command.tsx
import { Command as CommandPrimitive } from "cmdk";
import { Search } from "lucide-react";
import { forwardRef } from "react";

// src/components/ui/dialog.tsx
import { Close, Content, Description, Overlay, Portal, Root as Root6, Title, Trigger } from "@radix-ui/react-dialog";
import { XIcon } from "lucide-react";
import { jsx as jsx31, jsxs as jsxs16 } from "react/jsx-runtime";

// src/components/ui/command.tsx
import { jsx as jsx32, jsxs as jsxs17 } from "react/jsx-runtime";
var Command = forwardRef(
  ({ className, ...props }, ref) => {
    return /* @__PURE__ */ jsx32(
      CommandPrimitive,
      {
        ref,
        "data-slot": "command",
        className: cn(
          "bg-popover text-popover-foreground flex h-full w-full flex-col overflow-hidden rounded-md",
          className
        ),
        ...props
      }
    );
  }
);
var CommandInput = forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsxs17("div", { className: "flex items-center border-b px-3", ...{ "cmdk-input-wrapper": "" }, children: [
  /* @__PURE__ */ jsx32(Search, { className: "mr-2 h-4 w-4 shrink-0 opacity-50" }),
  /* @__PURE__ */ jsx32(
    CommandPrimitive.Input,
    {
      ref,
      className: cn(
        "flex h-10 w-full rounded-md bg-transparent py-3 text-sm outline-none placeholder:text-muted-foreground disabled:cursor-not-allowed disabled:opacity-50",
        className
      ),
      ...props
    }
  )
] }));
CommandInput.displayName = CommandPrimitive.Input.displayName;
function CommandList({ className, ...props }) {
  return /* @__PURE__ */ jsx32(
    CommandPrimitive.List,
    {
      "data-slot": "command-list",
      className: cn("max-h-[300px] scroll-py-1 overflow-x-hidden overflow-y-auto", className),
      ...props
    }
  );
}
function CommandGroup({ className, ...props }) {
  return /* @__PURE__ */ jsx32(
    CommandPrimitive.Group,
    {
      "data-slot": "command-group",
      className: cn(
        "text-foreground [&_[cmdk-group-heading]]:text-muted-foreground overflow-hidden p-1 [&_[cmdk-group-heading]]:px-2 [&_[cmdk-group-heading]]:py-1.5 [&_[cmdk-group-heading]]:text-xs [&_[cmdk-group-heading]]:font-medium",
        className
      ),
      ...props
    }
  );
}
function CommandItem({ className, ...props }) {
  return /* @__PURE__ */ jsx32(
    CommandPrimitive.Item,
    {
      "data-slot": "command-item",
      className: cn(
        "data-[selected=true]:bg-accent data-[selected=true]:text-accent-foreground [&_svg:not([class*='text-'])]:text-muted-foreground relative flex cursor-default items-center gap-2 rounded-sm px-2 py-1.5 text-sm outline-hidden select-none data-[disabled=true]:pointer-events-none data-[disabled=true]:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
        className
      ),
      ...props
    }
  );
}

// src/components/ui/fancy-multi-select.tsx
import { jsx as jsx33, jsxs as jsxs18 } from "react/jsx-runtime";
function FancyMultiSelect({
  multiple,
  items = [],
  selected,
  onValueChange,
  autoFocus = false,
  disabled = false,
  ariaDescribedby,
  ariaPlaceholder,
  onFocus,
  onBlur,
  className,
  id
}) {
  const inputRef = useRef(null);
  const [open, setOpen] = useState(false);
  const [inputValue, setInputValue] = useState("");
  const selectedItems = useMemo2(
    () => items.filter((item) => selected.some((selectedValue) => isEqual(item.value, selectedValue))),
    [items, selected]
  );
  const selectables = useMemo2(
    () => items.filter((framework) => !selectedItems.some((item) => isEqual(item.value, framework.value))),
    [items, selectedItems]
  );
  const handleUnselect = useCallback(
    (framework) => {
      if (disabled) {
        return;
      }
      const newSelected = selectedItems.filter((s) => !isEqual(s.value, framework.value));
      onValueChange?.(newSelected.map((item) => item.index));
    },
    [selectedItems, onValueChange, disabled]
  );
  const handleKeyDown = useCallback(
    (e) => {
      if (disabled || !inputRef.current || inputRef.current.value !== "") {
        return;
      }
      if (e.key === "Delete" || e.key === "Backspace") {
        const newSelected = selectedItems.slice(0, -1);
        onValueChange?.(newSelected.map((item) => item.index));
      } else if (e.key === "Escape") {
        inputRef.current.blur();
      }
    },
    [selectedItems, onValueChange, disabled]
  );
  const handleSelect = useCallback(
    (item) => {
      if (disabled) {
        return;
      }
      setInputValue("");
      const newSelected = multiple ? [...selectedItems, item] : [item];
      onValueChange?.(newSelected.map((item2) => item2.index));
    },
    [multiple, selectedItems, onValueChange, disabled]
  );
  const handleFocus = useCallback(
    (e) => {
      if (!disabled) {
        setOpen(true);
      }
      onFocus?.(e);
    },
    [disabled, onFocus]
  );
  return /* @__PURE__ */ jsxs18(
    Command,
    {
      onKeyDown: handleKeyDown,
      className: cn("overflow-visible bg-transparent", className),
      autoFocus,
      "aria-disabled": disabled,
      onBlur,
      onFocus: handleFocus,
      "aria-describedby": ariaDescribedby,
      "aria-placeholder": ariaPlaceholder,
      children: [
        /* @__PURE__ */ jsx33(
          "div",
          {
            className: cn(
              "group border border-input px-3 py-2 text-sm ring-offset-background rounded-md focus-within:ring-1 focus-within:ring-ring focus-within:ring-offset-1",
              disabled && "opacity-50 cursor-not-allowed"
            ),
            children: /* @__PURE__ */ jsxs18("div", { className: "flex gap-1 flex-wrap", children: [
              selectedItems.map((item) => /* @__PURE__ */ jsxs18(Badge, { variant: "secondary", children: [
                item.label,
                /* @__PURE__ */ jsx33(
                  "button",
                  {
                    type: "button",
                    className: "rtl:mr-1 ltr:ml-1 ring-offset-background rounded-full outline-none focus:ring-1 focus:ring-ring focus:ring-offset-1",
                    onKeyDown: (e) => e.key === "Enter" && !disabled && handleUnselect(item),
                    onMouseDown: (e) => {
                      e.preventDefault();
                      e.stopPropagation();
                    },
                    onClick: () => handleUnselect(item),
                    disabled,
                    children: /* @__PURE__ */ jsx33(
                      X,
                      {
                        className: cn(
                          "h-3 w-3 text-muted-foreground hover:text-foreground",
                          disabled && "pointer-events-none"
                        )
                      }
                    )
                  }
                )
              ] }, item.value)),
              /* @__PURE__ */ jsx33(
                CommandPrimitive2.Input,
                {
                  ref: inputRef,
                  value: inputValue,
                  onValueChange: setInputValue,
                  onBlur: () => setOpen(false),
                  onFocus: () => !disabled && setOpen(true),
                  placeholder: "Select ...",
                  className: "rtl:mr-2 ltr:ml-2 bg-transparent outline-none placeholder:text-muted-foreground flex-1",
                  disabled,
                  "aria-controls": `command-item-input-${id}`,
                  "aria-labelledby": `command-item-input-${id}`,
                  id: `command-item-input-${id}`
                }
              )
            ] })
          }
        ),
        open && !disabled && selectables.length > 0 && /* @__PURE__ */ jsx33("div", { className: "relative mt-2", children: /* @__PURE__ */ jsx33("div", { className: "absolute w-full z-10 top-0 rounded-md border bg-popover text-popover-foreground shadow-md outline-none animate-in", children: /* @__PURE__ */ jsx33(CommandGroup, { className: "h-full overflow-auto", children: /* @__PURE__ */ jsx33(CommandList, { children: selectables.map((item) => /* @__PURE__ */ jsx33(
          CommandItem,
          {
            disabled: item.disabled,
            onMouseDown: (e) => {
              e.preventDefault();
              e.stopPropagation();
            },
            "aria-controls": `${item.value}-command-item`,
            "aria-labelledby": `${item.value}-command-item`,
            id: `${item.value}-command-item`,
            onSelect: () => handleSelect(item),
            className: "cursor-pointer",
            children: item.label
          },
          `${item.value}-command-item`
        )) }) }) }) })
      ]
    }
  );
}

// src/components/ui/fancy-select.tsx
import { Check, ChevronDown as ChevronDown2 } from "lucide-react";
import { useEffect, useRef as useRef2, useState as useState2 } from "react";
import { jsx as jsx34, jsxs as jsxs19 } from "react/jsx-runtime";
function FancySelect({
  items,
  selected,
  onValueChange,
  autoFocus = false,
  disabled = false,
  required = false,
  placeholder = "Select...",
  ariaDescribedby,
  ariaPlaceholder,
  onFocus,
  onBlur,
  className
}) {
  const [open, setOpen] = useState2(false);
  const selectedItem = items?.find((item) => item.value === selected);
  const selectedRef = useRef2(null);
  const containerRef = useRef2(null);
  useEffect(() => {
    if (open && selectedRef.current) {
      requestAnimationFrame(() => {
        selectedRef.current?.scrollIntoView({ block: "nearest" });
      });
    }
  }, [open]);
  const handleBlur = (e) => {
    if (containerRef.current && !containerRef.current.contains(e.relatedTarget)) {
      setOpen(false);
    }
    onBlur?.(e);
  };
  return /* @__PURE__ */ jsxs19(
    Command,
    {
      ref: containerRef,
      className: cn("overflow-visible bg-transparent", className),
      autoFocus,
      "aria-disabled": disabled,
      onBlur: handleBlur,
      onFocus,
      "aria-describedby": ariaDescribedby,
      "aria-placeholder": ariaPlaceholder,
      children: [
        /* @__PURE__ */ jsxs19(
          "div",
          {
            onClick: () => !disabled && setOpen(!open),
            className: cn(
              "flex h-9 w-full items-center justify-between gap-2 whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50",
              !selectedItem && required && "border-red-500",
              disabled && "opacity-50 cursor-not-allowed",
              className
            ),
            children: [
              /* @__PURE__ */ jsx34("span", { className: cn("flex-1 line-clamp-1", !selectedItem && "text-muted-foreground"), children: selectedItem?.label || placeholder }),
              /* @__PURE__ */ jsx34(ChevronDown2, { className: "h-4 w-4 opacity-50" })
            ]
          }
        ),
        /* @__PURE__ */ jsx34("div", { className: "relative", children: open && items && items.length > 0 ? /* @__PURE__ */ jsx34(
          "div",
          {
            style: { top: "0.5rem" },
            className: "absolute w-full z-10 rounded-md border bg-popover text-popover-foreground shadow-md outline-none",
            children: /* @__PURE__ */ jsx34(CommandGroup, { className: "h-full overflow-auto", children: /* @__PURE__ */ jsx34(CommandList, { children: items.map((item) => /* @__PURE__ */ jsxs19(
              CommandItem,
              {
                ref: item.value === selected ? selectedRef : void 0,
                onMouseDown: (e) => {
                  e.preventDefault();
                  e.stopPropagation();
                },
                onSelect: () => {
                  if (!item.disabled) {
                    onValueChange?.(item.value);
                    setOpen(false);
                  }
                },
                className: cn(
                  "cursor-pointer relative flex items-center justify-between rounded-sm py-1.5 gap-2 rtl:flex-row-reverse",
                  item.value === selected && "font-semibold",
                  item.disabled && "opacity-50 cursor-not-allowed"
                ),
                children: [
                  /* @__PURE__ */ jsx34("span", { children: item.label }),
                  /* @__PURE__ */ jsx34("span", { className: "flex h-3.5 w-3.5 items-center justify-center", children: item.value === selected && /* @__PURE__ */ jsx34(Check, { className: "h-4 w-4" }) })
                ]
              },
              item.value
            )) }) })
          }
        ) : null })
      ]
    }
  );
}

// src/SelectWidget/SelectWidget.tsx
import { jsx as jsx35 } from "react/jsx-runtime";
function SelectWidget({
  id,
  options,
  required,
  disabled,
  readonly,
  value,
  multiple,
  autofocus,
  onChange,
  onBlur,
  onFocus,
  defaultValue,
  placeholder,
  rawErrors = [],
  className
}) {
  const { enumOptions, enumDisabled, emptyValue: optEmptyValue } = options;
  const _onFancyFocus = () => {
    onFocus(id, enumOptionsValueForIndex3(value, enumOptions, optEmptyValue));
  };
  const _onFancyBlur = () => {
    onBlur(id, enumOptionsValueForIndex3(value, enumOptions, optEmptyValue));
  };
  const items = enumOptions?.map(({ value: value2, label }, index) => ({
    value: multiple ? value2 : index.toString(),
    label,
    index,
    disabled: Array.isArray(enumDisabled) && enumDisabled.includes(value2)
  }));
  const cnClassName = cn({ "border-destructive": rawErrors.length > 0 }, className);
  return /* @__PURE__ */ jsx35("div", { className: "p-0.5", children: !multiple ? /* @__PURE__ */ jsx35(
    FancySelect,
    {
      items,
      selected: enumOptionsIndexForValue(value ?? defaultValue, enumOptions, false),
      onValueChange: (selectedValue) => {
        onChange(enumOptionsValueForIndex3(selectedValue, enumOptions, optEmptyValue));
      },
      autoFocus: autofocus,
      disabled: disabled || readonly,
      required,
      placeholder,
      className: cnClassName,
      onFocus: _onFancyFocus,
      onBlur: _onFancyBlur,
      ariaDescribedby: ariaDescribedByIds6(id)
    }
  ) : /* @__PURE__ */ jsx35(
    FancyMultiSelect,
    {
      id,
      autoFocus: autofocus,
      disabled: disabled || readonly,
      multiple: true,
      className: cnClassName,
      items,
      selected: value,
      onValueChange: (values) => {
        onChange(enumOptionsValueForIndex3(values, enumOptions, optEmptyValue));
      },
      onFocus: _onFancyFocus,
      onBlur: _onFancyBlur
    }
  ) });
}

// src/TextareaWidget/TextareaWidget.tsx
import { ariaDescribedByIds as ariaDescribedByIds7 } from "@rjsf/utils";

// src/components/ui/textarea.tsx
import { jsx as jsx36 } from "react/jsx-runtime";
function Textarea({ className, ...props }) {
  return /* @__PURE__ */ jsx36(
    "textarea",
    {
      "data-slot": "textarea",
      className: cn(
        "border-input placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 flex field-sizing-content min-h-16 w-full rounded-md border bg-transparent px-3 py-2 text-base shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 md:text-sm",
        className
      ),
      ...props
    }
  );
}

// src/TextareaWidget/TextareaWidget.tsx
import { jsx as jsx37 } from "react/jsx-runtime";
function TextareaWidget({
  id,
  htmlName,
  placeholder,
  value,
  required,
  disabled,
  autofocus,
  readonly,
  onBlur,
  onFocus,
  onChange,
  options,
  className
}) {
  const _onChange = ({ target: { value: value2 } }) => onChange(value2 === "" ? options.emptyValue : value2);
  const _onBlur = ({ target }) => onBlur(id, target && target.value);
  const _onFocus = ({ target }) => onFocus(id, target && target.value);
  return /* @__PURE__ */ jsx37("div", { className: "flex p-0.5", children: /* @__PURE__ */ jsx37(
    Textarea,
    {
      id,
      name: htmlName || id,
      placeholder,
      disabled,
      readOnly: readonly,
      value: value ?? "",
      required,
      autoFocus: autofocus,
      rows: options.rows || 5,
      onChange: _onChange,
      onBlur: _onBlur,
      onFocus: _onFocus,
      "aria-describedby": ariaDescribedByIds7(id),
      className
    }
  ) });
}

// src/Widgets/Widgets.ts
function generateWidgets() {
  return {
    CheckboxWidget,
    CheckboxesWidget,
    RadioWidget,
    RangeWidget,
    SelectWidget,
    TextareaWidget
  };
}
var Widgets_default = generateWidgets();

// src/Theme/Theme.tsx
function generateTheme() {
  return {
    templates: generateTemplates(),
    widgets: generateWidgets()
  };
}
var Theme_default = generateTheme();

// src/Form/Form.tsx
function generateForm() {
  return withTheme(generateTheme());
}
var Form_default = generateForm();

// src/index.ts
var index_default = Form_default;
export {
  Form_default as Form,
  Templates_default as Templates,
  Theme_default as Theme,
  Widgets_default as Widgets,
  index_default as default,
  generateForm,
  generateTemplates,
  generateTheme,
  generateWidgets
};
