// @rjsf/shadcn/dist/rjsf-shadcn.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@rjsf/core'), require('@rjsf/utils'), require('lucide-react'), require('@radix-ui/react-slot'), require('class-variance-authority'), require('clsx'), require('tailwind-merge'), require('react/jsx-runtime'), require('@radix-ui/react-separator'), require('@radix-ui/react-checkbox'), require('@radix-ui/react-icons'), require('@radix-ui/react-label'), require('@radix-ui/react-radio-group'), require('lodash/pick'), require('@radix-ui/react-slider'), require('react'), require('cmdk'), require('lodash/isEqual'), require('@radix-ui/react-dialog')) :
  typeof define === 'function' && define.amd ? define(['exports', '@rjsf/core', '@rjsf/utils', 'lucide-react', '@radix-ui/react-slot', 'class-variance-authority', 'clsx', 'tailwind-merge', 'react/jsx-runtime', '@radix-ui/react-separator', '@radix-ui/react-checkbox', '@radix-ui/react-icons', '@radix-ui/react-label', '@radix-ui/react-radio-group', 'lodash/pick', '@radix-ui/react-slider', 'react', 'cmdk', 'lodash/isEqual', '@radix-ui/react-dialog'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global["@rjsf/rjsf-shadcn"] = {}, global.core, global.utils, global.lucideReact, global.reactSlot, global.classVarianceAuthority, global.clsx, global.tailwindMerge, global.jsxRuntime, global.reactSeparator, global.reactCheckbox, global.reactIcons, global.reactLabel, global.reactRadioGroup, global._pick, global.reactSlider, global.react, global.cmdk, global.isEqual));
})(this, (function (exports, core, utils, lucideReact, reactSlot, classVarianceAuthority, clsx, tailwindMerge, jsxRuntime, reactSeparator, reactCheckbox, reactIcons, reactLabel, reactRadioGroup, _pick, reactSlider, react, cmdk, isEqual) { 'use strict';

  // src/Form/Form.tsx
  function cn(...inputs) {
    return tailwindMerge.twMerge(clsx.clsx(inputs));
  }
  var buttonVariants = classVarianceAuthority.cva(
    "inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive",
    {
      variants: {
        variant: {
          default: "bg-primary text-primary-foreground shadow-xs hover:bg-primary/90",
          destructive: "bg-destructive text-white shadow-xs hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60",
          outline: "border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50",
          secondary: "bg-secondary text-secondary-foreground shadow-xs hover:bg-secondary/80",
          ghost: "hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50",
          link: "text-primary underline-offset-4 hover:underline"
        },
        size: {
          default: "h-9 px-4 py-2 has-[>svg]:px-3",
          sm: "h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5",
          lg: "h-10 rounded-md px-6 has-[>svg]:px-4",
          xs: "h-6 rounded-md gap-1 px-2 has-[>svg]:px-2",
          icon: "size-9"
        }
      },
      defaultVariants: {
        variant: "default",
        size: "default"
      }
    }
  );
  function Button({
    className,
    variant,
    size,
    asChild = false,
    ...props
  }) {
    const Comp = asChild ? reactSlot.Slot : "button";
    return /* @__PURE__ */ jsxRuntime.jsx(Comp, { "data-slot": "button", className: cn(buttonVariants({ variant, size, className })), ...props });
  }
  function AddButton({
    uiSchema,
    registry,
    className,
    ...props
  }) {
    const { translateString } = registry;
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "p-0 m-0", children: /* @__PURE__ */ jsxRuntime.jsxs(Button, { ...props, className: cn("w-fit gap-2", className), variant: "outline", type: "button", children: [
      /* @__PURE__ */ jsxRuntime.jsx(lucideReact.PlusCircle, { size: 16 }),
      " ",
      translateString(utils.TranslatableString.AddItemButton)
    ] }) });
  }
  function ArrayFieldItemTemplate(props) {
    const { children, buttonsProps, displayLabel, hasDescription, hasToolbar, uiSchema, registry } = props;
    const uiOptions = utils.getUiOptions(uiSchema);
    const ArrayFieldItemButtonsTemplate = utils.getTemplate(
      "ArrayFieldItemButtonsTemplate",
      registry,
      uiOptions
    );
    const margin = hasDescription ? -6 : 22;
    return /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "mb-2 flex flex-row flex-wrap items-center", children: [
      /* @__PURE__ */ jsxRuntime.jsx("div", { className: "grow shrink", children }),
      /* @__PURE__ */ jsxRuntime.jsx("div", { className: "flex items-end justify-end p-0.5", children: hasToolbar && /* @__PURE__ */ jsxRuntime.jsx(
        "div",
        {
          className: "flex gap-2",
          style: { marginLeft: "5px", marginTop: displayLabel ? `${margin}px` : void 0 },
          children: /* @__PURE__ */ jsxRuntime.jsx(ArrayFieldItemButtonsTemplate, { ...buttonsProps })
        }
      ) })
    ] }) });
  }
  function ArrayFieldTemplate(props) {
    const {
      canAdd,
      disabled,
      fieldPathId,
      uiSchema,
      items,
      optionalDataControl,
      onAddClick,
      readonly,
      registry,
      required,
      schema,
      title
    } = props;
    const uiOptions = utils.getUiOptions(uiSchema);
    const ArrayFieldDescriptionTemplate = utils.getTemplate(
      "ArrayFieldDescriptionTemplate",
      registry,
      uiOptions
    );
    const ArrayFieldTitleTemplate = utils.getTemplate(
      "ArrayFieldTitleTemplate",
      registry,
      uiOptions
    );
    const showOptionalDataControlInTitle = !readonly && !disabled;
    const {
      ButtonTemplates: { AddButton: AddButton2 }
    } = registry.templates;
    return /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx("div", { className: "m-0 flex p-0", children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "m-0 w-full p-0", children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        ArrayFieldTitleTemplate,
        {
          fieldPathId,
          title: uiOptions.title || title,
          schema,
          uiSchema,
          required,
          registry,
          optionalDataControl: showOptionalDataControlInTitle ? optionalDataControl : void 0
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        ArrayFieldDescriptionTemplate,
        {
          fieldPathId,
          description: uiOptions.description || schema.description,
          schema,
          uiSchema,
          registry
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "p-0 m-0 w-full mb-2", children: [
        !showOptionalDataControlInTitle ? optionalDataControl : void 0,
        items,
        canAdd && /* @__PURE__ */ jsxRuntime.jsx("div", { className: "mt-2 flex justify-end", children: /* @__PURE__ */ jsxRuntime.jsx(
          AddButton2,
          {
            id: utils.buttonId(fieldPathId, "add"),
            className: "rjsf-array-item-add",
            onClick: onAddClick,
            disabled: disabled || readonly,
            uiSchema,
            registry
          }
        ) })
      ] }, `array-item-list-${fieldPathId.$id}`)
    ] }) }) });
  }
  function Input({ className, type, ...props }) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      "input",
      {
        type,
        "data-slot": "input",
        className: cn(
          "file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm",
          "focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px]",
          "aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive",
          className
        ),
        ...props
      }
    );
  }
  function BaseInputTemplate({
    id,
    htmlName,
    placeholder,
    required,
    readonly,
    disabled,
    type,
    value,
    onChange,
    onChangeOverride,
    onBlur,
    onFocus,
    autofocus,
    options,
    schema,
    rawErrors = [],
    children,
    extraProps,
    className
  }) {
    const inputProps = {
      ...extraProps,
      ...utils.getInputProps(schema, type, options)
    };
    const _onChange = ({ target: { value: value2 } }) => onChange(value2 === "" ? options.emptyValue : value2);
    const _onBlur = ({ target }) => onBlur(id, target && target.value);
    const _onFocus = ({ target }) => onFocus(id, target && target.value);
    return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "p-0.5", children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        Input,
        {
          id,
          name: htmlName || id,
          type,
          placeholder,
          autoFocus: autofocus,
          required,
          disabled,
          readOnly: readonly,
          className: cn({ "border-destructive focus-visible:ring-0": rawErrors.length > 0 }, className),
          list: schema.examples ? utils.examplesId(id) : void 0,
          ...inputProps,
          value: value || value === 0 ? value : "",
          onChange: onChangeOverride || _onChange,
          onBlur: _onBlur,
          onFocus: _onFocus,
          "aria-describedby": utils.ariaDescribedByIds(id, !!schema.examples)
        }
      ),
      children,
      Array.isArray(schema.examples) ? /* @__PURE__ */ jsxRuntime.jsx("datalist", { id: utils.examplesId(id), children: schema.examples.concat(schema.default && !schema.examples.includes(schema.default) ? [schema.default] : []).map((example) => {
        return /* @__PURE__ */ jsxRuntime.jsx("option", { value: example }, example);
      }) }) : null
    ] });
  }
  function DescriptionField({ id, description, registry, uiSchema }) {
    if (!description) {
      return null;
    }
    return /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx("div", { id, className: "text-sm text-muted-foreground", children: /* @__PURE__ */ jsxRuntime.jsx(core.RichDescription, { description, registry, uiSchema }) }) });
  }
  var alertVariants = classVarianceAuthority.cva(
    "relative w-full rounded-lg border px-4 py-3 text-sm grid has-[>svg]:grid-cols-[calc(var(--spacing)*4)_1fr] grid-cols-[0_1fr] has-[>svg]:gap-x-3 gap-y-0.5 items-start [&>svg]:size-4 [&>svg]:translate-y-0.5 [&>svg]:text-current",
    {
      variants: {
        variant: {
          default: "bg-card text-card-foreground",
          destructive: "text-destructive bg-card [&>svg]:text-current *:data-[slot=alert-description]:text-destructive/90"
        }
      },
      defaultVariants: {
        variant: "default"
      }
    }
  );
  function Alert({ className, variant, ...props }) {
    return /* @__PURE__ */ jsxRuntime.jsx("div", { "data-slot": "alert", role: "alert", className: cn(alertVariants({ variant }), className), ...props });
  }
  function AlertTitle({ className, ...props }) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        "data-slot": "alert-title",
        className: cn("col-start-2 line-clamp-1 min-h-4 font-medium tracking-tight", className),
        ...props
      }
    );
  }
  function AlertDescription({ className, ...props }) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        "data-slot": "alert-description",
        className: cn(
          "text-muted-foreground col-start-2 grid justify-items-start gap-1 text-sm [&_p]:leading-relaxed",
          className
        ),
        ...props
      }
    );
  }
  function ErrorList({
    errors,
    registry
  }) {
    const { translateString } = registry;
    return /* @__PURE__ */ jsxRuntime.jsxs(Alert, { variant: "destructive", className: "mb-2", children: [
      /* @__PURE__ */ jsxRuntime.jsx(lucideReact.AlertCircle, { className: "h-4 w-4" }),
      /* @__PURE__ */ jsxRuntime.jsx(AlertTitle, { children: translateString(utils.TranslatableString.ErrorsLabel) }),
      /* @__PURE__ */ jsxRuntime.jsx(AlertDescription, { className: "flex flex-col gap-1", children: errors.map((error, i) => {
        return /* @__PURE__ */ jsxRuntime.jsxs("span", { children: [
          "\u2022 ",
          error.stack
        ] }, i);
      }) })
    ] });
  }
  function FieldErrorTemplate(props) {
    const { errors = [], fieldPathId } = props;
    if (errors.length === 0) {
      return null;
    }
    const id = utils.errorId(fieldPathId);
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "flex flex-col gap-1", id, children: errors.map((error, i) => {
      return /* @__PURE__ */ jsxRuntime.jsx("span", { className: "text-xs font-medium text-destructive mb-1", children: error }, i);
    }) });
  }
  function FieldHelpTemplate(props) {
    const { fieldPathId, help, uiSchema, registry, hasErrors } = props;
    if (!help) {
      return null;
    }
    return /* @__PURE__ */ jsxRuntime.jsx(
      "span",
      {
        className: cn("text-xs font-medium text-muted-foreground", { "text-destructive": hasErrors }),
        id: utils.helpId(fieldPathId),
        children: /* @__PURE__ */ jsxRuntime.jsx(core.RichHelp, { help, registry, uiSchema })
      }
    );
  }
  function FieldTemplate({
    id,
    children,
    displayLabel,
    rawErrors = [],
    errors,
    help,
    description,
    rawDescription,
    classNames,
    style,
    disabled,
    label,
    hidden,
    onKeyRename,
    onKeyRenameBlur,
    onRemoveProperty,
    readonly,
    required,
    schema,
    uiSchema,
    registry
  }) {
    const uiOptions = utils.getUiOptions(uiSchema);
    const WrapIfAdditionalTemplate2 = utils.getTemplate(
      "WrapIfAdditionalTemplate",
      registry,
      uiOptions
    );
    if (hidden) {
      return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "hidden", children });
    }
    const isCheckbox = uiOptions.widget === "checkbox";
    return /* @__PURE__ */ jsxRuntime.jsx(
      WrapIfAdditionalTemplate2,
      {
        classNames,
        style,
        disabled,
        id,
        label,
        displayLabel,
        onKeyRename,
        onKeyRenameBlur,
        onRemoveProperty,
        rawDescription,
        readonly,
        required,
        schema,
        uiSchema,
        registry,
        children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex flex-col gap-2", children: [
          displayLabel && !isCheckbox && /* @__PURE__ */ jsxRuntime.jsxs(
            "label",
            {
              className: cn(
                "text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70",
                { " text-destructive": rawErrors.length > 0 }
              ),
              htmlFor: id,
              children: [
                label,
                required ? "*" : null
              ]
            }
          ),
          children,
          displayLabel && rawDescription && !isCheckbox && /* @__PURE__ */ jsxRuntime.jsx(
            "span",
            {
              className: cn("text-xs font-medium text-muted-foreground", { " text-destructive": rawErrors.length > 0 }),
              children: description
            }
          ),
          errors,
          help
        ] })
      }
    );
  }
  function GridTemplate(props) {
    const { children, column, className, ...rest } = props;
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: cn("grid gap-2", className), ...rest, children });
  }
  function IconButton(props) {
    const { icon, iconType, className, uiSchema, registry, ...otherProps } = props;
    return /* @__PURE__ */ jsxRuntime.jsx(Button, { size: "icon", variant: "outline", className, ...otherProps, type: "button", children: icon });
  }
  function CopyButton(props) {
    const {
      registry: { translateString }
    } = props;
    return /* @__PURE__ */ jsxRuntime.jsx(IconButton, { title: translateString(utils.TranslatableString.CopyButton), ...props, icon: /* @__PURE__ */ jsxRuntime.jsx(lucideReact.Copy, { className: "h-4 w-4" }) });
  }
  function MoveDownButton(props) {
    const {
      registry: { translateString }
    } = props;
    return /* @__PURE__ */ jsxRuntime.jsx(
      IconButton,
      {
        title: translateString(utils.TranslatableString.MoveDownButton),
        ...props,
        icon: /* @__PURE__ */ jsxRuntime.jsx(lucideReact.ChevronDown, { className: "h-4 w-4" })
      }
    );
  }
  function MoveUpButton(props) {
    const {
      registry: { translateString }
    } = props;
    return /* @__PURE__ */ jsxRuntime.jsx(
      IconButton,
      {
        title: translateString(utils.TranslatableString.MoveUpButton),
        ...props,
        icon: /* @__PURE__ */ jsxRuntime.jsx(lucideReact.ChevronUp, { className: "h-4 w-4" })
      }
    );
  }
  function RemoveButton(props) {
    const {
      registry: { translateString }
    } = props;
    return /* @__PURE__ */ jsxRuntime.jsx(
      IconButton,
      {
        title: translateString(utils.TranslatableString.RemoveButton),
        ...props,
        className: "border-destructive",
        icon: /* @__PURE__ */ jsxRuntime.jsx(lucideReact.Trash2, { className: "h-4 w-4 stroke-destructive" })
      }
    );
  }
  function MultiSchemaFieldTemplate({ selector, optionSchemaField }) {
    return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: cn("p-4 border rounded-md bg-background shadow-sm"), children: [
      /* @__PURE__ */ jsxRuntime.jsx("div", { className: cn("mb-4"), children: selector }),
      optionSchemaField
    ] });
  }
  function ObjectFieldTemplate({
    description,
    title,
    properties,
    required,
    uiSchema,
    fieldPathId,
    schema,
    formData,
    optionalDataControl,
    onAddProperty,
    disabled,
    readonly,
    registry
  }) {
    const uiOptions = utils.getUiOptions(uiSchema);
    const TitleFieldTemplate = utils.getTemplate("TitleFieldTemplate", registry, uiOptions);
    const DescriptionFieldTemplate = utils.getTemplate(
      "DescriptionFieldTemplate",
      registry,
      uiOptions
    );
    const showOptionalDataControlInTitle = !readonly && !disabled;
    const {
      ButtonTemplates: { AddButton: AddButton2 }
    } = registry.templates;
    return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
      title && /* @__PURE__ */ jsxRuntime.jsx(
        TitleFieldTemplate,
        {
          id: utils.titleId(fieldPathId),
          title,
          required,
          schema,
          uiSchema,
          registry,
          optionalDataControl: showOptionalDataControlInTitle ? optionalDataControl : void 0
        }
      ),
      description && /* @__PURE__ */ jsxRuntime.jsx(
        DescriptionFieldTemplate,
        {
          id: utils.descriptionId(fieldPathId),
          description,
          schema,
          uiSchema,
          registry
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex flex-col gap-2", children: [
        !showOptionalDataControlInTitle ? optionalDataControl : void 0,
        properties.map((element, index) => /* @__PURE__ */ jsxRuntime.jsx("div", { className: `${element.hidden ? "hidden" : ""} flex`, children: /* @__PURE__ */ jsxRuntime.jsx("div", { className: "w-full", children: element.content }) }, index)),
        utils.canExpand(schema, uiSchema, formData) ? /* @__PURE__ */ jsxRuntime.jsx("div", { className: "mt-2 flex justify-end", children: /* @__PURE__ */ jsxRuntime.jsx(
          AddButton2,
          {
            id: utils.buttonId(fieldPathId, "add"),
            onClick: onAddProperty,
            disabled: disabled || readonly,
            className: "rjsf-object-property-expand",
            uiSchema,
            registry
          }
        ) }) : null
      ] })
    ] });
  }
  function OptionalDataControlsTemplate(props) {
    const { id, registry, label, onAddClick, onRemoveClick } = props;
    if (onAddClick) {
      return /* @__PURE__ */ jsxRuntime.jsx(
        IconButton,
        {
          id,
          registry,
          className: "rjsf-add-optional-data",
          onClick: onAddClick,
          title: label,
          icon: /* @__PURE__ */ jsxRuntime.jsx(lucideReact.PlusCircle, {}),
          size: "xs"
        }
      );
    } else if (onRemoveClick) {
      return /* @__PURE__ */ jsxRuntime.jsx(
        RemoveButton,
        {
          id,
          registry,
          className: "rjsf-remove-optional-data",
          onClick: onRemoveClick,
          title: label,
          size: "xs"
        }
      );
    }
    return /* @__PURE__ */ jsxRuntime.jsx("em", { id, children: label });
  }
  function SubmitButton(props) {
    const { submitText, norender, props: submitButtonProps } = utils.getSubmitButtonOptions(props.uiSchema);
    if (norender) {
      return null;
    }
    return /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx(Button, { type: "submit", ...submitButtonProps, className: cn("my-2", submitButtonProps?.className), children: submitText }) });
  }
  function Separator({
    className,
    orientation = "horizontal",
    decorative = true,
    ...props
  }) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      reactSeparator.Root,
      {
        "data-slot": "separator",
        decorative,
        orientation,
        className: cn(
          "bg-border shrink-0 data-[orientation=horizontal]:h-px data-[orientation=horizontal]:w-full data-[orientation=vertical]:h-full data-[orientation=vertical]:w-px",
          className
        ),
        ...props
      }
    );
  }
  function TitleField({
    id,
    title,
    uiSchema,
    optionalDataControl
  }) {
    const uiOptions = utils.getUiOptions(uiSchema);
    let heading = /* @__PURE__ */ jsxRuntime.jsx("h5", { children: uiOptions.title || title });
    if (optionalDataControl) {
      heading = /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex flex-row", children: [
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "flex-grow", children: heading }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "flex", children: optionalDataControl })
      ] });
    }
    return /* @__PURE__ */ jsxRuntime.jsxs("div", { id, className: "my-1 flex flex-col gap-0.5", children: [
      heading,
      /* @__PURE__ */ jsxRuntime.jsx(Separator, { dir: "horizontal", style: { height: "1px" }, className: "my-1" })
    ] });
  }
  function WrapIfAdditionalTemplate({
    classNames,
    style,
    children,
    disabled,
    id,
    label,
    displayLabel,
    onRemoveProperty,
    onKeyRenameBlur,
    rawDescription,
    readonly,
    required,
    schema,
    uiSchema,
    registry
  }) {
    const { templates, translateString } = registry;
    const { RemoveButton: RemoveButton2 } = templates.ButtonTemplates;
    const keyLabel = translateString(utils.TranslatableString.KeyLabel, [label]);
    const additional = utils.ADDITIONAL_PROPERTY_FLAG in schema;
    if (!additional) {
      return /* @__PURE__ */ jsxRuntime.jsx("div", { className: classNames, style, children });
    }
    const marginDesc = rawDescription ? -28 : 0;
    const margin = displayLabel ? 22 + marginDesc : 0;
    const keyId = `${id}-key`;
    return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
      /* @__PURE__ */ jsxRuntime.jsxs("div", { className: `grid grid-cols-12 col-span-12 items-center gap-2 ${classNames}`, style, children: [
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "grid gap-2 col-span-5", children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex flex-col gap-2", children: [
          displayLabel && /* @__PURE__ */ jsxRuntime.jsx("label", { htmlFor: keyId, className: "text-sm font-medium text-muted-foreground leading-none", children: keyLabel }),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "pl-0.5", children: /* @__PURE__ */ jsxRuntime.jsx(
            Input,
            {
              required,
              defaultValue: label,
              disabled: disabled || readonly,
              id: keyId,
              name: keyId,
              onBlur: !readonly ? onKeyRenameBlur : void 0,
              type: "text",
              className: "w-full border shadow-sm"
            }
          ) }),
          !!rawDescription && /* @__PURE__ */ jsxRuntime.jsx("span", { className: "text-xs font-medium text-muted-foreground", children: /* @__PURE__ */ jsxRuntime.jsx("div", { className: "text-sm text-muted-foreground", children: "\xA0" }) })
        ] }) }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "grid gap-2 col-span-6 pr-0.5", children }),
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: "grid gap-2 col-span-1", style: { marginTop: `${margin}px` }, children: /* @__PURE__ */ jsxRuntime.jsx(
          RemoveButton2,
          {
            id: utils.buttonId(id, "remove"),
            iconType: "block",
            className: "rjsf-object-property-remove w-full",
            disabled: disabled || readonly,
            onClick: onRemoveProperty,
            uiSchema,
            registry
          }
        ) })
      ] }),
      /* @__PURE__ */ jsxRuntime.jsx(Separator, { dir: "horizontal", className: "mt-2" })
    ] });
  }

  // src/Templates/Templates.ts
  function generateTemplates() {
    return {
      ArrayFieldItemTemplate,
      ArrayFieldTemplate,
      BaseInputTemplate,
      ButtonTemplates: {
        AddButton,
        CopyButton,
        MoveDownButton,
        MoveUpButton,
        RemoveButton,
        SubmitButton
      },
      DescriptionFieldTemplate: DescriptionField,
      ErrorListTemplate: ErrorList,
      FieldErrorTemplate,
      FieldHelpTemplate,
      FieldTemplate,
      GridTemplate,
      MultiSchemaFieldTemplate,
      ObjectFieldTemplate,
      OptionalDataControlsTemplate,
      TitleFieldTemplate: TitleField,
      WrapIfAdditionalTemplate
    };
  }
  var Templates_default = generateTemplates();
  function Checkbox({ className, ...props }) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      reactCheckbox.Root,
      {
        "data-slot": "checkbox",
        className: cn(
          "peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50",
          className
        ),
        ...props,
        children: /* @__PURE__ */ jsxRuntime.jsx(
          reactCheckbox.Indicator,
          {
            "data-slot": "checkbox-indicator",
            className: "flex items-center justify-center text-current transition-none",
            children: /* @__PURE__ */ jsxRuntime.jsx(reactIcons.CheckIcon, { className: "size-3.5" })
          }
        )
      }
    );
  }
  function Label({ className, ...props }) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      reactLabel.Root,
      {
        "data-slot": "label",
        className: cn(
          "flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50",
          className
        ),
        ...props
      }
    );
  }
  function CheckboxWidget(props) {
    const {
      id,
      htmlName,
      value,
      disabled,
      readonly,
      label,
      hideLabel,
      schema,
      autofocus,
      options,
      onChange,
      onBlur,
      onFocus,
      registry,
      uiSchema,
      className
    } = props;
    const required = utils.schemaRequiresTrueValue(schema);
    const DescriptionFieldTemplate = utils.getTemplate(
      "DescriptionFieldTemplate",
      registry,
      options
    );
    const _onChange = (checked) => onChange(checked);
    const _onBlur = () => onBlur(id, value);
    const _onFocus = () => onFocus(id, value);
    const description = options.description || schema.description;
    return /* @__PURE__ */ jsxRuntime.jsxs(
      "div",
      {
        className: `relative ${disabled || readonly ? "cursor-not-allowed opacity-50" : ""}`,
        "aria-describedby": utils.ariaDescribedByIds(id),
        children: [
          !hideLabel && description && /* @__PURE__ */ jsxRuntime.jsx(
            DescriptionFieldTemplate,
            {
              id: utils.descriptionId(id),
              description,
              schema,
              uiSchema,
              registry
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex items-center gap-2 my-2", children: [
            /* @__PURE__ */ jsxRuntime.jsx(
              Checkbox,
              {
                id,
                name: htmlName || id,
                checked: typeof value === "undefined" ? false : Boolean(value),
                required,
                disabled: disabled || readonly,
                autoFocus: autofocus,
                onCheckedChange: _onChange,
                onBlur: _onBlur,
                onFocus: _onFocus,
                className
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(Label, { className: "leading-tight", htmlFor: id, children: utils.labelValue(label, hideLabel || !label) })
          ] })
        ]
      }
    );
  }
  function CheckboxesWidget({
    id,
    htmlName,
    disabled,
    options,
    value,
    autofocus,
    readonly,
    required,
    onChange,
    onBlur,
    onFocus,
    className
  }) {
    const { enumOptions, enumDisabled, inline, emptyValue } = options;
    const checkboxesValues = Array.isArray(value) ? value : [value];
    const _onBlur = ({ target }) => onBlur(id, utils.enumOptionsValueForIndex(target && target.value, enumOptions, emptyValue));
    const _onFocus = ({ target }) => onFocus(id, utils.enumOptionsValueForIndex(target && target.value, enumOptions, emptyValue));
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: cn({ "flex flex-col gap-2": !inline, "flex flex-row gap-4 flex-wrap": inline }), children: Array.isArray(enumOptions) && enumOptions.map((option, index) => {
      const checked = utils.enumOptionsIsSelected(option.value, checkboxesValues);
      const itemDisabled = Array.isArray(enumDisabled) && enumDisabled.indexOf(option.value) !== -1;
      const indexOptionId = utils.optionId(id, index);
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex items-center gap-2", children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          Checkbox,
          {
            id: indexOptionId,
            name: htmlName || id,
            required,
            disabled: disabled || itemDisabled || readonly,
            onCheckedChange: (state) => {
              if (state) {
                onChange(utils.enumOptionsSelectValue(index, checkboxesValues, enumOptions));
              } else {
                onChange(utils.enumOptionsDeselectValue(index, checkboxesValues, enumOptions));
              }
            },
            className,
            checked,
            autoFocus: autofocus && index === 0,
            onBlur: _onBlur,
            onFocus: _onFocus,
            "aria-describedby": utils.ariaDescribedByIds(id)
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(Label, { className: "leading-tight", htmlFor: utils.optionId(id, index), children: option.label })
      ] }, indexOptionId);
    }) });
  }
  function RadioGroup({ className, ...props }) {
    return /* @__PURE__ */ jsxRuntime.jsx(reactRadioGroup.Root, { "data-slot": "radio-group", className: cn("grid gap-3", className), ...props });
  }
  function RadioGroupItem({ className, ...props }) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      reactRadioGroup.Item,
      {
        "data-slot": "radio-group-item",
        className: cn(
          "border-input text-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 aspect-square size-4 shrink-0 rounded-full border shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50",
          className
        ),
        ...props,
        children: /* @__PURE__ */ jsxRuntime.jsx(reactRadioGroup.Indicator, { "data-slot": "radio-group-indicator", className: "relative flex items-center justify-center", children: /* @__PURE__ */ jsxRuntime.jsx(lucideReact.CircleIcon, { className: "fill-primary absolute top-1/2 left-1/2 size-2 -translate-x-1/2 -translate-y-1/2" }) })
      }
    );
  }
  function RadioWidget({
    id,
    options,
    value,
    required,
    disabled,
    readonly,
    onChange,
    onBlur,
    onFocus,
    className
  }) {
    const { enumOptions, enumDisabled, emptyValue } = options;
    const _onChange = (value2) => onChange(utils.enumOptionsValueForIndex(value2, enumOptions, emptyValue));
    const _onBlur = ({ target }) => onBlur(id, utils.enumOptionsValueForIndex(target && target.value, enumOptions, emptyValue));
    const _onFocus = ({ target }) => onFocus(id, utils.enumOptionsValueForIndex(target && target.value, enumOptions, emptyValue));
    const inline = Boolean(options && options.inline);
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "mb-0", children: /* @__PURE__ */ jsxRuntime.jsx(
      RadioGroup,
      {
        defaultValue: value?.toString(),
        required,
        disabled: disabled || readonly,
        onValueChange: (e) => {
          _onChange(e);
        },
        onBlur: _onBlur,
        onFocus: _onFocus,
        "aria-describedby": utils.ariaDescribedByIds(id),
        orientation: inline ? "horizontal" : "vertical",
        className: cn("flex flex-wrap", { "flex-col": !inline }, className),
        children: Array.isArray(enumOptions) && enumOptions.map((option, index) => {
          const itemDisabled = Array.isArray(enumDisabled) && enumDisabled.indexOf(option.value) !== -1;
          const checked = utils.enumOptionsIsSelected(option.value, value);
          return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex items-center gap-2", children: [
            /* @__PURE__ */ jsxRuntime.jsx(
              RadioGroupItem,
              {
                checked,
                value: index.toString(),
                id: utils.optionId(id, index),
                disabled: itemDisabled
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(Label, { className: "leading-tight", htmlFor: utils.optionId(id, index), children: option.label })
          ] }, utils.optionId(id, index));
        })
      }
    ) });
  }
  function Slider({ className, defaultValue, value, min = 0, max = 100, ...props }) {
    const _values = react.useMemo(
      () => Array.isArray(value) ? value : Array.isArray(defaultValue) ? defaultValue : [min, max],
      [value, defaultValue, min, max]
    );
    return /* @__PURE__ */ jsxRuntime.jsxs(
      reactSlider.Root,
      {
        "data-slot": "slider",
        defaultValue,
        value,
        min,
        max,
        className: cn(
          "relative flex w-full touch-none items-center select-none data-[disabled]:opacity-50 data-[orientation=vertical]:h-full data-[orientation=vertical]:min-h-44 data-[orientation=vertical]:w-auto data-[orientation=vertical]:flex-col",
          className
        ),
        ...props,
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            reactSlider.Track,
            {
              "data-slot": "slider-track",
              className: cn(
                "bg-muted relative grow overflow-hidden rounded-full data-[orientation=horizontal]:h-1.5 data-[orientation=horizontal]:w-full data-[orientation=vertical]:h-full data-[orientation=vertical]:w-1.5"
              ),
              children: /* @__PURE__ */ jsxRuntime.jsx(
                reactSlider.Range,
                {
                  "data-slot": "slider-range",
                  className: cn("bg-primary absolute data-[orientation=horizontal]:h-full data-[orientation=vertical]:w-full")
                }
              )
            }
          ),
          Array.from({ length: _values.length }, (_, index) => /* @__PURE__ */ jsxRuntime.jsx(
            reactSlider.Thumb,
            {
              "data-slot": "slider-thumb",
              className: "border-primary bg-background ring-ring/50 block size-4 shrink-0 rounded-full border shadow-sm transition-[color,box-shadow] hover:ring-4 focus-visible:ring-4 focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50"
            },
            index
          ))
        ]
      }
    );
  }
  var allowedProps = [
    "name",
    "min",
    "max",
    "step",
    "orientation",
    "disabled",
    "defaultValue",
    "value",
    "onValueChange",
    "className",
    "dir",
    "inverted",
    "minStepsBetweenThumbs"
  ];
  function RangeWidget({
    value,
    readonly,
    disabled,
    options,
    schema,
    onChange,
    label,
    id
  }) {
    const _onChange = (value2) => onChange(value2[0]);
    const sliderProps = { ...utils.rangeSpec(schema) };
    const uiProps = { id, ..._pick(options.props || {}, allowedProps) };
    return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        Slider,
        {
          disabled: disabled || readonly,
          min: sliderProps.min,
          max: sliderProps.max,
          step: sliderProps.step,
          value: [value],
          onValueChange: _onChange,
          ...uiProps,
          "aria-describedby": utils.ariaDescribedByIds(id)
        }
      ),
      value
    ] });
  }
  var badgeVariants = classVarianceAuthority.cva(
    "inline-flex items-center justify-center rounded-md border px-2 py-0.5 text-xs font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden",
    {
      variants: {
        variant: {
          default: "border-transparent bg-primary text-primary-foreground [a&]:hover:bg-primary/90",
          secondary: "border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90",
          destructive: "border-transparent bg-destructive text-white [a&]:hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60",
          outline: "text-foreground [a&]:hover:bg-accent [a&]:hover:text-accent-foreground"
        }
      },
      defaultVariants: {
        variant: "default"
      }
    }
  );
  function Badge({
    className,
    variant,
    asChild = false,
    ...props
  }) {
    const Comp = asChild ? reactSlot.Slot : "span";
    return /* @__PURE__ */ jsxRuntime.jsx(Comp, { "data-slot": "badge", className: cn(badgeVariants({ variant }), className), ...props });
  }
  var Command = react.forwardRef(
    ({ className, ...props }, ref) => {
      return /* @__PURE__ */ jsxRuntime.jsx(
        cmdk.Command,
        {
          ref,
          "data-slot": "command",
          className: cn(
            "bg-popover text-popover-foreground flex h-full w-full flex-col overflow-hidden rounded-md",
            className
          ),
          ...props
        }
      );
    }
  );
  var CommandInput = react.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex items-center border-b px-3", ...{ "cmdk-input-wrapper": "" }, children: [
    /* @__PURE__ */ jsxRuntime.jsx(lucideReact.Search, { className: "mr-2 h-4 w-4 shrink-0 opacity-50" }),
    /* @__PURE__ */ jsxRuntime.jsx(
      cmdk.Command.Input,
      {
        ref,
        className: cn(
          "flex h-10 w-full rounded-md bg-transparent py-3 text-sm outline-none placeholder:text-muted-foreground disabled:cursor-not-allowed disabled:opacity-50",
          className
        ),
        ...props
      }
    )
  ] }));
  CommandInput.displayName = cmdk.Command.Input.displayName;
  function CommandList({ className, ...props }) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      cmdk.Command.List,
      {
        "data-slot": "command-list",
        className: cn("max-h-[300px] scroll-py-1 overflow-x-hidden overflow-y-auto", className),
        ...props
      }
    );
  }
  function CommandGroup({ className, ...props }) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      cmdk.Command.Group,
      {
        "data-slot": "command-group",
        className: cn(
          "text-foreground [&_[cmdk-group-heading]]:text-muted-foreground overflow-hidden p-1 [&_[cmdk-group-heading]]:px-2 [&_[cmdk-group-heading]]:py-1.5 [&_[cmdk-group-heading]]:text-xs [&_[cmdk-group-heading]]:font-medium",
          className
        ),
        ...props
      }
    );
  }
  function CommandItem({ className, ...props }) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      cmdk.Command.Item,
      {
        "data-slot": "command-item",
        className: cn(
          "data-[selected=true]:bg-accent data-[selected=true]:text-accent-foreground [&_svg:not([class*='text-'])]:text-muted-foreground relative flex cursor-default items-center gap-2 rounded-sm px-2 py-1.5 text-sm outline-hidden select-none data-[disabled=true]:pointer-events-none data-[disabled=true]:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
          className
        ),
        ...props
      }
    );
  }
  function FancyMultiSelect({
    multiple,
    items = [],
    selected,
    onValueChange,
    autoFocus = false,
    disabled = false,
    ariaDescribedby,
    ariaPlaceholder,
    onFocus,
    onBlur,
    className,
    id
  }) {
    const inputRef = react.useRef(null);
    const [open, setOpen] = react.useState(false);
    const [inputValue, setInputValue] = react.useState("");
    const selectedItems = react.useMemo(
      () => items.filter((item) => selected.some((selectedValue) => isEqual(item.value, selectedValue))),
      [items, selected]
    );
    const selectables = react.useMemo(
      () => items.filter((framework) => !selectedItems.some((item) => isEqual(item.value, framework.value))),
      [items, selectedItems]
    );
    const handleUnselect = react.useCallback(
      (framework) => {
        if (disabled) {
          return;
        }
        const newSelected = selectedItems.filter((s) => !isEqual(s.value, framework.value));
        onValueChange?.(newSelected.map((item) => item.index));
      },
      [selectedItems, onValueChange, disabled]
    );
    const handleKeyDown = react.useCallback(
      (e) => {
        if (disabled || !inputRef.current || inputRef.current.value !== "") {
          return;
        }
        if (e.key === "Delete" || e.key === "Backspace") {
          const newSelected = selectedItems.slice(0, -1);
          onValueChange?.(newSelected.map((item) => item.index));
        } else if (e.key === "Escape") {
          inputRef.current.blur();
        }
      },
      [selectedItems, onValueChange, disabled]
    );
    const handleSelect = react.useCallback(
      (item) => {
        if (disabled) {
          return;
        }
        setInputValue("");
        const newSelected = multiple ? [...selectedItems, item] : [item];
        onValueChange?.(newSelected.map((item2) => item2.index));
      },
      [multiple, selectedItems, onValueChange, disabled]
    );
    const handleFocus = react.useCallback(
      (e) => {
        if (!disabled) {
          setOpen(true);
        }
        onFocus?.(e);
      },
      [disabled, onFocus]
    );
    return /* @__PURE__ */ jsxRuntime.jsxs(
      Command,
      {
        onKeyDown: handleKeyDown,
        className: cn("overflow-visible bg-transparent", className),
        autoFocus,
        "aria-disabled": disabled,
        onBlur,
        onFocus: handleFocus,
        "aria-describedby": ariaDescribedby,
        "aria-placeholder": ariaPlaceholder,
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            "div",
            {
              className: cn(
                "group border border-input px-3 py-2 text-sm ring-offset-background rounded-md focus-within:ring-1 focus-within:ring-ring focus-within:ring-offset-1",
                disabled && "opacity-50 cursor-not-allowed"
              ),
              children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex gap-1 flex-wrap", children: [
                selectedItems.map((item) => /* @__PURE__ */ jsxRuntime.jsxs(Badge, { variant: "secondary", children: [
                  item.label,
                  /* @__PURE__ */ jsxRuntime.jsx(
                    "button",
                    {
                      type: "button",
                      className: "rtl:mr-1 ltr:ml-1 ring-offset-background rounded-full outline-none focus:ring-1 focus:ring-ring focus:ring-offset-1",
                      onKeyDown: (e) => e.key === "Enter" && !disabled && handleUnselect(item),
                      onMouseDown: (e) => {
                        e.preventDefault();
                        e.stopPropagation();
                      },
                      onClick: () => handleUnselect(item),
                      disabled,
                      children: /* @__PURE__ */ jsxRuntime.jsx(
                        lucideReact.X,
                        {
                          className: cn(
                            "h-3 w-3 text-muted-foreground hover:text-foreground",
                            disabled && "pointer-events-none"
                          )
                        }
                      )
                    }
                  )
                ] }, item.value)),
                /* @__PURE__ */ jsxRuntime.jsx(
                  cmdk.Command.Input,
                  {
                    ref: inputRef,
                    value: inputValue,
                    onValueChange: setInputValue,
                    onBlur: () => setOpen(false),
                    onFocus: () => !disabled && setOpen(true),
                    placeholder: "Select ...",
                    className: "rtl:mr-2 ltr:ml-2 bg-transparent outline-none placeholder:text-muted-foreground flex-1",
                    disabled,
                    "aria-controls": `command-item-input-${id}`,
                    "aria-labelledby": `command-item-input-${id}`,
                    id: `command-item-input-${id}`
                  }
                )
              ] })
            }
          ),
          open && !disabled && selectables.length > 0 && /* @__PURE__ */ jsxRuntime.jsx("div", { className: "relative mt-2", children: /* @__PURE__ */ jsxRuntime.jsx("div", { className: "absolute w-full z-10 top-0 rounded-md border bg-popover text-popover-foreground shadow-md outline-none animate-in", children: /* @__PURE__ */ jsxRuntime.jsx(CommandGroup, { className: "h-full overflow-auto", children: /* @__PURE__ */ jsxRuntime.jsx(CommandList, { children: selectables.map((item) => /* @__PURE__ */ jsxRuntime.jsx(
            CommandItem,
            {
              disabled: item.disabled,
              onMouseDown: (e) => {
                e.preventDefault();
                e.stopPropagation();
              },
              "aria-controls": `${item.value}-command-item`,
              "aria-labelledby": `${item.value}-command-item`,
              id: `${item.value}-command-item`,
              onSelect: () => handleSelect(item),
              className: "cursor-pointer",
              children: item.label
            },
            `${item.value}-command-item`
          )) }) }) }) })
        ]
      }
    );
  }
  function FancySelect({
    items,
    selected,
    onValueChange,
    autoFocus = false,
    disabled = false,
    required = false,
    placeholder = "Select...",
    ariaDescribedby,
    ariaPlaceholder,
    onFocus,
    onBlur,
    className
  }) {
    const [open, setOpen] = react.useState(false);
    const selectedItem = items?.find((item) => item.value === selected);
    const selectedRef = react.useRef(null);
    const containerRef = react.useRef(null);
    react.useEffect(() => {
      if (open && selectedRef.current) {
        requestAnimationFrame(() => {
          selectedRef.current?.scrollIntoView({ block: "nearest" });
        });
      }
    }, [open]);
    const handleBlur = (e) => {
      if (containerRef.current && !containerRef.current.contains(e.relatedTarget)) {
        setOpen(false);
      }
      onBlur?.(e);
    };
    return /* @__PURE__ */ jsxRuntime.jsxs(
      Command,
      {
        ref: containerRef,
        className: cn("overflow-visible bg-transparent", className),
        autoFocus,
        "aria-disabled": disabled,
        onBlur: handleBlur,
        onFocus,
        "aria-describedby": ariaDescribedby,
        "aria-placeholder": ariaPlaceholder,
        children: [
          /* @__PURE__ */ jsxRuntime.jsxs(
            "div",
            {
              onClick: () => !disabled && setOpen(!open),
              className: cn(
                "flex h-9 w-full items-center justify-between gap-2 whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50",
                !selectedItem && required && "border-red-500",
                disabled && "opacity-50 cursor-not-allowed",
                className
              ),
              children: [
                /* @__PURE__ */ jsxRuntime.jsx("span", { className: cn("flex-1 line-clamp-1", !selectedItem && "text-muted-foreground"), children: selectedItem?.label || placeholder }),
                /* @__PURE__ */ jsxRuntime.jsx(lucideReact.ChevronDown, { className: "h-4 w-4 opacity-50" })
              ]
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: "relative", children: open && items && items.length > 0 ? /* @__PURE__ */ jsxRuntime.jsx(
            "div",
            {
              style: { top: "0.5rem" },
              className: "absolute w-full z-10 rounded-md border bg-popover text-popover-foreground shadow-md outline-none",
              children: /* @__PURE__ */ jsxRuntime.jsx(CommandGroup, { className: "h-full overflow-auto", children: /* @__PURE__ */ jsxRuntime.jsx(CommandList, { children: items.map((item) => /* @__PURE__ */ jsxRuntime.jsxs(
                CommandItem,
                {
                  ref: item.value === selected ? selectedRef : void 0,
                  onMouseDown: (e) => {
                    e.preventDefault();
                    e.stopPropagation();
                  },
                  onSelect: () => {
                    if (!item.disabled) {
                      onValueChange?.(item.value);
                      setOpen(false);
                    }
                  },
                  className: cn(
                    "cursor-pointer relative flex items-center justify-between rounded-sm py-1.5 gap-2 rtl:flex-row-reverse",
                    item.value === selected && "font-semibold",
                    item.disabled && "opacity-50 cursor-not-allowed"
                  ),
                  children: [
                    /* @__PURE__ */ jsxRuntime.jsx("span", { children: item.label }),
                    /* @__PURE__ */ jsxRuntime.jsx("span", { className: "flex h-3.5 w-3.5 items-center justify-center", children: item.value === selected && /* @__PURE__ */ jsxRuntime.jsx(lucideReact.Check, { className: "h-4 w-4" }) })
                  ]
                },
                item.value
              )) }) })
            }
          ) : null })
        ]
      }
    );
  }
  function SelectWidget({
    id,
    options,
    required,
    disabled,
    readonly,
    value,
    multiple,
    autofocus,
    onChange,
    onBlur,
    onFocus,
    defaultValue,
    placeholder,
    rawErrors = [],
    className
  }) {
    const { enumOptions, enumDisabled, emptyValue: optEmptyValue } = options;
    const _onFancyFocus = () => {
      onFocus(id, utils.enumOptionsValueForIndex(value, enumOptions, optEmptyValue));
    };
    const _onFancyBlur = () => {
      onBlur(id, utils.enumOptionsValueForIndex(value, enumOptions, optEmptyValue));
    };
    const items = enumOptions?.map(({ value: value2, label }, index) => ({
      value: multiple ? value2 : index.toString(),
      label,
      index,
      disabled: Array.isArray(enumDisabled) && enumDisabled.includes(value2)
    }));
    const cnClassName = cn({ "border-destructive": rawErrors.length > 0 }, className);
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "p-0.5", children: !multiple ? /* @__PURE__ */ jsxRuntime.jsx(
      FancySelect,
      {
        items,
        selected: utils.enumOptionsIndexForValue(value ?? defaultValue, enumOptions, false),
        onValueChange: (selectedValue) => {
          onChange(utils.enumOptionsValueForIndex(selectedValue, enumOptions, optEmptyValue));
        },
        autoFocus: autofocus,
        disabled: disabled || readonly,
        required,
        placeholder,
        className: cnClassName,
        onFocus: _onFancyFocus,
        onBlur: _onFancyBlur,
        ariaDescribedby: utils.ariaDescribedByIds(id)
      }
    ) : /* @__PURE__ */ jsxRuntime.jsx(
      FancyMultiSelect,
      {
        id,
        autoFocus: autofocus,
        disabled: disabled || readonly,
        multiple: true,
        className: cnClassName,
        items,
        selected: value,
        onValueChange: (values) => {
          onChange(utils.enumOptionsValueForIndex(values, enumOptions, optEmptyValue));
        },
        onFocus: _onFancyFocus,
        onBlur: _onFancyBlur
      }
    ) });
  }
  function Textarea({ className, ...props }) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      "textarea",
      {
        "data-slot": "textarea",
        className: cn(
          "border-input placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 flex field-sizing-content min-h-16 w-full rounded-md border bg-transparent px-3 py-2 text-base shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 md:text-sm",
          className
        ),
        ...props
      }
    );
  }
  function TextareaWidget({
    id,
    htmlName,
    placeholder,
    value,
    required,
    disabled,
    autofocus,
    readonly,
    onBlur,
    onFocus,
    onChange,
    options,
    className
  }) {
    const _onChange = ({ target: { value: value2 } }) => onChange(value2 === "" ? options.emptyValue : value2);
    const _onBlur = ({ target }) => onBlur(id, target && target.value);
    const _onFocus = ({ target }) => onFocus(id, target && target.value);
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "flex p-0.5", children: /* @__PURE__ */ jsxRuntime.jsx(
      Textarea,
      {
        id,
        name: htmlName || id,
        placeholder,
        disabled,
        readOnly: readonly,
        value: value ?? "",
        required,
        autoFocus: autofocus,
        rows: options.rows || 5,
        onChange: _onChange,
        onBlur: _onBlur,
        onFocus: _onFocus,
        "aria-describedby": utils.ariaDescribedByIds(id),
        className
      }
    ) });
  }

  // src/Widgets/Widgets.ts
  function generateWidgets() {
    return {
      CheckboxWidget,
      CheckboxesWidget,
      RadioWidget,
      RangeWidget,
      SelectWidget,
      TextareaWidget
    };
  }
  var Widgets_default = generateWidgets();

  // src/Theme/Theme.tsx
  function generateTheme() {
    return {
      templates: generateTemplates(),
      widgets: generateWidgets()
    };
  }
  var Theme_default = generateTheme();

  // src/Form/Form.tsx
  function generateForm() {
    return core.withTheme(generateTheme());
  }
  var Form_default = generateForm();

  // src/index.ts
  var index_default = Form_default;

  exports.Form = Form_default;
  exports.Templates = Templates_default;
  exports.Theme = Theme_default;
  exports.Widgets = Widgets_default;
  exports.default = index_default;
  exports.generateForm = generateForm;
  exports.generateTemplates = generateTemplates;
  exports.generateTheme = generateTheme;
  exports.generateWidgets = generateWidgets;

  Object.defineProperty(exports, '__esModule', { value: true });

}));
